{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "cca26c9b-632f-4bed-a4bd-06b89cecf937",
   "metadata": {},
   "outputs": [],
   "source": [
    "import boto3\n",
    "import numpy as np\n",
    "import os\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import tarfile\n",
    "from cellxgene_schema.write_labels import AnnDataLabelAppender\n",
    "from urllib.parse import quote\n",
    "from cellxgene_ontology_guide.ontology_parser import OntologyParser\n",
    "ontology_parser = OntologyParser(schema_version=\"v6.0.0\")\n",
    "from cellxgene_ontology_guide.supported_versions import CXGSchema, load_supported_versions\n",
    "import requests\n",
    "from io import BytesIO\n",
    "\n",
    "\n",
    "s3client = boto3.client('s3')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "40f3f071-0b7c-4ef8-b7cb-49cdad9c153e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def download_files(s_dir):\n",
    "    mx_h5 = f'{s_dir}/count/sample_filtered_feature_bc_matrix.h5'\n",
    "    metrics_csv = f'{s_dir}/metrics_summary.csv'\n",
    "    cri_tar = f'{s_dir}/count/crispr_analysis.tar.gz'\n",
    "    \n",
    "    for file_path in [mx_h5, metrics_csv, cri_tar]:\n",
    "        f = file_path.split('/')[-1]\n",
    "        s3client.download_file(bucket_name, file_path, f)\n",
    "\n",
    "\n",
    "def custom_var_to_obs(adata):\n",
    "    moved = []\n",
    "    for gene_index in adata.var[\n",
    "        (adata.var['feature_types'] == 'Gene Expression') &\n",
    "        (adata.var['gene_ids'].str.startswith('ENS') == False)\n",
    "    ].index:\n",
    "        gene_id = adata.var.loc[gene_index]['gene_ids']\n",
    "        gene_values = adata.X[:, adata.var.index.get_loc(gene_index)]\n",
    "        adata.obs[gene_id] = gene_values.A.flatten() if hasattr(gene_values, 'A') else gene_values.toarray().flatten().tolist()\n",
    "        moved.append(gene_id)\n",
    "\n",
    "    adata.var.set_index('gene_ids', inplace=True)\n",
    "    var_to_keep = [i for i in adata.var.index if i not in moved]\n",
    "    adata = adata[:, var_to_keep]\n",
    "    adata.var = adata.var.replace('', np.nan).dropna(axis=1, how='all')\n",
    "\n",
    "    return adata\n",
    "\n",
    "\n",
    "def gather_crispr():\n",
    "    tar = tarfile.open('crispr_analysis.tar.gz', 'r:gz')\n",
    "    f = tar.extractfile('protospacer_calls_per_cell.csv')\n",
    "    df = pd.read_csv(f).rename(columns={'num_umis':'num_umis_guide_id'})\n",
    "    tar.close()\n",
    "    df['genetic_perturbation_guide_id'] = df['feature_call'].apply(lambda x: x.replace('|',' || '))\n",
    "    df = df[['cell_barcode','genetic_perturbation_guide_id','num_umis_guide_id']].set_index('cell_barcode')\n",
    "\n",
    "    return df\n",
    "\n",
    "\n",
    "def gather_metrics(samp):\n",
    "    df = pd.read_csv('metrics_summary.csv')\n",
    "    df['Metric'] = df.apply(lambda x: f\"{x['Metric Name']}, {x['Library Type']}, {x['Category'].replace('Cells','Sample')}\", axis=1)\n",
    "\n",
    "    probe_barcodes = ' || '.join(df[\n",
    "        (df['Metric Name'] == 'Sample ID') &\n",
    "        (df['Metric Value'] == samp)\n",
    "    ]['Group Name'].unique())\n",
    "\n",
    "    df = df[\n",
    "        (df['Grouped By'].isin(['Fastq ID','Probe barcode ID']) == False) &\n",
    "        (df['Category'].isin(['Library','Cells']))\n",
    "    ]\n",
    "    df = df[['Metric','Metric Value']].set_index('Metric').transpose()\n",
    "\n",
    "    for c in df.columns:\n",
    "        v = df[c].iloc[0]\n",
    "        if v.endswith('%'):\n",
    "            df[c] = df[c].apply(lambda x: float(x.rstrip('%')) / 100)\n",
    "        elif '%' in v:\n",
    "            df.drop(columns=c, inplace=True)\n",
    "        else:\n",
    "            df[c] = df[c].apply(lambda x: int(x.replace(',','')))\n",
    "\n",
    "    df['Probe barcode IDs'] = probe_barcodes\n",
    "\n",
    "    keep = {\n",
    "        'Confidently mapped reads in cells, Gene Expression, Sample': 'fraction_mapped_reads_in_cells_gex',\n",
    "        'Median genes per cell, Gene Expression, Sample': 'median_genes_per_cell',\n",
    "        'Cells, CRISPR Guide Capture, Sample': 'cell_count_cri',\n",
    "        'Cells, Gene Expression, Sample': 'cell_count_gex',\n",
    "        'Valid GEM barcodes, CRISPR Guide Capture, Library': 'fraction_valid_barcodes_cri',\n",
    "        'Valid GEM barcodes, Gene Expression, Library': 'fraction_valid_barcodes_gex',\n",
    "        'Reads confidently mapped to probe set, Gene Expression, Sample': 'frac_reads_mapped_gex',\n",
    "        'Confidently mapped to genome, Gene Expression, Library': 'frac_reads_mapped_gex',\n",
    "        'Mean reads per cell, Gene Expression, Sample': 'mean_reads_per_cell_gex'\n",
    "    }\n",
    "\n",
    "    df = df[[f for f in keep.keys() if f in df.columns]].rename(columns=keep)\n",
    "\n",
    "    return df\n",
    "\n",
    "\n",
    "def cxg_add_labels(adata):\n",
    "    adata.obs['cell_type_ontology_term_id'] = 'unknown'\n",
    "    labeler = AnnDataLabelAppender(adata)\n",
    "    labeler._add_labels()\n",
    "    adata.obs.drop(columns=['cell_type_ontology_term_id','cell_type'],inplace=True)\n",
    "\n",
    "    schema_v = labeler.schema_version\n",
    "    adata.uns['schema_version'] = schema_v\n",
    "    adata.uns['schema_reference'] = labeler._build_schema_reference_url(schema_v)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "dcca356f-0b0c-48cb-9ced-bc50f06f0678",
   "metadata": {},
   "outputs": [],
   "source": [
    "#read in the sample metadata\n",
    "sheet = ''\n",
    "gid = ''\n",
    "url = f'https://docs.google.com/spreadsheets/d/{sheet}/export?format=csv&gid={gid}'\n",
    "response = requests.get(url)\n",
    "sample_df = pd.read_csv(BytesIO(response.content), comment=\"#\", dtype=str).dropna(axis=1,how='all')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "cfef2216-6069-401e-b506-a2263164bdc5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Matching 'EFO' term id not found for label '10x flex' in column 'assay'\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>sample_probe_barcode</th>\n",
       "      <th>donor_id</th>\n",
       "      <th>tissue_type</th>\n",
       "      <th>suspension_type</th>\n",
       "      <th>perturbation_strategy</th>\n",
       "      <th>experimental_perturbation_type</th>\n",
       "      <th>experimental_perturbation</th>\n",
       "      <th>experimental_perturbation_time_point</th>\n",
       "      <th>organism_ontology_term_id</th>\n",
       "      <th>sex_ontology_term_id</th>\n",
       "      <th>self_reported_ethnicity_ontology_term_id</th>\n",
       "      <th>disease_ontology_term_id</th>\n",
       "      <th>assay_ontology_term_id</th>\n",
       "      <th>development_stage_ontology_term_id</th>\n",
       "      <th>tissue_ontology_term_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CD4i_R1L01_D1_Rest</td>\n",
       "      <td>BC001+CR001|BC002+CR002|BC003+CR003|BC004+CR004</td>\n",
       "      <td>CE0008162</td>\n",
       "      <td>primary cell culture</td>\n",
       "      <td>cell</td>\n",
       "      <td>CRISPRi</td>\n",
       "      <td>stimulation</td>\n",
       "      <td>IL2</td>\n",
       "      <td>Rest</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>PATO:0000383</td>\n",
       "      <td>HANCESTRO:0014</td>\n",
       "      <td>HANCESTRO:0568</td>\n",
       "      <td>n/a</td>\n",
       "      <td>HsapDv:0000123</td>\n",
       "      <td>CD4-positive, alpha-beta T cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CD4i_R1L01_D1_Stim8hr</td>\n",
       "      <td>BC009+CR009|BC010+CR010|BC011+CR011|BC012+CR012</td>\n",
       "      <td>CE0008162</td>\n",
       "      <td>primary cell culture</td>\n",
       "      <td>cell</td>\n",
       "      <td>CRISPRi</td>\n",
       "      <td>stimulation</td>\n",
       "      <td>IL2</td>\n",
       "      <td>Stim8hr</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>PATO:0000383</td>\n",
       "      <td>HANCESTRO:0014</td>\n",
       "      <td>HANCESTRO:0568</td>\n",
       "      <td>n/a</td>\n",
       "      <td>HsapDv:0000123</td>\n",
       "      <td>CD4-positive, alpha-beta T cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CD4i_R1L01_D2_Rest</td>\n",
       "      <td>BC005+CR005|BC006+CR006|BC007+CR007|BC008+CR008</td>\n",
       "      <td>CE0010866</td>\n",
       "      <td>primary cell culture</td>\n",
       "      <td>cell</td>\n",
       "      <td>CRISPRi</td>\n",
       "      <td>stimulation</td>\n",
       "      <td>IL2</td>\n",
       "      <td>Rest</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>PATO:0000383</td>\n",
       "      <td>HANCESTRO:0568</td>\n",
       "      <td>HANCESTRO:0568</td>\n",
       "      <td>n/a</td>\n",
       "      <td>HsapDv:0000116</td>\n",
       "      <td>CD4-positive, alpha-beta T cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CD4i_R1L01_D2_Stim8hr</td>\n",
       "      <td>BC013+CR013|BC014+CR014|BC015+CR015|BC016+CR016</td>\n",
       "      <td>CE0010866</td>\n",
       "      <td>primary cell culture</td>\n",
       "      <td>cell</td>\n",
       "      <td>CRISPRi</td>\n",
       "      <td>stimulation</td>\n",
       "      <td>IL2</td>\n",
       "      <td>Stim8hr</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>PATO:0000383</td>\n",
       "      <td>HANCESTRO:0568</td>\n",
       "      <td>HANCESTRO:0568</td>\n",
       "      <td>n/a</td>\n",
       "      <td>HsapDv:0000116</td>\n",
       "      <td>CD4-positive, alpha-beta T cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CD4i_R1L02_D1_Rest</td>\n",
       "      <td>BC001+CR001|BC002+CR002|BC003+CR003|BC004+CR004</td>\n",
       "      <td>CE0008162</td>\n",
       "      <td>primary cell culture</td>\n",
       "      <td>cell</td>\n",
       "      <td>CRISPRi</td>\n",
       "      <td>stimulation</td>\n",
       "      <td>IL2</td>\n",
       "      <td>Rest</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>PATO:0000383</td>\n",
       "      <td>HANCESTRO:0014</td>\n",
       "      <td>HANCESTRO:0568</td>\n",
       "      <td>n/a</td>\n",
       "      <td>HsapDv:0000123</td>\n",
       "      <td>CD4-positive, alpha-beta T cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>CD4i_R1L22_D2_Stim8hr</td>\n",
       "      <td>BC013+CR013|BC014+CR014|BC015+CR015|BC016+CR016</td>\n",
       "      <td>CE0010866</td>\n",
       "      <td>primary cell culture</td>\n",
       "      <td>cell</td>\n",
       "      <td>CRISPRi</td>\n",
       "      <td>stimulation</td>\n",
       "      <td>IL2</td>\n",
       "      <td>Stim8hr</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>PATO:0000383</td>\n",
       "      <td>HANCESTRO:0568</td>\n",
       "      <td>HANCESTRO:0568</td>\n",
       "      <td>n/a</td>\n",
       "      <td>HsapDv:0000116</td>\n",
       "      <td>CD4-positive, alpha-beta T cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>CD4i_R1L23_D1_Rest</td>\n",
       "      <td>BC001+CR001|BC002+CR002|BC003+CR003|BC004+CR004</td>\n",
       "      <td>CE0008162</td>\n",
       "      <td>primary cell culture</td>\n",
       "      <td>cell</td>\n",
       "      <td>CRISPRi</td>\n",
       "      <td>stimulation</td>\n",
       "      <td>IL2</td>\n",
       "      <td>Rest</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>PATO:0000383</td>\n",
       "      <td>HANCESTRO:0014</td>\n",
       "      <td>HANCESTRO:0568</td>\n",
       "      <td>n/a</td>\n",
       "      <td>HsapDv:0000123</td>\n",
       "      <td>CD4-positive, alpha-beta T cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>CD4i_R1L23_D1_Stim8hr</td>\n",
       "      <td>BC009+CR009|BC010+CR010|BC011+CR011|BC012+CR012</td>\n",
       "      <td>CE0008162</td>\n",
       "      <td>primary cell culture</td>\n",
       "      <td>cell</td>\n",
       "      <td>CRISPRi</td>\n",
       "      <td>stimulation</td>\n",
       "      <td>IL2</td>\n",
       "      <td>Stim8hr</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>PATO:0000383</td>\n",
       "      <td>HANCESTRO:0014</td>\n",
       "      <td>HANCESTRO:0568</td>\n",
       "      <td>n/a</td>\n",
       "      <td>HsapDv:0000123</td>\n",
       "      <td>CD4-positive, alpha-beta T cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>CD4i_R1L23_D2_Rest</td>\n",
       "      <td>BC005+CR005|BC006+CR006|BC007+CR007|BC008+CR008</td>\n",
       "      <td>CE0010866</td>\n",
       "      <td>primary cell culture</td>\n",
       "      <td>cell</td>\n",
       "      <td>CRISPRi</td>\n",
       "      <td>stimulation</td>\n",
       "      <td>IL2</td>\n",
       "      <td>Rest</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>PATO:0000383</td>\n",
       "      <td>HANCESTRO:0568</td>\n",
       "      <td>HANCESTRO:0568</td>\n",
       "      <td>n/a</td>\n",
       "      <td>HsapDv:0000116</td>\n",
       "      <td>CD4-positive, alpha-beta T cell</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>CD4i_R1L23_D2_Stim8hr</td>\n",
       "      <td>BC013+CR013|BC014+CR014|BC015+CR015|BC016+CR016</td>\n",
       "      <td>CE0010866</td>\n",
       "      <td>primary cell culture</td>\n",
       "      <td>cell</td>\n",
       "      <td>CRISPRi</td>\n",
       "      <td>stimulation</td>\n",
       "      <td>IL2</td>\n",
       "      <td>Stim8hr</td>\n",
       "      <td>NCBITaxon:9606</td>\n",
       "      <td>PATO:0000383</td>\n",
       "      <td>HANCESTRO:0568</td>\n",
       "      <td>HANCESTRO:0568</td>\n",
       "      <td>n/a</td>\n",
       "      <td>HsapDv:0000116</td>\n",
       "      <td>CD4-positive, alpha-beta T cell</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>92 rows Ã— 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              sample_name                             sample_probe_barcode  \\\n",
       "0      CD4i_R1L01_D1_Rest  BC001+CR001|BC002+CR002|BC003+CR003|BC004+CR004   \n",
       "1   CD4i_R1L01_D1_Stim8hr  BC009+CR009|BC010+CR010|BC011+CR011|BC012+CR012   \n",
       "2      CD4i_R1L01_D2_Rest  BC005+CR005|BC006+CR006|BC007+CR007|BC008+CR008   \n",
       "3   CD4i_R1L01_D2_Stim8hr  BC013+CR013|BC014+CR014|BC015+CR015|BC016+CR016   \n",
       "4      CD4i_R1L02_D1_Rest  BC001+CR001|BC002+CR002|BC003+CR003|BC004+CR004   \n",
       "..                    ...                                              ...   \n",
       "87  CD4i_R1L22_D2_Stim8hr  BC013+CR013|BC014+CR014|BC015+CR015|BC016+CR016   \n",
       "88     CD4i_R1L23_D1_Rest  BC001+CR001|BC002+CR002|BC003+CR003|BC004+CR004   \n",
       "89  CD4i_R1L23_D1_Stim8hr  BC009+CR009|BC010+CR010|BC011+CR011|BC012+CR012   \n",
       "90     CD4i_R1L23_D2_Rest  BC005+CR005|BC006+CR006|BC007+CR007|BC008+CR008   \n",
       "91  CD4i_R1L23_D2_Stim8hr  BC013+CR013|BC014+CR014|BC015+CR015|BC016+CR016   \n",
       "\n",
       "     donor_id           tissue_type suspension_type perturbation_strategy  \\\n",
       "0   CE0008162  primary cell culture            cell               CRISPRi   \n",
       "1   CE0008162  primary cell culture            cell               CRISPRi   \n",
       "2   CE0010866  primary cell culture            cell               CRISPRi   \n",
       "3   CE0010866  primary cell culture            cell               CRISPRi   \n",
       "4   CE0008162  primary cell culture            cell               CRISPRi   \n",
       "..        ...                   ...             ...                   ...   \n",
       "87  CE0010866  primary cell culture            cell               CRISPRi   \n",
       "88  CE0008162  primary cell culture            cell               CRISPRi   \n",
       "89  CE0008162  primary cell culture            cell               CRISPRi   \n",
       "90  CE0010866  primary cell culture            cell               CRISPRi   \n",
       "91  CE0010866  primary cell culture            cell               CRISPRi   \n",
       "\n",
       "   experimental_perturbation_type experimental_perturbation  \\\n",
       "0                     stimulation                       IL2   \n",
       "1                     stimulation                       IL2   \n",
       "2                     stimulation                       IL2   \n",
       "3                     stimulation                       IL2   \n",
       "4                     stimulation                       IL2   \n",
       "..                            ...                       ...   \n",
       "87                    stimulation                       IL2   \n",
       "88                    stimulation                       IL2   \n",
       "89                    stimulation                       IL2   \n",
       "90                    stimulation                       IL2   \n",
       "91                    stimulation                       IL2   \n",
       "\n",
       "   experimental_perturbation_time_point organism_ontology_term_id  \\\n",
       "0                                  Rest            NCBITaxon:9606   \n",
       "1                               Stim8hr            NCBITaxon:9606   \n",
       "2                                  Rest            NCBITaxon:9606   \n",
       "3                               Stim8hr            NCBITaxon:9606   \n",
       "4                                  Rest            NCBITaxon:9606   \n",
       "..                                  ...                       ...   \n",
       "87                              Stim8hr            NCBITaxon:9606   \n",
       "88                                 Rest            NCBITaxon:9606   \n",
       "89                              Stim8hr            NCBITaxon:9606   \n",
       "90                                 Rest            NCBITaxon:9606   \n",
       "91                              Stim8hr            NCBITaxon:9606   \n",
       "\n",
       "   sex_ontology_term_id self_reported_ethnicity_ontology_term_id  \\\n",
       "0          PATO:0000383                           HANCESTRO:0014   \n",
       "1          PATO:0000383                           HANCESTRO:0014   \n",
       "2          PATO:0000383                           HANCESTRO:0568   \n",
       "3          PATO:0000383                           HANCESTRO:0568   \n",
       "4          PATO:0000383                           HANCESTRO:0014   \n",
       "..                  ...                                      ...   \n",
       "87         PATO:0000383                           HANCESTRO:0568   \n",
       "88         PATO:0000383                           HANCESTRO:0014   \n",
       "89         PATO:0000383                           HANCESTRO:0014   \n",
       "90         PATO:0000383                           HANCESTRO:0568   \n",
       "91         PATO:0000383                           HANCESTRO:0568   \n",
       "\n",
       "   disease_ontology_term_id assay_ontology_term_id  \\\n",
       "0            HANCESTRO:0568                    n/a   \n",
       "1            HANCESTRO:0568                    n/a   \n",
       "2            HANCESTRO:0568                    n/a   \n",
       "3            HANCESTRO:0568                    n/a   \n",
       "4            HANCESTRO:0568                    n/a   \n",
       "..                      ...                    ...   \n",
       "87           HANCESTRO:0568                    n/a   \n",
       "88           HANCESTRO:0568                    n/a   \n",
       "89           HANCESTRO:0568                    n/a   \n",
       "90           HANCESTRO:0568                    n/a   \n",
       "91           HANCESTRO:0568                    n/a   \n",
       "\n",
       "   development_stage_ontology_term_id          tissue_ontology_term_id  \n",
       "0                      HsapDv:0000123  CD4-positive, alpha-beta T cell  \n",
       "1                      HsapDv:0000123  CD4-positive, alpha-beta T cell  \n",
       "2                      HsapDv:0000116  CD4-positive, alpha-beta T cell  \n",
       "3                      HsapDv:0000116  CD4-positive, alpha-beta T cell  \n",
       "4                      HsapDv:0000123  CD4-positive, alpha-beta T cell  \n",
       "..                                ...                              ...  \n",
       "87                     HsapDv:0000116  CD4-positive, alpha-beta T cell  \n",
       "88                     HsapDv:0000123  CD4-positive, alpha-beta T cell  \n",
       "89                     HsapDv:0000123  CD4-positive, alpha-beta T cell  \n",
       "90                     HsapDv:0000116  CD4-positive, alpha-beta T cell  \n",
       "91                     HsapDv:0000116  CD4-positive, alpha-beta T cell  \n",
       "\n",
       "[92 rows x 16 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Mapping ontologies\n",
    "\n",
    "col_ont_map = {\n",
    "    'organism':'NCBITaxon',\n",
    "    'sex':'PATO',\n",
    "    'self_reported_ethnicity':'HANCESTRO',\n",
    "    'disease':'MONDO',\n",
    "    'assay':'EFO',\n",
    "    'development_stage':{'NCBITaxon:6239':'WBls', # C. Elegans\n",
    "                         'NCBITaxon:7227':'FBdv', # Drosophila\n",
    "                         'NCBITaxon:10090':'MmusDv', # Mouse\n",
    "                         'NCBITaxon:7955':'ZFS', # Zebrafish\n",
    "                         'other':'HsapDv' # For all other organisms, use HsapDv\n",
    "                        },\n",
    "    'tissue':{'NCBITaxon:6239':'WBbt', # C. Elegans\n",
    "              'NCBITaxon:7227':'FBbt', # Drosophila\n",
    "              'NCBITaxon:7955':'ZFA', # Zebrafish\n",
    "              'other':'UBERON' # For all other organisms, use UBERON\n",
    "             }\n",
    "}\n",
    "\n",
    "\n",
    "for col in col_ont_map:\n",
    "    map_dict = {}\n",
    "    for label in sample_df[col].unique():\n",
    "        if col == 'disease' and label == 'normal': # normal is not in MONDO ontology\n",
    "            map_dict[label] = 'PATO:0000461'\n",
    "        elif label in ['unknown','n/a']: # unknown and n/a won't be in ontologies, pass along\n",
    "            map_dict[label] = label\n",
    "        elif col in ['tissue','development_stage']:\n",
    "            if col == 'tissue':\n",
    "                # Find what tissue type is at label row\n",
    "                if sample_df.loc[sample_df[col] == label, 'tissue_type'].tolist()[0] != 'tissue':\n",
    "                    map_dict[label] = label # Don't map cell type in tissue\n",
    "                    continue\n",
    "            # Find what organism term id is at label row\n",
    "            org_term_id = sample_df.loc[sample_df[col] == label, 'organism_ontology_term_id'].tolist()[0]\n",
    "            if org_term_id in col_ont_map[col]:\n",
    "                # Get ontology of specific organism and map label\n",
    "                species_ont = col_ont_map[col][sample_df.loc[org_term_id]]\n",
    "                term_id = ontology_parser.get_term_id_by_label(label, species_ont)\n",
    "            else:\n",
    "                term_id = ontology_parser.get_term_id_by_label(label, col_ont_map[col]['other'])\n",
    "        else:\n",
    "            term_id = ontology_parser.get_term_id_by_label(label, col_ont_map[col])\n",
    "        if term_id == None:\n",
    "            print(f\"Matching '{col_ont_map[col]}' term id not found for label '{label}' in column '{col}'\")\n",
    "        map_dict[label] = term_id\n",
    "    sample_df[col + '_ontology_term_id'] = sample_df[col].map(map_dict)\n",
    "    del sample_df[col]\n",
    "    \n",
    "# Blank fields in worksheet result in NaN values in dataframe, replacing these with n/a ? \n",
    "# Could also replace with unknown for certain columns using fillna options?\n",
    "sample_df.fillna('n/a', inplace=True)\n",
    "sample_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f6914a64-c8dd-4866-8a57-c7ca0f8746e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "bucket_name = 'czi-psomagen'\n",
    "lab = 'weissman'\n",
    "proj = 'scaling-in-vivo-perturb-seq-in-the-liver-and-beyond'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d78fdbfc-45b7-4069-a300-09aeb813f756",
   "metadata": {},
   "outputs": [
    {
     "ename": "ClientError",
     "evalue": "An error occurred (AccessDenied) when calling the ListObjects operation: Access Denied",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mClientError\u001b[39m                               Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[6]\u001b[39m\u001b[32m, line 2\u001b[39m\n\u001b[32m      1\u001b[39m my_dir = \u001b[33mf\u001b[39m\u001b[33m'\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mlab\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m-\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mproj\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m/\u001b[39m\u001b[33m'\u001b[39m\n\u001b[32m----> \u001b[39m\u001b[32m2\u001b[39m r = \u001b[43ms3client\u001b[49m\u001b[43m.\u001b[49m\u001b[43mlist_objects\u001b[49m\u001b[43m(\u001b[49m\u001b[43mBucket\u001b[49m\u001b[43m=\u001b[49m\u001b[43mbucket_name\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mPrefix\u001b[49m\u001b[43m=\u001b[49m\u001b[43mmy_dir\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mDelimiter\u001b[49m\u001b[43m=\u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43m/\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[32m      3\u001b[39m orders = [o[\u001b[33m'\u001b[39m\u001b[33mPrefix\u001b[39m\u001b[33m'\u001b[39m].replace(my_dir,\u001b[33m'\u001b[39m\u001b[33m'\u001b[39m) \u001b[38;5;28;01mfor\u001b[39;00m o \u001b[38;5;129;01min\u001b[39;00m r[\u001b[33m'\u001b[39m\u001b[33mCommonPrefixes\u001b[39m\u001b[33m'\u001b[39m]]\n\u001b[32m      5\u001b[39m samples = []\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/opt/anaconda3/envs/lattice/lib/python3.11/site-packages/botocore/client.py:570\u001b[39m, in \u001b[36mClientCreator._create_api_method.<locals>._api_call\u001b[39m\u001b[34m(self, *args, **kwargs)\u001b[39m\n\u001b[32m    566\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m(\n\u001b[32m    567\u001b[39m         \u001b[33mf\u001b[39m\u001b[33m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mpy_operation_name\u001b[38;5;132;01m}\u001b[39;00m\u001b[33m() only accepts keyword arguments.\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m    568\u001b[39m     )\n\u001b[32m    569\u001b[39m \u001b[38;5;66;03m# The \"self\" in this scope is referring to the BaseClient.\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m570\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[43m.\u001b[49m\u001b[43m_make_api_call\u001b[49m\u001b[43m(\u001b[49m\u001b[43moperation_name\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/opt/anaconda3/envs/lattice/lib/python3.11/site-packages/botocore/context.py:124\u001b[39m, in \u001b[36mwith_current_context.<locals>.decorator.<locals>.wrapper\u001b[39m\u001b[34m(*args, **kwargs)\u001b[39m\n\u001b[32m    122\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m hook:\n\u001b[32m    123\u001b[39m     hook()\n\u001b[32m--> \u001b[39m\u001b[32m124\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mfunc\u001b[49m\u001b[43m(\u001b[49m\u001b[43m*\u001b[49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43m*\u001b[49m\u001b[43m*\u001b[49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m/opt/anaconda3/envs/lattice/lib/python3.11/site-packages/botocore/client.py:1031\u001b[39m, in \u001b[36mBaseClient._make_api_call\u001b[39m\u001b[34m(self, operation_name, api_params)\u001b[39m\n\u001b[32m   1027\u001b[39m     error_code = error_info.get(\u001b[33m\"\u001b[39m\u001b[33mQueryErrorCode\u001b[39m\u001b[33m\"\u001b[39m) \u001b[38;5;129;01mor\u001b[39;00m error_info.get(\n\u001b[32m   1028\u001b[39m         \u001b[33m\"\u001b[39m\u001b[33mCode\u001b[39m\u001b[33m\"\u001b[39m\n\u001b[32m   1029\u001b[39m     )\n\u001b[32m   1030\u001b[39m     error_class = \u001b[38;5;28mself\u001b[39m.exceptions.from_code(error_code)\n\u001b[32m-> \u001b[39m\u001b[32m1031\u001b[39m     \u001b[38;5;28;01mraise\u001b[39;00m error_class(parsed_response, operation_name)\n\u001b[32m   1032\u001b[39m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m   1033\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m parsed_response\n",
      "\u001b[31mClientError\u001b[39m: An error occurred (AccessDenied) when calling the ListObjects operation: Access Denied"
     ]
    }
   ],
   "source": [
    "my_dir = f'{lab}-{proj}/'\n",
    "r = s3client.list_objects(Bucket=bucket_name, Prefix=my_dir, Delimiter='/')\n",
    "orders = [o['Prefix'].replace(my_dir,'') for o in r['CommonPrefixes']]\n",
    "\n",
    "samples = []\n",
    "for o in orders:\n",
    "    r = s3client.list_objects(Bucket=bucket_name, Prefix=f'{my_dir}{o}', Delimiter='/')\n",
    "    libs = [l['Prefix'].replace(f'{my_dir}{o}','') for l in r['CommonPrefixes']]\n",
    "    for l in libs:\n",
    "        r = s3client.list_objects(Bucket=bucket_name, Prefix=f'{my_dir}{o}{l}processed/cellranger/', Delimiter='/')\n",
    "        dates = [d['Prefix'].replace(f'{my_dir}{o}{l}processed/cellranger/','') for d in r['CommonPrefixes']]\n",
    "        for d in dates:\n",
    "            r = s3client.list_objects(Bucket=bucket_name, Prefix=f'{my_dir}{o}{l}processed/cellranger/{d}outs/per_sample_outs/', Delimiter='/')\n",
    "            subs = [s['Prefix'].replace(f'{my_dir}{o}{l}processed/cellranger/{d}outs/per_sample_outs/','') for s in r['CommonPrefixes']]\n",
    "            for s in subs:\n",
    "                samples.append({\n",
    "                    'order': o,\n",
    "                    'library': l.rstrip('/'),\n",
    "                    'date': d,\n",
    "                    'sample': s.rstrip('/')\n",
    "                })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff24c2b5-49db-4f56-b6d1-afe038409ae3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#identify which sample to curate\n",
    "n = 0\n",
    "s = samples[n]\n",
    "s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35d329fa-62b9-49da-aedb-c844aec6407b",
   "metadata": {},
   "outputs": [],
   "source": [
    "order = s['order']\n",
    "lib = s['library']\n",
    "run_date = s['date']\n",
    "samp = s['sample']\n",
    "\n",
    "s_dir = f'{lab}-{proj}/{order}{lib}/processed/cellranger/{run_date}outs/per_sample_outs/{samp}'\n",
    "download_files(s_dir)\n",
    "\n",
    "adata = sc.read_10x_h5('sample_filtered_feature_bc_matrix.h5', gex_only=True)\n",
    "\n",
    "adata = custom_var_to_obs(adata)\n",
    "\n",
    "adata.obs['lab'] = lab\n",
    "adata.obs['project'] = proj\n",
    "adata.obs['sample_name'] = samp\n",
    "\n",
    "adata.obs = adata.obs.merge(sample_df, on='sample_name', how='left').set_index(adata.obs.index)\n",
    "adata.uns['organism_ontology_term_id'] = adata.obs['organism_ontology_term_id'].unique()[0]\n",
    "adata.obs.drop(columns=['organism_ontology_term_id'],inplace=True)\n",
    "\n",
    "crispr_df = gather_crispr()\n",
    "adata.obs = adata.obs.merge(\n",
    "    crispr_df, left_index=True, right_index=True, how='left'\n",
    ").set_index(adata.obs.index)\n",
    "\n",
    "metrics_df = gather_metrics(samp)\n",
    "for c in metrics_df.columns:\n",
    "    adata.obs[c] = metrics_df[c].values[0]\n",
    "\n",
    "cxg_add_labels(adata)\n",
    "\n",
    "#startswith('mt-') works for human & mouse, more attn needed for other organisms\n",
    "adata.var['mt'] = adata.var['feature_name'].str.lower().str.startswith('mt-')\n",
    "sc.pp.calculate_qc_metrics(adata, qc_vars=['mt'], inplace=True)\n",
    "\n",
    "adata.write(filename=f'{samp}_curated.h5ad', compression='gzip')\n",
    "\n",
    "for f in ['sample_filtered_feature_bc_matrix.h5','crispr_analysis.tar.gz','metrics_summary.csv']:\n",
    "    os.remove(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97f384b9-e302-4acc-82d3-aac3904176d9",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
