{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "bdd18dbd-2754-4d6e-bb48-188296a8b201",
   "metadata": {},
   "outputs": [],
   "source": [
    "import boto3\n",
    "import numpy as np\n",
    "import os\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import tarfile\n",
    "import requests\n",
    "import re\n",
    "import json\n",
    "import argparse\n",
    "import sys\n",
    "import fsspec\n",
    "import anndata as ad\n",
    "\n",
    "from io import BytesIO\n",
    "from dataclasses import dataclass, field\n",
    "from urllib.request import Request,urlopen\n",
    "from bs4 import BeautifulSoup\n",
    "from cellxgene_schema.write_labels import AnnDataLabelAppender\n",
    "from urllib.parse import quote\n",
    "from cellxgene_ontology_guide.supported_versions import CXGSchema, load_supported_versions\n",
    "from cellxgene_ontology_guide.ontology_parser import OntologyParser\n",
    "ontology_parser = OntologyParser()\n",
    "fs = fsspec.filesystem('s3')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "98e1aaef-010b-464c-b822-3e0456e5510f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading R096-25-0255\n",
      "Loading R096-25-0256\n",
      "Loading R096-25-0257\n",
      "Loading R096-25-0258\n"
     ]
    }
   ],
   "source": [
    "uri = 's3://czi-novogene/trapnell-seahub-bcp/NVUS2024101701-04/RNA3_096/processed/Run_2025-12-15_biohub/'\n",
    "uri_obj = URIPath(full_uri=uri)\n",
    "\n",
    "sample_metadata = LatticeMetadata(sheet_id='', tab_name='sample template', uri=uri_obj)\n",
    "sample_metadata.metadata_df\n",
    "\n",
    "if is_single(uri_obj):\n",
    "    experiment_04 = SingleSublibrary(metadata=sample_metadata, uri=uri_obj)\n",
    "else:\n",
    "    experiment_04 = MultiSublibrary(metadata=sample_metadata, uri=uri_obj)\n",
    "\n",
    "experiment_04.download_files()\n",
    "experiment_04.get_sample_ids_map()\n",
    "\n",
    "experiment_04.add_pseudosamples()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0760cdb6-0b73-4953-95eb-d9d7284ce264",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'bl1': 'R096-25-0255',\n",
       " 'bl2': 'R096-25-0256',\n",
       " 'bl3': 'R096-25-0257',\n",
       " 'bl4': 'R096-25-0258'}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "b7eb32c5-a465-472d-a86d-1ff774ec9456",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SingleSublibrary(metadata=LatticeMetadata(sheet_id='1iSivqO9pWOKt6LxzNYvCTpCDeWDYnTrkk3W1Hz-fJIQ', tab_name='sample template', uri=URIPath(full_uri='s3://czi-novogene/trapnell-seahub-bcp/NVUS2024101701-04/RNA3_096/processed/Run_2025-12-15_biohub/')), uri=URIPath(full_uri='s3://czi-novogene/trapnell-seahub-bcp/NVUS2024101701-04/RNA3_096/processed/Run_2025-12-15_biohub/'), all_pseudosamples=[Pseudosample(local_h5ad='temp_plate/R096-25-0255.QSR-8_anndata.h5ad', parent=..., pseudosample_id='bl1', sample_id='R096-25-0255', allcells='temp_plate/R096-25-0255.QSR-8.allCells.csv', metadata_sample=             sample_name !experiment_name experiment_name hash_index  \\\n",
       "0    CHEM13_SCALE_A1_Bl1  CHEM13_RNA3-096        RNA3_096   SCALE-A1   \n",
       "1    CHEM13_SCALE_B1_Bl1  CHEM13_RNA3-096        RNA3_096   SCALE-B1   \n",
       "2    CHEM13_SCALE_C1_Bl1  CHEM13_RNA3-096        RNA3_096   SCALE-C1   \n",
       "3    CHEM13_SCALE_D1_Bl1  CHEM13_RNA3-096        RNA3_096   SCALE-D1   \n",
       "4    CHEM13_SCALE_E1_Bl1  CHEM13_RNA3-096        RNA3_096   SCALE-E1   \n",
       "..                   ...              ...             ...        ...   \n",
       "75  CHEM13_SCALE_D10_Bl1  CHEM13_RNA3-096        RNA3_096  SCALE-D10   \n",
       "76  CHEM13_SCALE_E10_Bl1  CHEM13_RNA3-096        RNA3_096  SCALE-E10   \n",
       "77  CHEM13_SCALE_F10_Bl1  CHEM13_RNA3-096        RNA3_096  SCALE-F10   \n",
       "78  CHEM13_SCALE_G10_Bl1  CHEM13_RNA3-096        RNA3_096  SCALE-G10   \n",
       "79  CHEM13_SCALE_H10_Bl1  CHEM13_RNA3-096        RNA3_096  SCALE-H10   \n",
       "\n",
       "       nuclear_hash_oligo                                           RT_index  \\\n",
       "0   TGCGTTCCGG;GCTCATGCTT  SCALEQUANT-A1,SCALEQUANT-B1,SCALEQUANT-C1,SCAL...   \n",
       "1   TGCGTTCCGG;AGGATGGACG  SCALEQUANT-A1,SCALEQUANT-B1,SCALEQUANT-C1,SCAL...   \n",
       "2   TGCGTTCCGG;ACTGCTCCTA  SCALEQUANT-A1,SCALEQUANT-B1,SCALEQUANT-C1,SCAL...   \n",
       "3   TGCGTTCCGG;CGAACTAAGG  SCALEQUANT-A1,SCALEQUANT-B1,SCALEQUANT-C1,SCAL...   \n",
       "4   TGCGTTCCGG;GGAATCAGTA  SCALEQUANT-A1,SCALEQUANT-B1,SCALEQUANT-C1,SCAL...   \n",
       "..                    ...                                                ...   \n",
       "75  CCAGACCATG;CGAACTAAGG  SCALEQUANT-A1,SCALEQUANT-B1,SCALEQUANT-C1,SCAL...   \n",
       "76  CCAGACCATG;GGAATCAGTA  SCALEQUANT-A1,SCALEQUANT-B1,SCALEQUANT-C1,SCAL...   \n",
       "77  CCAGACCATG;GGCGGACGTT  SCALEQUANT-A1,SCALEQUANT-B1,SCALEQUANT-C1,SCAL...   \n",
       "78  CCAGACCATG;GGCTGAATTG  SCALEQUANT-A1,SCALEQUANT-B1,SCALEQUANT-C1,SCAL...   \n",
       "79  CCAGACCATG;GAGCGCGGCA  SCALEQUANT-A1,SCALEQUANT-B1,SCALEQUANT-C1,SCAL...   \n",
       "\n",
       "                                       RT_index_oligo  lab_name JUMBO-sci  \\\n",
       "0   CAAGAGTC,ATCTGAGA,CCACTTCT,GGATGTTA,ATTGAGCG,A...  Trapnell     FALSE   \n",
       "1   CAAGAGTC,ATCTGAGA,CCACTTCT,GGATGTTA,ATTGAGCG,A...  Trapnell     FALSE   \n",
       "2   CAAGAGTC,ATCTGAGA,CCACTTCT,GGATGTTA,ATTGAGCG,A...  Trapnell     FALSE   \n",
       "3   CAAGAGTC,ATCTGAGA,CCACTTCT,GGATGTTA,ATTGAGCG,A...  Trapnell     FALSE   \n",
       "4   CAAGAGTC,ATCTGAGA,CCACTTCT,GGATGTTA,ATTGAGCG,A...  Trapnell     FALSE   \n",
       "..                                                ...       ...       ...   \n",
       "75  CAAGAGTC,ATCTGAGA,CCACTTCT,GGATGTTA,ATTGAGCG,A...  Trapnell     FALSE   \n",
       "76  CAAGAGTC,ATCTGAGA,CCACTTCT,GGATGTTA,ATTGAGCG,A...  Trapnell     FALSE   \n",
       "77  CAAGAGTC,ATCTGAGA,CCACTTCT,GGATGTTA,ATTGAGCG,A...  Trapnell     FALSE   \n",
       "78  CAAGAGTC,ATCTGAGA,CCACTTCT,GGATGTTA,ATTGAGCG,A...  Trapnell     FALSE   \n",
       "79  CAAGAGTC,ATCTGAGA,CCACTTCT,GGATGTTA,ATTGAGCG,A...  Trapnell     FALSE   \n",
       "\n",
       "   pseudosample  ... development_stage (time)  \\\n",
       "0           bl1  ...                   72 hpf   \n",
       "1           bl1  ...                   72 hpf   \n",
       "2           bl1  ...                   72 hpf   \n",
       "3           bl1  ...                   72 hpf   \n",
       "4           bl1  ...                   72 hpf   \n",
       "..          ...  ...                      ...   \n",
       "75          bl1  ...                   72 hpf   \n",
       "76          bl1  ...                   72 hpf   \n",
       "77          bl1  ...                   72 hpf   \n",
       "78          bl1  ...                   72 hpf   \n",
       "79          bl1  ...                   72 hpf   \n",
       "\n",
       "                                               !assay  \\\n",
       "0   ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "1   ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "2   ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "3   ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "4   ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "..                                                ...   \n",
       "75  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "76  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "77  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "78  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "79  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "\n",
       "                                  assay tissue_type suspension_type  \\\n",
       "0   ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "1   ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "2   ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "3   ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "4   ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "..                                  ...         ...             ...   \n",
       "75  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "76  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "77  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "78  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "79  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "\n",
       "   !genetic_perturbation_strategy experimental_perturbation_type  \\\n",
       "0                no perturbations                       chemical   \n",
       "1                 no perturbation                       chemical   \n",
       "2                 no perturbation                       chemical   \n",
       "3                 no perturbation                       chemical   \n",
       "4                 no perturbation                       chemical   \n",
       "..                            ...                            ...   \n",
       "75                no perturbation                       chemical   \n",
       "76                no perturbation                       chemical   \n",
       "77                no perturbation                       chemical   \n",
       "78                no perturbation                       chemical   \n",
       "79                no perturbation                       chemical   \n",
       "\n",
       "   experimental_perturbation experimental_perturbation_concentration  \\\n",
       "0                   SB505124                           50 micromolar   \n",
       "1                   SB505124                           50 micromolar   \n",
       "2                   SB505124                           50 micromolar   \n",
       "3                   SB505124                           50 micromolar   \n",
       "4                   SB505124                           50 micromolar   \n",
       "..                       ...                                     ...   \n",
       "75                      DMSO                            0 micromolar   \n",
       "76                      DMSO                            0 micromolar   \n",
       "77                      DMSO                            0 micromolar   \n",
       "78                      DMSO                            0 micromolar   \n",
       "79                      DMSO                            0 micromolar   \n",
       "\n",
       "   experimental_perturbation_time_point  \n",
       "0                                48 hpf  \n",
       "1                                48 hpf  \n",
       "2                                48 hpf  \n",
       "3                                48 hpf  \n",
       "4                                48 hpf  \n",
       "..                                  ...  \n",
       "75                               60 hpf  \n",
       "76                               60 hpf  \n",
       "77                               60 hpf  \n",
       "78                               60 hpf  \n",
       "79                               60 hpf  \n",
       "\n",
       "[80 rows x 34 columns]), Pseudosample(local_h5ad='temp_plate/R096-25-0256.QSR-8_anndata.h5ad', parent=..., pseudosample_id='bl2', sample_id='R096-25-0256', allcells='temp_plate/R096-25-0256.QSR-8.allCells.csv', metadata_sample=              sample_name !experiment_name experiment_name hash_index  \\\n",
       "80   CHEM13_SCALE_A11_Bl2  CHEM13_RNA3-096        RNA3_096  SCALE-A11   \n",
       "81   CHEM13_SCALE_B11_Bl2  CHEM13_RNA3-096        RNA3_096  SCALE-B11   \n",
       "82   CHEM13_SCALE_C11_Bl2  CHEM13_RNA3-096        RNA3_096  SCALE-C11   \n",
       "83   CHEM13_SCALE_D11_Bl2  CHEM13_RNA3-096        RNA3_096  SCALE-D11   \n",
       "84   CHEM13_SCALE_E11_Bl2  CHEM13_RNA3-096        RNA3_096  SCALE-E11   \n",
       "85   CHEM13_SCALE_F11_Bl2  CHEM13_RNA3-096        RNA3_096  SCALE-F11   \n",
       "86   CHEM13_SCALE_G11_Bl2  CHEM13_RNA3-096        RNA3_096  SCALE-G11   \n",
       "87   CHEM13_SCALE_H11_Bl2  CHEM13_RNA3-096        RNA3_096  SCALE-H11   \n",
       "88   CHEM13_SCALE_A12_Bl2  CHEM13_RNA3-096        RNA3_096  SCALE-A12   \n",
       "89   CHEM13_SCALE_B12_Bl2  CHEM13_RNA3-096        RNA3_096  SCALE-B12   \n",
       "90   CHEM13_SCALE_C12_Bl2  CHEM13_RNA3-096        RNA3_096  SCALE-C12   \n",
       "91   CHEM13_SCALE_D12_Bl2  CHEM13_RNA3-096        RNA3_096  SCALE-D12   \n",
       "92   CHEM13_SCALE_E12_Bl2  CHEM13_RNA3-096        RNA3_096  SCALE-E12   \n",
       "93   CHEM13_SCALE_F12_Bl2  CHEM13_RNA3-096        RNA3_096  SCALE-F12   \n",
       "94   CHEM13_SCALE_G12_Bl2  CHEM13_RNA3-096        RNA3_096  SCALE-G12   \n",
       "95   CHEM13_SCALE_H12_Bl2  CHEM13_RNA3-096        RNA3_096  SCALE-H12   \n",
       "96    CHEM13_SCALE_A1_Bl2  CHEM13_RNA3-096        RNA3_096   SCALE-A1   \n",
       "97    CHEM13_SCALE_B1_Bl2  CHEM13_RNA3-096        RNA3_096   SCALE-B1   \n",
       "98    CHEM13_SCALE_C1_Bl2  CHEM13_RNA3-096        RNA3_096   SCALE-C1   \n",
       "99    CHEM13_SCALE_D1_Bl2  CHEM13_RNA3-096        RNA3_096   SCALE-D1   \n",
       "100   CHEM13_SCALE_E1_Bl2  CHEM13_RNA3-096        RNA3_096   SCALE-E1   \n",
       "101   CHEM13_SCALE_F1_Bl2  CHEM13_RNA3-096        RNA3_096   SCALE-F1   \n",
       "102   CHEM13_SCALE_G1_Bl2  CHEM13_RNA3-096        RNA3_096   SCALE-G1   \n",
       "103   CHEM13_SCALE_H1_Bl2  CHEM13_RNA3-096        RNA3_096   SCALE-H1   \n",
       "104   CHEM13_SCALE_A2_Bl2  CHEM13_RNA3-096        RNA3_096   SCALE-A2   \n",
       "105   CHEM13_SCALE_B2_Bl2  CHEM13_RNA3-096        RNA3_096   SCALE-B2   \n",
       "106   CHEM13_SCALE_C2_Bl2  CHEM13_RNA3-096        RNA3_096   SCALE-C2   \n",
       "107   CHEM13_SCALE_D2_Bl2  CHEM13_RNA3-096        RNA3_096   SCALE-D2   \n",
       "108   CHEM13_SCALE_E2_Bl2  CHEM13_RNA3-096        RNA3_096   SCALE-E2   \n",
       "109   CHEM13_SCALE_F2_Bl2  CHEM13_RNA3-096        RNA3_096   SCALE-F2   \n",
       "110   CHEM13_SCALE_G2_Bl2  CHEM13_RNA3-096        RNA3_096   SCALE-G2   \n",
       "111   CHEM13_SCALE_H2_Bl2  CHEM13_RNA3-096        RNA3_096   SCALE-H2   \n",
       "112   CHEM13_SCALE_A3_Bl2  CHEM13_RNA3-096        RNA3_096   SCALE-A3   \n",
       "113   CHEM13_SCALE_B3_Bl2  CHEM13_RNA3-096        RNA3_096   SCALE-B3   \n",
       "114   CHEM13_SCALE_C3_Bl2  CHEM13_RNA3-096        RNA3_096   SCALE-C3   \n",
       "115   CHEM13_SCALE_D3_Bl2  CHEM13_RNA3-096        RNA3_096   SCALE-D3   \n",
       "116   CHEM13_SCALE_E3_Bl2  CHEM13_RNA3-096        RNA3_096   SCALE-E3   \n",
       "117   CHEM13_SCALE_F3_Bl2  CHEM13_RNA3-096        RNA3_096   SCALE-F3   \n",
       "118   CHEM13_SCALE_G3_Bl2  CHEM13_RNA3-096        RNA3_096   SCALE-G3   \n",
       "119   CHEM13_SCALE_H3_Bl2  CHEM13_RNA3-096        RNA3_096   SCALE-H3   \n",
       "\n",
       "        nuclear_hash_oligo                                           RT_index  \\\n",
       "80   CTAGCGAGCC;GCTCATGCTT  SCALEQUANT-D1,SCALEQUANT-G1,SCALEQUANT-H1,SCAL...   \n",
       "81   CTAGCGAGCC;AGGATGGACG  SCALEQUANT-D1,SCALEQUANT-G1,SCALEQUANT-H1,SCAL...   \n",
       "82   CTAGCGAGCC;ACTGCTCCTA  SCALEQUANT-D1,SCALEQUANT-G1,SCALEQUANT-H1,SCAL...   \n",
       "83   CTAGCGAGCC;CGAACTAAGG  SCALEQUANT-D1,SCALEQUANT-G1,SCALEQUANT-H1,SCAL...   \n",
       "84   CTAGCGAGCC;GGAATCAGTA  SCALEQUANT-D1,SCALEQUANT-G1,SCALEQUANT-H1,SCAL...   \n",
       "85   CTAGCGAGCC;GGCGGACGTT  SCALEQUANT-D1,SCALEQUANT-G1,SCALEQUANT-H1,SCAL...   \n",
       "86   CTAGCGAGCC;GGCTGAATTG  SCALEQUANT-D1,SCALEQUANT-G1,SCALEQUANT-H1,SCAL...   \n",
       "87   CTAGCGAGCC;GAGCGCGGCA  SCALEQUANT-D1,SCALEQUANT-G1,SCALEQUANT-H1,SCAL...   \n",
       "88   GCTGCTGAAT;GCTCATGCTT  SCALEQUANT-D1,SCALEQUANT-G1,SCALEQUANT-H1,SCAL...   \n",
       "89   GCTGCTGAAT;AGGATGGACG  SCALEQUANT-D1,SCALEQUANT-G1,SCALEQUANT-H1,SCAL...   \n",
       "90   GCTGCTGAAT;ACTGCTCCTA  SCALEQUANT-D1,SCALEQUANT-G1,SCALEQUANT-H1,SCAL...   \n",
       "91   GCTGCTGAAT;CGAACTAAGG  SCALEQUANT-D1,SCALEQUANT-G1,SCALEQUANT-H1,SCAL...   \n",
       "92   GCTGCTGAAT;GGAATCAGTA  SCALEQUANT-D1,SCALEQUANT-G1,SCALEQUANT-H1,SCAL...   \n",
       "93   GCTGCTGAAT;GGCGGACGTT  SCALEQUANT-D1,SCALEQUANT-G1,SCALEQUANT-H1,SCAL...   \n",
       "94   GCTGCTGAAT;GGCTGAATTG  SCALEQUANT-D1,SCALEQUANT-G1,SCALEQUANT-H1,SCAL...   \n",
       "95   GCTGCTGAAT;GAGCGCGGCA  SCALEQUANT-D1,SCALEQUANT-G1,SCALEQUANT-H1,SCAL...   \n",
       "96   TGCGTTCCGG;GCTCATGCTT  SCALEQUANT-D1,SCALEQUANT-G1,SCALEQUANT-H1,SCAL...   \n",
       "97   TGCGTTCCGG;AGGATGGACG  SCALEQUANT-D1,SCALEQUANT-G1,SCALEQUANT-H1,SCAL...   \n",
       "98   TGCGTTCCGG;ACTGCTCCTA  SCALEQUANT-D1,SCALEQUANT-G1,SCALEQUANT-H1,SCAL...   \n",
       "99   TGCGTTCCGG;CGAACTAAGG  SCALEQUANT-D1,SCALEQUANT-G1,SCALEQUANT-H1,SCAL...   \n",
       "100  TGCGTTCCGG;GGAATCAGTA  SCALEQUANT-D1,SCALEQUANT-G1,SCALEQUANT-H1,SCAL...   \n",
       "101  TGCGTTCCGG;GGCGGACGTT  SCALEQUANT-D1,SCALEQUANT-G1,SCALEQUANT-H1,SCAL...   \n",
       "102  TGCGTTCCGG;GGCTGAATTG  SCALEQUANT-D1,SCALEQUANT-G1,SCALEQUANT-H1,SCAL...   \n",
       "103  TGCGTTCCGG;GAGCGCGGCA  SCALEQUANT-D1,SCALEQUANT-G1,SCALEQUANT-H1,SCAL...   \n",
       "104  CAATATGCGT;GCTCATGCTT  SCALEQUANT-D1,SCALEQUANT-G1,SCALEQUANT-H1,SCAL...   \n",
       "105  AGGATGGACG;CAATATGCGT  SCALEQUANT-D1,SCALEQUANT-G1,SCALEQUANT-H1,SCAL...   \n",
       "106  ACTGCTCCTA;CAATATGCGT  SCALEQUANT-D1,SCALEQUANT-G1,SCALEQUANT-H1,SCAL...   \n",
       "107  CGAACTAAGG;CAATATGCGT  SCALEQUANT-D1,SCALEQUANT-G1,SCALEQUANT-H1,SCAL...   \n",
       "108  GGAATCAGTA;CAATATGCGT  SCALEQUANT-D1,SCALEQUANT-G1,SCALEQUANT-H1,SCAL...   \n",
       "109  GGCGGACGTT;CAATATGCGT  SCALEQUANT-D1,SCALEQUANT-G1,SCALEQUANT-H1,SCAL...   \n",
       "110  GGCTGAATTG;CAATATGCGT  SCALEQUANT-D1,SCALEQUANT-G1,SCALEQUANT-H1,SCAL...   \n",
       "111  GAGCGCGGCA;CAATATGCGT  SCALEQUANT-D1,SCALEQUANT-G1,SCALEQUANT-H1,SCAL...   \n",
       "112  GCTCATGCTT;TGCTGGATAC  SCALEQUANT-D1,SCALEQUANT-G1,SCALEQUANT-H1,SCAL...   \n",
       "113  AGGATGGACG;TGCTGGATAC  SCALEQUANT-D1,SCALEQUANT-G1,SCALEQUANT-H1,SCAL...   \n",
       "114  ACTGCTCCTA;TGCTGGATAC  SCALEQUANT-D1,SCALEQUANT-G1,SCALEQUANT-H1,SCAL...   \n",
       "115  CGAACTAAGG;TGCTGGATAC  SCALEQUANT-D1,SCALEQUANT-G1,SCALEQUANT-H1,SCAL...   \n",
       "116  GGAATCAGTA;TGCTGGATAC  SCALEQUANT-D1,SCALEQUANT-G1,SCALEQUANT-H1,SCAL...   \n",
       "117  GGCGGACGTT;TGCTGGATAC  SCALEQUANT-D1,SCALEQUANT-G1,SCALEQUANT-H1,SCAL...   \n",
       "118  GGCTGAATTG;TGCTGGATAC  SCALEQUANT-D1,SCALEQUANT-G1,SCALEQUANT-H1,SCAL...   \n",
       "119  GAGCGCGGCA;TGCTGGATAC  SCALEQUANT-D1,SCALEQUANT-G1,SCALEQUANT-H1,SCAL...   \n",
       "\n",
       "                                        RT_index_oligo  lab_name JUMBO-sci  \\\n",
       "80   CTTAGCCA,TGCGCATA,CAGAGGTA,GTCTTGAC,ACTGGAAG,G...  Trapnell     FALSE   \n",
       "81   CTTAGCCA,TGCGCATA,CAGAGGTA,GTCTTGAC,ACTGGAAG,G...  Trapnell     FALSE   \n",
       "82   CTTAGCCA,TGCGCATA,CAGAGGTA,GTCTTGAC,ACTGGAAG,G...  Trapnell     FALSE   \n",
       "83   CTTAGCCA,TGCGCATA,CAGAGGTA,GTCTTGAC,ACTGGAAG,G...  Trapnell     FALSE   \n",
       "84   CTTAGCCA,TGCGCATA,CAGAGGTA,GTCTTGAC,ACTGGAAG,G...  Trapnell     FALSE   \n",
       "85   CTTAGCCA,TGCGCATA,CAGAGGTA,GTCTTGAC,ACTGGAAG,G...  Trapnell     FALSE   \n",
       "86   CTTAGCCA,TGCGCATA,CAGAGGTA,GTCTTGAC,ACTGGAAG,G...  Trapnell     FALSE   \n",
       "87   CTTAGCCA,TGCGCATA,CAGAGGTA,GTCTTGAC,ACTGGAAG,G...  Trapnell     FALSE   \n",
       "88   CTTAGCCA,TGCGCATA,CAGAGGTA,GTCTTGAC,ACTGGAAG,G...  Trapnell     FALSE   \n",
       "89   CTTAGCCA,TGCGCATA,CAGAGGTA,GTCTTGAC,ACTGGAAG,G...  Trapnell     FALSE   \n",
       "90   CTTAGCCA,TGCGCATA,CAGAGGTA,GTCTTGAC,ACTGGAAG,G...  Trapnell     FALSE   \n",
       "91   CTTAGCCA,TGCGCATA,CAGAGGTA,GTCTTGAC,ACTGGAAG,G...  Trapnell     FALSE   \n",
       "92   CTTAGCCA,TGCGCATA,CAGAGGTA,GTCTTGAC,ACTGGAAG,G...  Trapnell     FALSE   \n",
       "93   CTTAGCCA,TGCGCATA,CAGAGGTA,GTCTTGAC,ACTGGAAG,G...  Trapnell     FALSE   \n",
       "94   CTTAGCCA,TGCGCATA,CAGAGGTA,GTCTTGAC,ACTGGAAG,G...  Trapnell     FALSE   \n",
       "95   CTTAGCCA,TGCGCATA,CAGAGGTA,GTCTTGAC,ACTGGAAG,G...  Trapnell     FALSE   \n",
       "96   CTTAGCCA,TGCGCATA,CAGAGGTA,GTCTTGAC,ACTGGAAG,G...  Trapnell     FALSE   \n",
       "97   CTTAGCCA,TGCGCATA,CAGAGGTA,GTCTTGAC,ACTGGAAG,G...  Trapnell     FALSE   \n",
       "98   CTTAGCCA,TGCGCATA,CAGAGGTA,GTCTTGAC,ACTGGAAG,G...  Trapnell     FALSE   \n",
       "99   CTTAGCCA,TGCGCATA,CAGAGGTA,GTCTTGAC,ACTGGAAG,G...  Trapnell     FALSE   \n",
       "100  CTTAGCCA,TGCGCATA,CAGAGGTA,GTCTTGAC,ACTGGAAG,G...  Trapnell     FALSE   \n",
       "101  CTTAGCCA,TGCGCATA,CAGAGGTA,GTCTTGAC,ACTGGAAG,G...  Trapnell     FALSE   \n",
       "102  CTTAGCCA,TGCGCATA,CAGAGGTA,GTCTTGAC,ACTGGAAG,G...  Trapnell     FALSE   \n",
       "103  CTTAGCCA,TGCGCATA,CAGAGGTA,GTCTTGAC,ACTGGAAG,G...  Trapnell     FALSE   \n",
       "104  CTTAGCCA,TGCGCATA,CAGAGGTA,GTCTTGAC,ACTGGAAG,G...  Trapnell     FALSE   \n",
       "105  CTTAGCCA,TGCGCATA,CAGAGGTA,GTCTTGAC,ACTGGAAG,G...  Trapnell     FALSE   \n",
       "106  CTTAGCCA,TGCGCATA,CAGAGGTA,GTCTTGAC,ACTGGAAG,G...  Trapnell     FALSE   \n",
       "107  CTTAGCCA,TGCGCATA,CAGAGGTA,GTCTTGAC,ACTGGAAG,G...  Trapnell     FALSE   \n",
       "108  CTTAGCCA,TGCGCATA,CAGAGGTA,GTCTTGAC,ACTGGAAG,G...  Trapnell     FALSE   \n",
       "109  CTTAGCCA,TGCGCATA,CAGAGGTA,GTCTTGAC,ACTGGAAG,G...  Trapnell     FALSE   \n",
       "110  CTTAGCCA,TGCGCATA,CAGAGGTA,GTCTTGAC,ACTGGAAG,G...  Trapnell     FALSE   \n",
       "111  CTTAGCCA,TGCGCATA,CAGAGGTA,GTCTTGAC,ACTGGAAG,G...  Trapnell     FALSE   \n",
       "112  CTTAGCCA,TGCGCATA,CAGAGGTA,GTCTTGAC,ACTGGAAG,G...  Trapnell     FALSE   \n",
       "113  CTTAGCCA,TGCGCATA,CAGAGGTA,GTCTTGAC,ACTGGAAG,G...  Trapnell     FALSE   \n",
       "114  CTTAGCCA,TGCGCATA,CAGAGGTA,GTCTTGAC,ACTGGAAG,G...  Trapnell     FALSE   \n",
       "115  CTTAGCCA,TGCGCATA,CAGAGGTA,GTCTTGAC,ACTGGAAG,G...  Trapnell     FALSE   \n",
       "116  CTTAGCCA,TGCGCATA,CAGAGGTA,GTCTTGAC,ACTGGAAG,G...  Trapnell     FALSE   \n",
       "117  CTTAGCCA,TGCGCATA,CAGAGGTA,GTCTTGAC,ACTGGAAG,G...  Trapnell     FALSE   \n",
       "118  CTTAGCCA,TGCGCATA,CAGAGGTA,GTCTTGAC,ACTGGAAG,G...  Trapnell     FALSE   \n",
       "119  CTTAGCCA,TGCGCATA,CAGAGGTA,GTCTTGAC,ACTGGAAG,G...  Trapnell     FALSE   \n",
       "\n",
       "    pseudosample  ... development_stage (time)  \\\n",
       "80           bl2  ...                   96 hpf   \n",
       "81           bl2  ...                   96 hpf   \n",
       "82           bl2  ...                   96 hpf   \n",
       "83           bl2  ...                   96 hpf   \n",
       "84           bl2  ...                   96 hpf   \n",
       "85           bl2  ...                   96 hpf   \n",
       "86           bl2  ...                   96 hpf   \n",
       "87           bl2  ...                   96 hpf   \n",
       "88           bl2  ...                   96 hpf   \n",
       "89           bl2  ...                   96 hpf   \n",
       "90           bl2  ...                   96 hpf   \n",
       "91           bl2  ...                   96 hpf   \n",
       "92           bl2  ...                   96 hpf   \n",
       "93           bl2  ...                   96 hpf   \n",
       "94           bl2  ...                   96 hpf   \n",
       "95           bl2  ...                   96 hpf   \n",
       "96           bl2  ...                   96 hpf   \n",
       "97           bl2  ...                   96 hpf   \n",
       "98           bl2  ...                   96 hpf   \n",
       "99           bl2  ...                   96 hpf   \n",
       "100          bl2  ...                   96 hpf   \n",
       "101          bl2  ...                   96 hpf   \n",
       "102          bl2  ...                   96 hpf   \n",
       "103          bl2  ...                   96 hpf   \n",
       "104          bl2  ...                   96 hpf   \n",
       "105          bl2  ...                   96 hpf   \n",
       "106          bl2  ...                   96 hpf   \n",
       "107          bl2  ...                   96 hpf   \n",
       "108          bl2  ...                   96 hpf   \n",
       "109          bl2  ...                   96 hpf   \n",
       "110          bl2  ...                   96 hpf   \n",
       "111          bl2  ...                   96 hpf   \n",
       "112          bl2  ...                   96 hpf   \n",
       "113          bl2  ...                   96 hpf   \n",
       "114          bl2  ...                   96 hpf   \n",
       "115          bl2  ...                   96 hpf   \n",
       "116          bl2  ...                   96 hpf   \n",
       "117          bl2  ...                   96 hpf   \n",
       "118          bl2  ...                   96 hpf   \n",
       "119          bl2  ...                   96 hpf   \n",
       "\n",
       "                                                !assay  \\\n",
       "80   ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "81   ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "82   ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "83   ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "84   ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "85   ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "86   ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "87   ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "88   ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "89   ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "90   ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "91   ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "92   ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "93   ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "94   ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "95   ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "96   ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "97   ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "98   ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "99   ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "100  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "101  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "102  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "103  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "104  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "105  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "106  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "107  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "108  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "109  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "110  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "111  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "112  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "113  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "114  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "115  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "116  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "117  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "118  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "119  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "\n",
       "                                   assay tissue_type suspension_type  \\\n",
       "80   ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "81   ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "82   ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "83   ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "84   ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "85   ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "86   ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "87   ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "88   ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "89   ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "90   ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "91   ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "92   ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "93   ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "94   ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "95   ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "96   ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "97   ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "98   ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "99   ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "100  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "101  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "102  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "103  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "104  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "105  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "106  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "107  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "108  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "109  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "110  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "111  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "112  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "113  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "114  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "115  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "116  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "117  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "118  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "119  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "\n",
       "    !genetic_perturbation_strategy experimental_perturbation_type  \\\n",
       "80                 no perturbation                       chemical   \n",
       "81                 no perturbation                       chemical   \n",
       "82                 no perturbation                       chemical   \n",
       "83                 no perturbation                       chemical   \n",
       "84                 no perturbation                       chemical   \n",
       "85                 no perturbation                       chemical   \n",
       "86                 no perturbation                       chemical   \n",
       "87                 no perturbation                       chemical   \n",
       "88                 no perturbation                       chemical   \n",
       "89                 no perturbation                       chemical   \n",
       "90                 no perturbation                       chemical   \n",
       "91                 no perturbation                       chemical   \n",
       "92                 no perturbation                       chemical   \n",
       "93                 no perturbation                       chemical   \n",
       "94                 no perturbation                       chemical   \n",
       "95                 no perturbation                       chemical   \n",
       "96                 no perturbation                       chemical   \n",
       "97                 no perturbation                       chemical   \n",
       "98                 no perturbation                       chemical   \n",
       "99                 no perturbation                       chemical   \n",
       "100                no perturbation                       chemical   \n",
       "101                no perturbation                       chemical   \n",
       "102                no perturbation                       chemical   \n",
       "103                no perturbation                       chemical   \n",
       "104                no perturbation                       chemical   \n",
       "105                no perturbation                       chemical   \n",
       "106                no perturbation                       chemical   \n",
       "107                no perturbation                       chemical   \n",
       "108                no perturbation                       chemical   \n",
       "109                no perturbation                       chemical   \n",
       "110                no perturbation                       chemical   \n",
       "111                no perturbation                       chemical   \n",
       "112                no perturbation                       chemical   \n",
       "113                no perturbation                       chemical   \n",
       "114                no perturbation                       chemical   \n",
       "115                no perturbation                       chemical   \n",
       "116                no perturbation                       chemical   \n",
       "117                no perturbation                       chemical   \n",
       "118                no perturbation                       chemical   \n",
       "119                no perturbation                       chemical   \n",
       "\n",
       "    experimental_perturbation experimental_perturbation_concentration  \\\n",
       "80                   SB505124                           50 micromolar   \n",
       "81                   SB505124                           50 micromolar   \n",
       "82                   SB505124                           50 micromolar   \n",
       "83                   SB505124                           50 micromolar   \n",
       "84                   SB505124                           50 micromolar   \n",
       "85                   SB505124                           50 micromolar   \n",
       "86                   SB505124                           50 micromolar   \n",
       "87                   SB505124                           50 micromolar   \n",
       "88                     SU5402                           20 micromolar   \n",
       "89                     SU5402                           20 micromolar   \n",
       "90                     SU5402                           20 micromolar   \n",
       "91                     SU5402                           20 micromolar   \n",
       "92                     SU5402                           20 micromolar   \n",
       "93                     SU5402                           20 micromolar   \n",
       "94                     SU5402                           20 micromolar   \n",
       "95                     SU5402                           20 micromolar   \n",
       "96                       DMH1                           40 micromolar   \n",
       "97                       DMH1                           40 micromolar   \n",
       "98                       DMH1                           40 micromolar   \n",
       "99                       DMH1                           40 micromolar   \n",
       "100                      DMH1                           40 micromolar   \n",
       "101                      DMH1                           40 micromolar   \n",
       "102                      DMH1                           40 micromolar   \n",
       "103                      DMH1                           40 micromolar   \n",
       "104                  LY411575                           10 micromolar   \n",
       "105                  LY411575                           10 micromolar   \n",
       "106                  LY411575                           10 micromolar   \n",
       "107                  LY411575                           10 micromolar   \n",
       "108                  LY411575                           10 micromolar   \n",
       "109                  LY411575                           10 micromolar   \n",
       "110                  LY411575                           10 micromolar   \n",
       "111                  LY411575                           10 micromolar   \n",
       "112                      DMSO                            0 micromolar   \n",
       "113                      DMSO                            0 micromolar   \n",
       "114                      DMSO                            0 micromolar   \n",
       "115                      DMSO                            0 micromolar   \n",
       "116                      DMSO                            0 micromolar   \n",
       "117                      DMSO                            0 micromolar   \n",
       "118                      DMSO                            0 micromolar   \n",
       "119                      DMSO                            0 micromolar   \n",
       "\n",
       "    experimental_perturbation_time_point  \n",
       "80                                48 hpf  \n",
       "81                                48 hpf  \n",
       "82                                48 hpf  \n",
       "83                                48 hpf  \n",
       "84                                48 hpf  \n",
       "85                                48 hpf  \n",
       "86                                48 hpf  \n",
       "87                                48 hpf  \n",
       "88                                48 hpf  \n",
       "89                                48 hpf  \n",
       "90                                48 hpf  \n",
       "91                                48 hpf  \n",
       "92                                48 hpf  \n",
       "93                                48 hpf  \n",
       "94                                48 hpf  \n",
       "95                                48 hpf  \n",
       "96                                48 hpf  \n",
       "97                                48 hpf  \n",
       "98                                48 hpf  \n",
       "99                                48 hpf  \n",
       "100                               48 hpf  \n",
       "101                               48 hpf  \n",
       "102                               48 hpf  \n",
       "103                               48 hpf  \n",
       "104                               48 hpf  \n",
       "105                               48 hpf  \n",
       "106                               48 hpf  \n",
       "107                               48 hpf  \n",
       "108                               48 hpf  \n",
       "109                               48 hpf  \n",
       "110                               48 hpf  \n",
       "111                               48 hpf  \n",
       "112                               48 hpf  \n",
       "113                               48 hpf  \n",
       "114                               48 hpf  \n",
       "115                               48 hpf  \n",
       "116                               48 hpf  \n",
       "117                               48 hpf  \n",
       "118                               48 hpf  \n",
       "119                               48 hpf  \n",
       "\n",
       "[40 rows x 34 columns]), Pseudosample(local_h5ad='temp_plate/R096-25-0257.QSR-8_anndata.h5ad', parent=..., pseudosample_id='bl3', sample_id='R096-25-0257', allcells='temp_plate/R096-25-0257.QSR-8.allCells.csv', metadata_sample=             sample_name !experiment_name experiment_name hash_index  \\\n",
       "120  CHEM13_SCALE_A4_Bl3  CHEM13_RNA3-096        RNA3_096   SCALE-A4   \n",
       "121  CHEM13_SCALE_B4_Bl3  CHEM13_RNA3-096        RNA3_096   SCALE-B4   \n",
       "122  CHEM13_SCALE_C4_Bl3  CHEM13_RNA3-096        RNA3_096   SCALE-C4   \n",
       "123  CHEM13_SCALE_D4_Bl3  CHEM13_RNA3-096        RNA3_096   SCALE-D4   \n",
       "124  CHEM13_SCALE_E4_Bl3  CHEM13_RNA3-096        RNA3_096   SCALE-E4   \n",
       "125  CHEM13_SCALE_F4_Bl3  CHEM13_RNA3-096        RNA3_096   SCALE-F4   \n",
       "126  CHEM13_SCALE_G4_Bl3  CHEM13_RNA3-096        RNA3_096   SCALE-G4   \n",
       "127  CHEM13_SCALE_H4_Bl3  CHEM13_RNA3-096        RNA3_096   SCALE-H4   \n",
       "128  CHEM13_SCALE_A5_Bl3  CHEM13_RNA3-096        RNA3_096   SCALE-A5   \n",
       "129  CHEM13_SCALE_B5_Bl3  CHEM13_RNA3-096        RNA3_096   SCALE-B5   \n",
       "130  CHEM13_SCALE_C5_Bl3  CHEM13_RNA3-096        RNA3_096   SCALE-C5   \n",
       "131  CHEM13_SCALE_D5_Bl3  CHEM13_RNA3-096        RNA3_096   SCALE-D5   \n",
       "132  CHEM13_SCALE_E5_Bl3  CHEM13_RNA3-096        RNA3_096   SCALE-E5   \n",
       "133  CHEM13_SCALE_F5_Bl3  CHEM13_RNA3-096        RNA3_096   SCALE-F5   \n",
       "134  CHEM13_SCALE_G5_Bl3  CHEM13_RNA3-096        RNA3_096   SCALE-G5   \n",
       "135  CHEM13_SCALE_H5_Bl3  CHEM13_RNA3-096        RNA3_096   SCALE-H5   \n",
       "136  CHEM13_SCALE_A6_Bl3  CHEM13_RNA3-096        RNA3_096   SCALE-A6   \n",
       "137  CHEM13_SCALE_B6_Bl3  CHEM13_RNA3-096        RNA3_096   SCALE-B6   \n",
       "138  CHEM13_SCALE_C6_Bl3  CHEM13_RNA3-096        RNA3_096   SCALE-C6   \n",
       "139  CHEM13_SCALE_D6_Bl3  CHEM13_RNA3-096        RNA3_096   SCALE-D6   \n",
       "140  CHEM13_SCALE_E6_Bl3  CHEM13_RNA3-096        RNA3_096   SCALE-E6   \n",
       "141  CHEM13_SCALE_F6_Bl3  CHEM13_RNA3-096        RNA3_096   SCALE-F6   \n",
       "142  CHEM13_SCALE_G6_Bl3  CHEM13_RNA3-096        RNA3_096   SCALE-G6   \n",
       "143  CHEM13_SCALE_H6_Bl3  CHEM13_RNA3-096        RNA3_096   SCALE-H6   \n",
       "144  CHEM13_SCALE_A7_Bl3  CHEM13_RNA3-096        RNA3_096   SCALE-A7   \n",
       "145  CHEM13_SCALE_B7_Bl3  CHEM13_RNA3-096        RNA3_096   SCALE-B7   \n",
       "146  CHEM13_SCALE_C7_Bl3  CHEM13_RNA3-096        RNA3_096   SCALE-C7   \n",
       "147  CHEM13_SCALE_D7_Bl3  CHEM13_RNA3-096        RNA3_096   SCALE-D7   \n",
       "148  CHEM13_SCALE_E7_Bl3  CHEM13_RNA3-096        RNA3_096   SCALE-E7   \n",
       "149  CHEM13_SCALE_F7_Bl3  CHEM13_RNA3-096        RNA3_096   SCALE-F7   \n",
       "150  CHEM13_SCALE_G7_Bl3  CHEM13_RNA3-096        RNA3_096   SCALE-G7   \n",
       "151  CHEM13_SCALE_H7_Bl3  CHEM13_RNA3-096        RNA3_096   SCALE-H7   \n",
       "152  CHEM13_SCALE_A8_Bl3  CHEM13_RNA3-096        RNA3_096   SCALE-A8   \n",
       "153  CHEM13_SCALE_B8_Bl3  CHEM13_RNA3-096        RNA3_096   SCALE-B8   \n",
       "154  CHEM13_SCALE_C8_Bl3  CHEM13_RNA3-096        RNA3_096   SCALE-C8   \n",
       "155  CHEM13_SCALE_D8_Bl3  CHEM13_RNA3-096        RNA3_096   SCALE-D8   \n",
       "156  CHEM13_SCALE_E8_Bl3  CHEM13_RNA3-096        RNA3_096   SCALE-E8   \n",
       "157  CHEM13_SCALE_F8_Bl3  CHEM13_RNA3-096        RNA3_096   SCALE-F8   \n",
       "158  CHEM13_SCALE_G8_Bl3  CHEM13_RNA3-096        RNA3_096   SCALE-G8   \n",
       "159  CHEM13_SCALE_H8_Bl3  CHEM13_RNA3-096        RNA3_096   SCALE-H8   \n",
       "\n",
       "        nuclear_hash_oligo                                           RT_index  \\\n",
       "120  GCTCATGCTT;CATTCCTGAT  SCALEQUANT-E1,SCALEQUANT-E2,SCALEQUANT-G2,SCAL...   \n",
       "121  AGGATGGACG;CATTCCTGAT  SCALEQUANT-E1,SCALEQUANT-E2,SCALEQUANT-G2,SCAL...   \n",
       "122  ACTGCTCCTA;CATTCCTGAT  SCALEQUANT-E1,SCALEQUANT-E2,SCALEQUANT-G2,SCAL...   \n",
       "123  CGAACTAAGG;CATTCCTGAT  SCALEQUANT-E1,SCALEQUANT-E2,SCALEQUANT-G2,SCAL...   \n",
       "124  GGAATCAGTA;CATTCCTGAT  SCALEQUANT-E1,SCALEQUANT-E2,SCALEQUANT-G2,SCAL...   \n",
       "125  GGCGGACGTT;CATTCCTGAT  SCALEQUANT-E1,SCALEQUANT-E2,SCALEQUANT-G2,SCAL...   \n",
       "126  GGCTGAATTG;CATTCCTGAT  SCALEQUANT-E1,SCALEQUANT-E2,SCALEQUANT-G2,SCAL...   \n",
       "127  GAGCGCGGCA;CATTCCTGAT  SCALEQUANT-E1,SCALEQUANT-E2,SCALEQUANT-G2,SCAL...   \n",
       "128  GCTCATGCTT;GAATTCGAGC  SCALEQUANT-E1,SCALEQUANT-E2,SCALEQUANT-G2,SCAL...   \n",
       "129  AGGATGGACG;GAATTCGAGC  SCALEQUANT-E1,SCALEQUANT-E2,SCALEQUANT-G2,SCAL...   \n",
       "130  ACTGCTCCTA;GAATTCGAGC  SCALEQUANT-E1,SCALEQUANT-E2,SCALEQUANT-G2,SCAL...   \n",
       "131  CGAACTAAGG;GAATTCGAGC  SCALEQUANT-E1,SCALEQUANT-E2,SCALEQUANT-G2,SCAL...   \n",
       "132  GGAATCAGTA;GAATTCGAGC  SCALEQUANT-E1,SCALEQUANT-E2,SCALEQUANT-G2,SCAL...   \n",
       "133  GGCGGACGTT;GAATTCGAGC  SCALEQUANT-E1,SCALEQUANT-E2,SCALEQUANT-G2,SCAL...   \n",
       "134  GGCTGAATTG;GAATTCGAGC  SCALEQUANT-E1,SCALEQUANT-E2,SCALEQUANT-G2,SCAL...   \n",
       "135  GAGCGCGGCA;GAATTCGAGC  SCALEQUANT-E1,SCALEQUANT-E2,SCALEQUANT-G2,SCAL...   \n",
       "136  GCTCATGCTT;AGAGTTGCAA  SCALEQUANT-E1,SCALEQUANT-E2,SCALEQUANT-G2,SCAL...   \n",
       "137  AGGATGGACG;AGAGTTGCAA  SCALEQUANT-E1,SCALEQUANT-E2,SCALEQUANT-G2,SCAL...   \n",
       "138  ACTGCTCCTA;AGAGTTGCAA  SCALEQUANT-E1,SCALEQUANT-E2,SCALEQUANT-G2,SCAL...   \n",
       "139  CGAACTAAGG;AGAGTTGCAA  SCALEQUANT-E1,SCALEQUANT-E2,SCALEQUANT-G2,SCAL...   \n",
       "140  GGAATCAGTA;AGAGTTGCAA  SCALEQUANT-E1,SCALEQUANT-E2,SCALEQUANT-G2,SCAL...   \n",
       "141  GGCGGACGTT;AGAGTTGCAA  SCALEQUANT-E1,SCALEQUANT-E2,SCALEQUANT-G2,SCAL...   \n",
       "142  GGCTGAATTG;AGAGTTGCAA  SCALEQUANT-E1,SCALEQUANT-E2,SCALEQUANT-G2,SCAL...   \n",
       "143  GAGCGCGGCA;AGAGTTGCAA  SCALEQUANT-E1,SCALEQUANT-E2,SCALEQUANT-G2,SCAL...   \n",
       "144  CCAGAGGAAG;GCTCATGCTT  SCALEQUANT-E1,SCALEQUANT-E2,SCALEQUANT-G2,SCAL...   \n",
       "145  CCAGAGGAAG;AGGATGGACG  SCALEQUANT-E1,SCALEQUANT-E2,SCALEQUANT-G2,SCAL...   \n",
       "146  CCAGAGGAAG;ACTGCTCCTA  SCALEQUANT-E1,SCALEQUANT-E2,SCALEQUANT-G2,SCAL...   \n",
       "147  CCAGAGGAAG;CGAACTAAGG  SCALEQUANT-E1,SCALEQUANT-E2,SCALEQUANT-G2,SCAL...   \n",
       "148  CCAGAGGAAG;GGAATCAGTA  SCALEQUANT-E1,SCALEQUANT-E2,SCALEQUANT-G2,SCAL...   \n",
       "149  CCAGAGGAAG;GGCGGACGTT  SCALEQUANT-E1,SCALEQUANT-E2,SCALEQUANT-G2,SCAL...   \n",
       "150  CCAGAGGAAG;GGCTGAATTG  SCALEQUANT-E1,SCALEQUANT-E2,SCALEQUANT-G2,SCAL...   \n",
       "151  CCAGAGGAAG;GAGCGCGGCA  SCALEQUANT-E1,SCALEQUANT-E2,SCALEQUANT-G2,SCAL...   \n",
       "152  AGAACCGACC;GCTCATGCTT  SCALEQUANT-E1,SCALEQUANT-E2,SCALEQUANT-G2,SCAL...   \n",
       "153  AGAACCGACC;AGGATGGACG  SCALEQUANT-E1,SCALEQUANT-E2,SCALEQUANT-G2,SCAL...   \n",
       "154  AGAACCGACC;ACTGCTCCTA  SCALEQUANT-E1,SCALEQUANT-E2,SCALEQUANT-G2,SCAL...   \n",
       "155  AGAACCGACC;CGAACTAAGG  SCALEQUANT-E1,SCALEQUANT-E2,SCALEQUANT-G2,SCAL...   \n",
       "156  AGAACCGACC;GGAATCAGTA  SCALEQUANT-E1,SCALEQUANT-E2,SCALEQUANT-G2,SCAL...   \n",
       "157  AGAACCGACC;GGCGGACGTT  SCALEQUANT-E1,SCALEQUANT-E2,SCALEQUANT-G2,SCAL...   \n",
       "158  AGAACCGACC;GGCTGAATTG  SCALEQUANT-E1,SCALEQUANT-E2,SCALEQUANT-G2,SCAL...   \n",
       "159  AGAACCGACC;GAGCGCGGCA  SCALEQUANT-E1,SCALEQUANT-E2,SCALEQUANT-G2,SCAL...   \n",
       "\n",
       "                                        RT_index_oligo  lab_name JUMBO-sci  \\\n",
       "120  ACTAACCA,GTAGACAC,GCTAAGTT,CCAATTAG,AACCGTGT,A...  Trapnell     FALSE   \n",
       "121  ACTAACCA,GTAGACAC,GCTAAGTT,CCAATTAG,AACCGTGT,A...  Trapnell     FALSE   \n",
       "122  ACTAACCA,GTAGACAC,GCTAAGTT,CCAATTAG,AACCGTGT,A...  Trapnell     FALSE   \n",
       "123  ACTAACCA,GTAGACAC,GCTAAGTT,CCAATTAG,AACCGTGT,A...  Trapnell     FALSE   \n",
       "124  ACTAACCA,GTAGACAC,GCTAAGTT,CCAATTAG,AACCGTGT,A...  Trapnell     FALSE   \n",
       "125  ACTAACCA,GTAGACAC,GCTAAGTT,CCAATTAG,AACCGTGT,A...  Trapnell     FALSE   \n",
       "126  ACTAACCA,GTAGACAC,GCTAAGTT,CCAATTAG,AACCGTGT,A...  Trapnell     FALSE   \n",
       "127  ACTAACCA,GTAGACAC,GCTAAGTT,CCAATTAG,AACCGTGT,A...  Trapnell     FALSE   \n",
       "128  ACTAACCA,GTAGACAC,GCTAAGTT,CCAATTAG,AACCGTGT,A...  Trapnell     FALSE   \n",
       "129  ACTAACCA,GTAGACAC,GCTAAGTT,CCAATTAG,AACCGTGT,A...  Trapnell     FALSE   \n",
       "130  ACTAACCA,GTAGACAC,GCTAAGTT,CCAATTAG,AACCGTGT,A...  Trapnell     FALSE   \n",
       "131  ACTAACCA,GTAGACAC,GCTAAGTT,CCAATTAG,AACCGTGT,A...  Trapnell     FALSE   \n",
       "132  ACTAACCA,GTAGACAC,GCTAAGTT,CCAATTAG,AACCGTGT,A...  Trapnell     FALSE   \n",
       "133  ACTAACCA,GTAGACAC,GCTAAGTT,CCAATTAG,AACCGTGT,A...  Trapnell     FALSE   \n",
       "134  ACTAACCA,GTAGACAC,GCTAAGTT,CCAATTAG,AACCGTGT,A...  Trapnell     FALSE   \n",
       "135  ACTAACCA,GTAGACAC,GCTAAGTT,CCAATTAG,AACCGTGT,A...  Trapnell     FALSE   \n",
       "136  ACTAACCA,GTAGACAC,GCTAAGTT,CCAATTAG,AACCGTGT,A...  Trapnell     FALSE   \n",
       "137  ACTAACCA,GTAGACAC,GCTAAGTT,CCAATTAG,AACCGTGT,A...  Trapnell     FALSE   \n",
       "138  ACTAACCA,GTAGACAC,GCTAAGTT,CCAATTAG,AACCGTGT,A...  Trapnell     FALSE   \n",
       "139  ACTAACCA,GTAGACAC,GCTAAGTT,CCAATTAG,AACCGTGT,A...  Trapnell     FALSE   \n",
       "140  ACTAACCA,GTAGACAC,GCTAAGTT,CCAATTAG,AACCGTGT,A...  Trapnell     FALSE   \n",
       "141  ACTAACCA,GTAGACAC,GCTAAGTT,CCAATTAG,AACCGTGT,A...  Trapnell     FALSE   \n",
       "142  ACTAACCA,GTAGACAC,GCTAAGTT,CCAATTAG,AACCGTGT,A...  Trapnell     FALSE   \n",
       "143  ACTAACCA,GTAGACAC,GCTAAGTT,CCAATTAG,AACCGTGT,A...  Trapnell     FALSE   \n",
       "144  ACTAACCA,GTAGACAC,GCTAAGTT,CCAATTAG,AACCGTGT,A...  Trapnell     FALSE   \n",
       "145  ACTAACCA,GTAGACAC,GCTAAGTT,CCAATTAG,AACCGTGT,A...  Trapnell     FALSE   \n",
       "146  ACTAACCA,GTAGACAC,GCTAAGTT,CCAATTAG,AACCGTGT,A...  Trapnell     FALSE   \n",
       "147  ACTAACCA,GTAGACAC,GCTAAGTT,CCAATTAG,AACCGTGT,A...  Trapnell     FALSE   \n",
       "148  ACTAACCA,GTAGACAC,GCTAAGTT,CCAATTAG,AACCGTGT,A...  Trapnell     FALSE   \n",
       "149  ACTAACCA,GTAGACAC,GCTAAGTT,CCAATTAG,AACCGTGT,A...  Trapnell     FALSE   \n",
       "150  ACTAACCA,GTAGACAC,GCTAAGTT,CCAATTAG,AACCGTGT,A...  Trapnell     FALSE   \n",
       "151  ACTAACCA,GTAGACAC,GCTAAGTT,CCAATTAG,AACCGTGT,A...  Trapnell     FALSE   \n",
       "152  ACTAACCA,GTAGACAC,GCTAAGTT,CCAATTAG,AACCGTGT,A...  Trapnell     FALSE   \n",
       "153  ACTAACCA,GTAGACAC,GCTAAGTT,CCAATTAG,AACCGTGT,A...  Trapnell     FALSE   \n",
       "154  ACTAACCA,GTAGACAC,GCTAAGTT,CCAATTAG,AACCGTGT,A...  Trapnell     FALSE   \n",
       "155  ACTAACCA,GTAGACAC,GCTAAGTT,CCAATTAG,AACCGTGT,A...  Trapnell     FALSE   \n",
       "156  ACTAACCA,GTAGACAC,GCTAAGTT,CCAATTAG,AACCGTGT,A...  Trapnell     FALSE   \n",
       "157  ACTAACCA,GTAGACAC,GCTAAGTT,CCAATTAG,AACCGTGT,A...  Trapnell     FALSE   \n",
       "158  ACTAACCA,GTAGACAC,GCTAAGTT,CCAATTAG,AACCGTGT,A...  Trapnell     FALSE   \n",
       "159  ACTAACCA,GTAGACAC,GCTAAGTT,CCAATTAG,AACCGTGT,A...  Trapnell     FALSE   \n",
       "\n",
       "    pseudosample  ... development_stage (time)  \\\n",
       "120          bl3  ...                   96 hpf   \n",
       "121          bl3  ...                   96 hpf   \n",
       "122          bl3  ...                   96 hpf   \n",
       "123          bl3  ...                   96 hpf   \n",
       "124          bl3  ...                   96 hpf   \n",
       "125          bl3  ...                   96 hpf   \n",
       "126          bl3  ...                   96 hpf   \n",
       "127          bl3  ...                   96 hpf   \n",
       "128          bl3  ...                   96 hpf   \n",
       "129          bl3  ...                   96 hpf   \n",
       "130          bl3  ...                   96 hpf   \n",
       "131          bl3  ...                   96 hpf   \n",
       "132          bl3  ...                   96 hpf   \n",
       "133          bl3  ...                   96 hpf   \n",
       "134          bl3  ...                   96 hpf   \n",
       "135          bl3  ...                   96 hpf   \n",
       "136          bl3  ...                   96 hpf   \n",
       "137          bl3  ...                   96 hpf   \n",
       "138          bl3  ...                   96 hpf   \n",
       "139          bl3  ...                   96 hpf   \n",
       "140          bl3  ...                   96 hpf   \n",
       "141          bl3  ...                   96 hpf   \n",
       "142          bl3  ...                   96 hpf   \n",
       "143          bl3  ...                   96 hpf   \n",
       "144          bl3  ...                   96 hpf   \n",
       "145          bl3  ...                   96 hpf   \n",
       "146          bl3  ...                   96 hpf   \n",
       "147          bl3  ...                   96 hpf   \n",
       "148          bl3  ...                   96 hpf   \n",
       "149          bl3  ...                   96 hpf   \n",
       "150          bl3  ...                   96 hpf   \n",
       "151          bl3  ...                   96 hpf   \n",
       "152          bl3  ...                   96 hpf   \n",
       "153          bl3  ...                   96 hpf   \n",
       "154          bl3  ...                   96 hpf   \n",
       "155          bl3  ...                   96 hpf   \n",
       "156          bl3  ...                   96 hpf   \n",
       "157          bl3  ...                   96 hpf   \n",
       "158          bl3  ...                   96 hpf   \n",
       "159          bl3  ...                   96 hpf   \n",
       "\n",
       "                                                !assay  \\\n",
       "120  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "121  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "122  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "123  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "124  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "125  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "126  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "127  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "128  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "129  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "130  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "131  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "132  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "133  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "134  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "135  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "136  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "137  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "138  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "139  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "140  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "141  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "142  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "143  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "144  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "145  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "146  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "147  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "148  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "149  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "150  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "151  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "152  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "153  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "154  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "155  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "156  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "157  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "158  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "159  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "\n",
       "                                   assay tissue_type suspension_type  \\\n",
       "120  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "121  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "122  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "123  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "124  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "125  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "126  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "127  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "128  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "129  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "130  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "131  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "132  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "133  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "134  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "135  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "136  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "137  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "138  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "139  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "140  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "141  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "142  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "143  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "144  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "145  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "146  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "147  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "148  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "149  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "150  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "151  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "152  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "153  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "154  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "155  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "156  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "157  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "158  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "159  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "\n",
       "    !genetic_perturbation_strategy experimental_perturbation_type  \\\n",
       "120                no perturbation                       chemical   \n",
       "121                no perturbation                       chemical   \n",
       "122                no perturbation                       chemical   \n",
       "123                no perturbation                       chemical   \n",
       "124                no perturbation                       chemical   \n",
       "125                no perturbation                       chemical   \n",
       "126                no perturbation                       chemical   \n",
       "127                no perturbation                       chemical   \n",
       "128                no perturbation                       chemical   \n",
       "129                no perturbation                       chemical   \n",
       "130                no perturbation                       chemical   \n",
       "131                no perturbation                       chemical   \n",
       "132                no perturbation                       chemical   \n",
       "133                no perturbation                       chemical   \n",
       "134                no perturbation                       chemical   \n",
       "135                no perturbation                       chemical   \n",
       "136                no perturbation                       chemical   \n",
       "137                no perturbation                       chemical   \n",
       "138                no perturbation                       chemical   \n",
       "139                no perturbation                       chemical   \n",
       "140                no perturbation                       chemical   \n",
       "141                no perturbation                       chemical   \n",
       "142                no perturbation                       chemical   \n",
       "143                no perturbation                       chemical   \n",
       "144                no perturbation                       chemical   \n",
       "145                no perturbation                       chemical   \n",
       "146                no perturbation                       chemical   \n",
       "147                no perturbation                       chemical   \n",
       "148                no perturbation                       chemical   \n",
       "149                no perturbation                       chemical   \n",
       "150                no perturbation                       chemical   \n",
       "151                no perturbation                       chemical   \n",
       "152                no perturbation                       chemical   \n",
       "153                no perturbation                       chemical   \n",
       "154                no perturbation                       chemical   \n",
       "155                no perturbation                       chemical   \n",
       "156                no perturbation                       chemical   \n",
       "157                no perturbation                       chemical   \n",
       "158                no perturbation                       chemical   \n",
       "159                no perturbation                       chemical   \n",
       "\n",
       "    experimental_perturbation experimental_perturbation_concentration  \\\n",
       "120                  SB505124                           50 micromolar   \n",
       "121                  SB505124                           50 micromolar   \n",
       "122                  SB505124                           50 micromolar   \n",
       "123                  SB505124                           50 micromolar   \n",
       "124                  SB505124                           50 micromolar   \n",
       "125                  SB505124                           50 micromolar   \n",
       "126                  SB505124                           50 micromolar   \n",
       "127                  SB505124                           50 micromolar   \n",
       "128                    SU5402                           20 micromolar   \n",
       "129                    SU5402                           20 micromolar   \n",
       "130                    SU5402                           20 micromolar   \n",
       "131                    SU5402                           20 micromolar   \n",
       "132                    SU5402                           20 micromolar   \n",
       "133                    SU5402                           20 micromolar   \n",
       "134                    SU5402                           20 micromolar   \n",
       "135                    SU5402                           20 micromolar   \n",
       "136                      DMH1                           40 micromolar   \n",
       "137                      DMH1                           40 micromolar   \n",
       "138                      DMH1                           40 micromolar   \n",
       "139                      DMH1                           40 micromolar   \n",
       "140                      DMH1                           40 micromolar   \n",
       "141                      DMH1                           40 micromolar   \n",
       "142                      DMH1                           40 micromolar   \n",
       "143                      DMH1                           40 micromolar   \n",
       "144                  LY411575                           10 micromolar   \n",
       "145                  LY411575                           10 micromolar   \n",
       "146                  LY411575                           10 micromolar   \n",
       "147                  LY411575                           10 micromolar   \n",
       "148                  LY411575                           10 micromolar   \n",
       "149                  LY411575                           10 micromolar   \n",
       "150                  LY411575                           10 micromolar   \n",
       "151                  LY411575                           10 micromolar   \n",
       "152                      DMSO                            0 micromolar   \n",
       "153                      DMSO                            0 micromolar   \n",
       "154                      DMSO                            0 micromolar   \n",
       "155                      DMSO                            0 micromolar   \n",
       "156                      DMSO                            0 micromolar   \n",
       "157                      DMSO                            0 micromolar   \n",
       "158                      DMSO                            0 micromolar   \n",
       "159                      DMSO                            0 micromolar   \n",
       "\n",
       "    experimental_perturbation_time_point  \n",
       "120                               60 hpf  \n",
       "121                               60 hpf  \n",
       "122                               60 hpf  \n",
       "123                               60 hpf  \n",
       "124                               60 hpf  \n",
       "125                               60 hpf  \n",
       "126                               60 hpf  \n",
       "127                               60 hpf  \n",
       "128                               60 hpf  \n",
       "129                               60 hpf  \n",
       "130                               60 hpf  \n",
       "131                               60 hpf  \n",
       "132                               60 hpf  \n",
       "133                               60 hpf  \n",
       "134                               60 hpf  \n",
       "135                               60 hpf  \n",
       "136                               60 hpf  \n",
       "137                               60 hpf  \n",
       "138                               60 hpf  \n",
       "139                               60 hpf  \n",
       "140                               60 hpf  \n",
       "141                               60 hpf  \n",
       "142                               60 hpf  \n",
       "143                               60 hpf  \n",
       "144                               60 hpf  \n",
       "145                               60 hpf  \n",
       "146                               60 hpf  \n",
       "147                               60 hpf  \n",
       "148                               60 hpf  \n",
       "149                               60 hpf  \n",
       "150                               60 hpf  \n",
       "151                               60 hpf  \n",
       "152                               60 hpf  \n",
       "153                               60 hpf  \n",
       "154                               60 hpf  \n",
       "155                               60 hpf  \n",
       "156                               60 hpf  \n",
       "157                               60 hpf  \n",
       "158                               60 hpf  \n",
       "159                               60 hpf  \n",
       "\n",
       "[40 rows x 34 columns]), Pseudosample(local_h5ad='temp_plate/R096-25-0258.QSR-8_anndata.h5ad', parent=..., pseudosample_id='bl4', sample_id='R096-25-0258', allcells='temp_plate/R096-25-0258.QSR-8.allCells.csv', metadata_sample=              sample_name !experiment_name experiment_name hash_index  \\\n",
       "160   CHEM13_SCALE_A9_Bl4  CHEM13_RNA3-096        RNA3_096   SCALE-A9   \n",
       "161   CHEM13_SCALE_B9_Bl4  CHEM13_RNA3-096        RNA3_096   SCALE-B9   \n",
       "162   CHEM13_SCALE_C9_Bl4  CHEM13_RNA3-096        RNA3_096   SCALE-C9   \n",
       "163   CHEM13_SCALE_D9_Bl4  CHEM13_RNA3-096        RNA3_096   SCALE-D9   \n",
       "164   CHEM13_SCALE_E9_Bl4  CHEM13_RNA3-096        RNA3_096   SCALE-E9   \n",
       "165   CHEM13_SCALE_F9_Bl4  CHEM13_RNA3-096        RNA3_096   SCALE-F9   \n",
       "166   CHEM13_SCALE_G9_Bl4  CHEM13_RNA3-096        RNA3_096   SCALE-G9   \n",
       "167   CHEM13_SCALE_H9_Bl4  CHEM13_RNA3-096        RNA3_096   SCALE-H9   \n",
       "168  CHEM13_SCALE_A10_Bl4  CHEM13_RNA3-096        RNA3_096  SCALE-A10   \n",
       "169  CHEM13_SCALE_B10_Bl4  CHEM13_RNA3-096        RNA3_096  SCALE-B10   \n",
       "170  CHEM13_SCALE_C10_Bl4  CHEM13_RNA3-096        RNA3_096  SCALE-C10   \n",
       "171  CHEM13_SCALE_D10_Bl4  CHEM13_RNA3-096        RNA3_096  SCALE-D10   \n",
       "172  CHEM13_SCALE_E10_Bl4  CHEM13_RNA3-096        RNA3_096  SCALE-E10   \n",
       "173  CHEM13_SCALE_F10_Bl4  CHEM13_RNA3-096        RNA3_096  SCALE-F10   \n",
       "174  CHEM13_SCALE_G10_Bl4  CHEM13_RNA3-096        RNA3_096  SCALE-G10   \n",
       "175  CHEM13_SCALE_H10_Bl4  CHEM13_RNA3-096        RNA3_096  SCALE-H10   \n",
       "176  CHEM13_SCALE_A11_Bl4  CHEM13_RNA3-096        RNA3_096  SCALE-A11   \n",
       "177  CHEM13_SCALE_B11_Bl4  CHEM13_RNA3-096        RNA3_096  SCALE-B11   \n",
       "178  CHEM13_SCALE_C11_Bl4  CHEM13_RNA3-096        RNA3_096  SCALE-C11   \n",
       "179  CHEM13_SCALE_D11_Bl4  CHEM13_RNA3-096        RNA3_096  SCALE-D11   \n",
       "180  CHEM13_SCALE_E11_Bl4  CHEM13_RNA3-096        RNA3_096  SCALE-E11   \n",
       "181  CHEM13_SCALE_F11_Bl4  CHEM13_RNA3-096        RNA3_096  SCALE-F11   \n",
       "182  CHEM13_SCALE_G11_Bl4  CHEM13_RNA3-096        RNA3_096  SCALE-G11   \n",
       "183  CHEM13_SCALE_H11_Bl4  CHEM13_RNA3-096        RNA3_096  SCALE-H11   \n",
       "184  CHEM13_SCALE_A12_Bl4  CHEM13_RNA3-096        RNA3_096  SCALE-A12   \n",
       "185  CHEM13_SCALE_B12_Bl4  CHEM13_RNA3-096        RNA3_096  SCALE-B12   \n",
       "186  CHEM13_SCALE_C12_Bl4  CHEM13_RNA3-096        RNA3_096  SCALE-C12   \n",
       "187  CHEM13_SCALE_D12_Bl4  CHEM13_RNA3-096        RNA3_096  SCALE-D12   \n",
       "188  CHEM13_SCALE_E12_Bl4  CHEM13_RNA3-096        RNA3_096  SCALE-E12   \n",
       "189  CHEM13_SCALE_F12_Bl4  CHEM13_RNA3-096        RNA3_096  SCALE-F12   \n",
       "190  CHEM13_SCALE_G12_Bl4  CHEM13_RNA3-096        RNA3_096  SCALE-G12   \n",
       "191  CHEM13_SCALE_H12_Bl4  CHEM13_RNA3-096        RNA3_096  SCALE-H12   \n",
       "192   CHEM13_SCALE_A1_Bl4  CHEM13_RNA3-096        RNA3_096   SCALE-A1   \n",
       "193   CHEM13_SCALE_B1_Bl4  CHEM13_RNA3-096        RNA3_096   SCALE-B1   \n",
       "194   CHEM13_SCALE_C1_Bl4  CHEM13_RNA3-096        RNA3_096   SCALE-C1   \n",
       "195   CHEM13_SCALE_D1_Bl4  CHEM13_RNA3-096        RNA3_096   SCALE-D1   \n",
       "196   CHEM13_SCALE_E1_Bl4  CHEM13_RNA3-096        RNA3_096   SCALE-E1   \n",
       "197   CHEM13_SCALE_F1_Bl4  CHEM13_RNA3-096        RNA3_096   SCALE-F1   \n",
       "198   CHEM13_SCALE_G1_Bl4  CHEM13_RNA3-096        RNA3_096   SCALE-G1   \n",
       "199   CHEM13_SCALE_H1_Bl4  CHEM13_RNA3-096        RNA3_096   SCALE-H1   \n",
       "\n",
       "        nuclear_hash_oligo                                           RT_index  \\\n",
       "160  CGCTCCATGA;GCTCATGCTT  SCALEQUANT-F1,SCALEQUANT-F2,SCALEQUANT-F3,SCAL...   \n",
       "161  CGCTCCATGA;AGGATGGACG  SCALEQUANT-F1,SCALEQUANT-F2,SCALEQUANT-F3,SCAL...   \n",
       "162  CGCTCCATGA;ACTGCTCCTA  SCALEQUANT-F1,SCALEQUANT-F2,SCALEQUANT-F3,SCAL...   \n",
       "163  CGCTCCATGA;CGAACTAAGG  SCALEQUANT-F1,SCALEQUANT-F2,SCALEQUANT-F3,SCAL...   \n",
       "164  CGCTCCATGA;GGAATCAGTA  SCALEQUANT-F1,SCALEQUANT-F2,SCALEQUANT-F3,SCAL...   \n",
       "165  CGCTCCATGA;GGCGGACGTT  SCALEQUANT-F1,SCALEQUANT-F2,SCALEQUANT-F3,SCAL...   \n",
       "166  CGCTCCATGA;GGCTGAATTG  SCALEQUANT-F1,SCALEQUANT-F2,SCALEQUANT-F3,SCAL...   \n",
       "167  CGCTCCATGA;GAGCGCGGCA  SCALEQUANT-F1,SCALEQUANT-F2,SCALEQUANT-F3,SCAL...   \n",
       "168  CCAGACCATG;GCTCATGCTT  SCALEQUANT-F1,SCALEQUANT-F2,SCALEQUANT-F3,SCAL...   \n",
       "169  CCAGACCATG;AGGATGGACG  SCALEQUANT-F1,SCALEQUANT-F2,SCALEQUANT-F3,SCAL...   \n",
       "170  CCAGACCATG;ACTGCTCCTA  SCALEQUANT-F1,SCALEQUANT-F2,SCALEQUANT-F3,SCAL...   \n",
       "171  CCAGACCATG;CGAACTAAGG  SCALEQUANT-F1,SCALEQUANT-F2,SCALEQUANT-F3,SCAL...   \n",
       "172  CCAGACCATG;GGAATCAGTA  SCALEQUANT-F1,SCALEQUANT-F2,SCALEQUANT-F3,SCAL...   \n",
       "173  CCAGACCATG;GGCGGACGTT  SCALEQUANT-F1,SCALEQUANT-F2,SCALEQUANT-F3,SCAL...   \n",
       "174  CCAGACCATG;GGCTGAATTG  SCALEQUANT-F1,SCALEQUANT-F2,SCALEQUANT-F3,SCAL...   \n",
       "175  CCAGACCATG;GAGCGCGGCA  SCALEQUANT-F1,SCALEQUANT-F2,SCALEQUANT-F3,SCAL...   \n",
       "176  CTAGCGAGCC;GCTCATGCTT  SCALEQUANT-F1,SCALEQUANT-F2,SCALEQUANT-F3,SCAL...   \n",
       "177  CTAGCGAGCC;AGGATGGACG  SCALEQUANT-F1,SCALEQUANT-F2,SCALEQUANT-F3,SCAL...   \n",
       "178  CTAGCGAGCC;ACTGCTCCTA  SCALEQUANT-F1,SCALEQUANT-F2,SCALEQUANT-F3,SCAL...   \n",
       "179  CTAGCGAGCC;CGAACTAAGG  SCALEQUANT-F1,SCALEQUANT-F2,SCALEQUANT-F3,SCAL...   \n",
       "180  CTAGCGAGCC;GGAATCAGTA  SCALEQUANT-F1,SCALEQUANT-F2,SCALEQUANT-F3,SCAL...   \n",
       "181  CTAGCGAGCC;GGCGGACGTT  SCALEQUANT-F1,SCALEQUANT-F2,SCALEQUANT-F3,SCAL...   \n",
       "182  CTAGCGAGCC;GGCTGAATTG  SCALEQUANT-F1,SCALEQUANT-F2,SCALEQUANT-F3,SCAL...   \n",
       "183  CTAGCGAGCC;GAGCGCGGCA  SCALEQUANT-F1,SCALEQUANT-F2,SCALEQUANT-F3,SCAL...   \n",
       "184  GCTGCTGAAT;GCTCATGCTT  SCALEQUANT-F1,SCALEQUANT-F2,SCALEQUANT-F3,SCAL...   \n",
       "185  GCTGCTGAAT;AGGATGGACG  SCALEQUANT-F1,SCALEQUANT-F2,SCALEQUANT-F3,SCAL...   \n",
       "186  GCTGCTGAAT;ACTGCTCCTA  SCALEQUANT-F1,SCALEQUANT-F2,SCALEQUANT-F3,SCAL...   \n",
       "187  GCTGCTGAAT;CGAACTAAGG  SCALEQUANT-F1,SCALEQUANT-F2,SCALEQUANT-F3,SCAL...   \n",
       "188  GCTGCTGAAT;GGAATCAGTA  SCALEQUANT-F1,SCALEQUANT-F2,SCALEQUANT-F3,SCAL...   \n",
       "189  GCTGCTGAAT;GGCGGACGTT  SCALEQUANT-F1,SCALEQUANT-F2,SCALEQUANT-F3,SCAL...   \n",
       "190  GCTGCTGAAT;GGCTGAATTG  SCALEQUANT-F1,SCALEQUANT-F2,SCALEQUANT-F3,SCAL...   \n",
       "191  GCTGCTGAAT;GAGCGCGGCA  SCALEQUANT-F1,SCALEQUANT-F2,SCALEQUANT-F3,SCAL...   \n",
       "192  TGCGTTCCGG;GCTCATGCTT  SCALEQUANT-F1,SCALEQUANT-F2,SCALEQUANT-F3,SCAL...   \n",
       "193  TGCGTTCCGG;AGGATGGACG  SCALEQUANT-F1,SCALEQUANT-F2,SCALEQUANT-F3,SCAL...   \n",
       "194  TGCGTTCCGG;ACTGCTCCTA  SCALEQUANT-F1,SCALEQUANT-F2,SCALEQUANT-F3,SCAL...   \n",
       "195  TGCGTTCCGG;CGAACTAAGG  SCALEQUANT-F1,SCALEQUANT-F2,SCALEQUANT-F3,SCAL...   \n",
       "196  TGCGTTCCGG;GGAATCAGTA  SCALEQUANT-F1,SCALEQUANT-F2,SCALEQUANT-F3,SCAL...   \n",
       "197  TGCGTTCCGG;GGCGGACGTT  SCALEQUANT-F1,SCALEQUANT-F2,SCALEQUANT-F3,SCAL...   \n",
       "198  TGCGTTCCGG;GGCTGAATTG  SCALEQUANT-F1,SCALEQUANT-F2,SCALEQUANT-F3,SCAL...   \n",
       "199  TGCGTTCCGG;GAGCGCGGCA  SCALEQUANT-F1,SCALEQUANT-F2,SCALEQUANT-F3,SCAL...   \n",
       "\n",
       "                                        RT_index_oligo  lab_name JUMBO-sci  \\\n",
       "160  GACAACTA,CTTATGCT,ACACCAAG,CTCTTCTC,AAGACGTT,A...  Trapnell     FALSE   \n",
       "161  GACAACTA,CTTATGCT,ACACCAAG,CTCTTCTC,AAGACGTT,A...  Trapnell     FALSE   \n",
       "162  GACAACTA,CTTATGCT,ACACCAAG,CTCTTCTC,AAGACGTT,A...  Trapnell     FALSE   \n",
       "163  GACAACTA,CTTATGCT,ACACCAAG,CTCTTCTC,AAGACGTT,A...  Trapnell     FALSE   \n",
       "164  GACAACTA,CTTATGCT,ACACCAAG,CTCTTCTC,AAGACGTT,A...  Trapnell     FALSE   \n",
       "165  GACAACTA,CTTATGCT,ACACCAAG,CTCTTCTC,AAGACGTT,A...  Trapnell     FALSE   \n",
       "166  GACAACTA,CTTATGCT,ACACCAAG,CTCTTCTC,AAGACGTT,A...  Trapnell     FALSE   \n",
       "167  GACAACTA,CTTATGCT,ACACCAAG,CTCTTCTC,AAGACGTT,A...  Trapnell     FALSE   \n",
       "168  GACAACTA,CTTATGCT,ACACCAAG,CTCTTCTC,AAGACGTT,A...  Trapnell     FALSE   \n",
       "169  GACAACTA,CTTATGCT,ACACCAAG,CTCTTCTC,AAGACGTT,A...  Trapnell     FALSE   \n",
       "170  GACAACTA,CTTATGCT,ACACCAAG,CTCTTCTC,AAGACGTT,A...  Trapnell     FALSE   \n",
       "171  GACAACTA,CTTATGCT,ACACCAAG,CTCTTCTC,AAGACGTT,A...  Trapnell     FALSE   \n",
       "172  GACAACTA,CTTATGCT,ACACCAAG,CTCTTCTC,AAGACGTT,A...  Trapnell     FALSE   \n",
       "173  GACAACTA,CTTATGCT,ACACCAAG,CTCTTCTC,AAGACGTT,A...  Trapnell     FALSE   \n",
       "174  GACAACTA,CTTATGCT,ACACCAAG,CTCTTCTC,AAGACGTT,A...  Trapnell     FALSE   \n",
       "175  GACAACTA,CTTATGCT,ACACCAAG,CTCTTCTC,AAGACGTT,A...  Trapnell     FALSE   \n",
       "176  GACAACTA,CTTATGCT,ACACCAAG,CTCTTCTC,AAGACGTT,A...  Trapnell     FALSE   \n",
       "177  GACAACTA,CTTATGCT,ACACCAAG,CTCTTCTC,AAGACGTT,A...  Trapnell     FALSE   \n",
       "178  GACAACTA,CTTATGCT,ACACCAAG,CTCTTCTC,AAGACGTT,A...  Trapnell     FALSE   \n",
       "179  GACAACTA,CTTATGCT,ACACCAAG,CTCTTCTC,AAGACGTT,A...  Trapnell     FALSE   \n",
       "180  GACAACTA,CTTATGCT,ACACCAAG,CTCTTCTC,AAGACGTT,A...  Trapnell     FALSE   \n",
       "181  GACAACTA,CTTATGCT,ACACCAAG,CTCTTCTC,AAGACGTT,A...  Trapnell     FALSE   \n",
       "182  GACAACTA,CTTATGCT,ACACCAAG,CTCTTCTC,AAGACGTT,A...  Trapnell     FALSE   \n",
       "183  GACAACTA,CTTATGCT,ACACCAAG,CTCTTCTC,AAGACGTT,A...  Trapnell     FALSE   \n",
       "184  GACAACTA,CTTATGCT,ACACCAAG,CTCTTCTC,AAGACGTT,A...  Trapnell     FALSE   \n",
       "185  GACAACTA,CTTATGCT,ACACCAAG,CTCTTCTC,AAGACGTT,A...  Trapnell     FALSE   \n",
       "186  GACAACTA,CTTATGCT,ACACCAAG,CTCTTCTC,AAGACGTT,A...  Trapnell     FALSE   \n",
       "187  GACAACTA,CTTATGCT,ACACCAAG,CTCTTCTC,AAGACGTT,A...  Trapnell     FALSE   \n",
       "188  GACAACTA,CTTATGCT,ACACCAAG,CTCTTCTC,AAGACGTT,A...  Trapnell     FALSE   \n",
       "189  GACAACTA,CTTATGCT,ACACCAAG,CTCTTCTC,AAGACGTT,A...  Trapnell     FALSE   \n",
       "190  GACAACTA,CTTATGCT,ACACCAAG,CTCTTCTC,AAGACGTT,A...  Trapnell     FALSE   \n",
       "191  GACAACTA,CTTATGCT,ACACCAAG,CTCTTCTC,AAGACGTT,A...  Trapnell     FALSE   \n",
       "192  GACAACTA,CTTATGCT,ACACCAAG,CTCTTCTC,AAGACGTT,A...  Trapnell     FALSE   \n",
       "193  GACAACTA,CTTATGCT,ACACCAAG,CTCTTCTC,AAGACGTT,A...  Trapnell     FALSE   \n",
       "194  GACAACTA,CTTATGCT,ACACCAAG,CTCTTCTC,AAGACGTT,A...  Trapnell     FALSE   \n",
       "195  GACAACTA,CTTATGCT,ACACCAAG,CTCTTCTC,AAGACGTT,A...  Trapnell     FALSE   \n",
       "196  GACAACTA,CTTATGCT,ACACCAAG,CTCTTCTC,AAGACGTT,A...  Trapnell     FALSE   \n",
       "197  GACAACTA,CTTATGCT,ACACCAAG,CTCTTCTC,AAGACGTT,A...  Trapnell     FALSE   \n",
       "198  GACAACTA,CTTATGCT,ACACCAAG,CTCTTCTC,AAGACGTT,A...  Trapnell     FALSE   \n",
       "199  GACAACTA,CTTATGCT,ACACCAAG,CTCTTCTC,AAGACGTT,A...  Trapnell     FALSE   \n",
       "\n",
       "    pseudosample  ... development_stage (time)  \\\n",
       "160          bl4  ...                   96 hpf   \n",
       "161          bl4  ...                   96 hpf   \n",
       "162          bl4  ...                   96 hpf   \n",
       "163          bl4  ...                   96 hpf   \n",
       "164          bl4  ...                   96 hpf   \n",
       "165          bl4  ...                   96 hpf   \n",
       "166          bl4  ...                   96 hpf   \n",
       "167          bl4  ...                   96 hpf   \n",
       "168          bl4  ...                   96 hpf   \n",
       "169          bl4  ...                   96 hpf   \n",
       "170          bl4  ...                   96 hpf   \n",
       "171          bl4  ...                   96 hpf   \n",
       "172          bl4  ...                   96 hpf   \n",
       "173          bl4  ...                   96 hpf   \n",
       "174          bl4  ...                   96 hpf   \n",
       "175          bl4  ...                   96 hpf   \n",
       "176          bl4  ...                   96 hpf   \n",
       "177          bl4  ...                   96 hpf   \n",
       "178          bl4  ...                   96 hpf   \n",
       "179          bl4  ...                   96 hpf   \n",
       "180          bl4  ...                   96 hpf   \n",
       "181          bl4  ...                   96 hpf   \n",
       "182          bl4  ...                   96 hpf   \n",
       "183          bl4  ...                   96 hpf   \n",
       "184          bl4  ...                   96 hpf   \n",
       "185          bl4  ...                   96 hpf   \n",
       "186          bl4  ...                   96 hpf   \n",
       "187          bl4  ...                   96 hpf   \n",
       "188          bl4  ...                   96 hpf   \n",
       "189          bl4  ...                   96 hpf   \n",
       "190          bl4  ...                   96 hpf   \n",
       "191          bl4  ...                   96 hpf   \n",
       "192          bl4  ...                   96 hpf   \n",
       "193          bl4  ...                   96 hpf   \n",
       "194          bl4  ...                   96 hpf   \n",
       "195          bl4  ...                   96 hpf   \n",
       "196          bl4  ...                   96 hpf   \n",
       "197          bl4  ...                   96 hpf   \n",
       "198          bl4  ...                   96 hpf   \n",
       "199          bl4  ...                   96 hpf   \n",
       "\n",
       "                                                !assay  \\\n",
       "160  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "161  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "162  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "163  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "164  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "165  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "166  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "167  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "168  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "169  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "170  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "171  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "172  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "173  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "174  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "175  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "176  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "177  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "178  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "179  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "180  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "181  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "182  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "183  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "184  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "185  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "186  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "187  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "188  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "189  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "190  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "191  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "192  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "193  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "194  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "195  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "196  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "197  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "198  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "199  ScalePlex QuantumScale Single Cell RNA sequencing   \n",
       "\n",
       "                                   assay tissue_type suspension_type  \\\n",
       "160  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "161  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "162  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "163  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "164  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "165  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "166  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "167  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "168  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "169  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "170  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "171  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "172  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "173  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "174  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "175  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "176  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "177  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "178  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "179  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "180  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "181  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "182  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "183  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "184  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "185  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "186  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "187  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "188  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "189  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "190  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "191  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "192  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "193  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "194  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "195  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "196  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "197  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "198  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "199  ScaleBio single cell RNA sequencing      tissue         nucleus   \n",
       "\n",
       "    !genetic_perturbation_strategy experimental_perturbation_type  \\\n",
       "160                no perturbation                       chemical   \n",
       "161                no perturbation                       chemical   \n",
       "162                no perturbation                       chemical   \n",
       "163                no perturbation                       chemical   \n",
       "164                no perturbation                       chemical   \n",
       "165                no perturbation                       chemical   \n",
       "166                no perturbation                       chemical   \n",
       "167                no perturbation                       chemical   \n",
       "168                no perturbation                       chemical   \n",
       "169                no perturbation                       chemical   \n",
       "170                no perturbation                       chemical   \n",
       "171                no perturbation                       chemical   \n",
       "172                no perturbation                       chemical   \n",
       "173                no perturbation                       chemical   \n",
       "174                no perturbation                       chemical   \n",
       "175                no perturbation                       chemical   \n",
       "176                no perturbation                       chemical   \n",
       "177                no perturbation                       chemical   \n",
       "178                no perturbation                       chemical   \n",
       "179                no perturbation                       chemical   \n",
       "180                no perturbation                       chemical   \n",
       "181                no perturbation                       chemical   \n",
       "182                no perturbation                       chemical   \n",
       "183                no perturbation                       chemical   \n",
       "184                no perturbation                       chemical   \n",
       "185                no perturbation                       chemical   \n",
       "186                no perturbation                       chemical   \n",
       "187                no perturbation                       chemical   \n",
       "188                no perturbation                       chemical   \n",
       "189                no perturbation                       chemical   \n",
       "190                no perturbation                       chemical   \n",
       "191                no perturbation                       chemical   \n",
       "192                no perturbation                       chemical   \n",
       "193                no perturbation                       chemical   \n",
       "194                no perturbation                       chemical   \n",
       "195                no perturbation                       chemical   \n",
       "196                no perturbation                       chemical   \n",
       "197                no perturbation                       chemical   \n",
       "198                no perturbation                       chemical   \n",
       "199                no perturbation                       chemical   \n",
       "\n",
       "    experimental_perturbation experimental_perturbation_concentration  \\\n",
       "160                  SB505124                           50 micromolar   \n",
       "161                  SB505124                           50 micromolar   \n",
       "162                  SB505124                           50 micromolar   \n",
       "163                  SB505124                           50 micromolar   \n",
       "164                  SB505124                           50 micromolar   \n",
       "165                  SB505124                           50 micromolar   \n",
       "166                  SB505124                           50 micromolar   \n",
       "167                  SB505124                           50 micromolar   \n",
       "168                    SU5402                           20 micromolar   \n",
       "169                    SU5402                           20 micromolar   \n",
       "170                    SU5402                           20 micromolar   \n",
       "171                    SU5402                           20 micromolar   \n",
       "172                    SU5402                           20 micromolar   \n",
       "173                    SU5402                           20 micromolar   \n",
       "174                    SU5402                           20 micromolar   \n",
       "175                    SU5402                           20 micromolar   \n",
       "176                      DMH1                           40 micromolar   \n",
       "177                      DMH1                           40 micromolar   \n",
       "178                      DMH1                           40 micromolar   \n",
       "179                      DMH1                           40 micromolar   \n",
       "180                      DMH1                           40 micromolar   \n",
       "181                      DMH1                           40 micromolar   \n",
       "182                      DMH1                           40 micromolar   \n",
       "183                      DMH1                           40 micromolar   \n",
       "184                  LY411575                           10 micromolar   \n",
       "185                  LY411575                           10 micromolar   \n",
       "186                  LY411575                           10 micromolar   \n",
       "187                  LY411575                           10 micromolar   \n",
       "188                  LY411575                           10 micromolar   \n",
       "189                  LY411575                           10 micromolar   \n",
       "190                  LY411575                           10 micromolar   \n",
       "191                  LY411575                           10 micromolar   \n",
       "192                      DMSO                            0 micromolar   \n",
       "193                      DMSO                            0 micromolar   \n",
       "194                      DMSO                            0 micromolar   \n",
       "195                      DMSO                            0 micromolar   \n",
       "196                      DMSO                            0 micromolar   \n",
       "197                      DMSO                            0 micromolar   \n",
       "198                      DMSO                            0 micromolar   \n",
       "199                      DMSO                            0 micromolar   \n",
       "\n",
       "    experimental_perturbation_time_point  \n",
       "160                               84 hpf  \n",
       "161                               84 hpf  \n",
       "162                               84 hpf  \n",
       "163                               84 hpf  \n",
       "164                               84 hpf  \n",
       "165                               84 hpf  \n",
       "166                               84 hpf  \n",
       "167                               84 hpf  \n",
       "168                               84 hpf  \n",
       "169                               84 hpf  \n",
       "170                               84 hpf  \n",
       "171                               84 hpf  \n",
       "172                               84 hpf  \n",
       "173                               84 hpf  \n",
       "174                               84 hpf  \n",
       "175                               84 hpf  \n",
       "176                               84 hpf  \n",
       "177                               84 hpf  \n",
       "178                               84 hpf  \n",
       "179                               84 hpf  \n",
       "180                               84 hpf  \n",
       "181                               84 hpf  \n",
       "182                               84 hpf  \n",
       "183                               84 hpf  \n",
       "184                               84 hpf  \n",
       "185                               84 hpf  \n",
       "186                               84 hpf  \n",
       "187                               84 hpf  \n",
       "188                               84 hpf  \n",
       "189                               84 hpf  \n",
       "190                               84 hpf  \n",
       "191                               84 hpf  \n",
       "192                               84 hpf  \n",
       "193                               84 hpf  \n",
       "194                               84 hpf  \n",
       "195                               84 hpf  \n",
       "196                               84 hpf  \n",
       "197                               84 hpf  \n",
       "198                               84 hpf  \n",
       "199                               84 hpf  \n",
       "\n",
       "[40 rows x 34 columns])], sample_map={'bl1': 'R096-25-0255', 'bl2': 'R096-25-0256', 'bl3': 'R096-25-0257', 'bl4': 'R096-25-0258'})"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "experiment_04"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "f161c979-c78b-4114-878e-897f5d9d376c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs  n_vars = 46006  31551\n",
       "    obs: 'cell_id', 'counts', 'genes', 'totalReads', 'countedReads', 'mappedReads', 'geneReads', 'exonReads', 'intronReads', 'antisenseReads', 'mitoReads', 'countedMultiGeneReads', 'Saturation', 'mitoProp', 'PCR', 'RT', 'bead_bc', 'sample', 'flags', 'passing_scaleplex', 'assigned_scaleplex'\n",
       "    var: 'gene_id', 'gene_symbol', 'annotation'"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "experiment_04.all_pseudosamples[0].adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "60245d65-7c40-412d-9bb9-48f77d9ae52b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs  n_vars = 46006  31551\n",
       "    obs: 'cell_id', 'cell_counts', 'cell_genes', 'cell_totalReads', 'cell_countedReads', 'cell_mappedReads', 'cell_geneReads', 'cell_exonReads', 'cell_intronReads', 'cell_antisenseReads', 'cell_mitoReads', 'cell_countedMultiGeneReads', 'cell_Saturation', 'cell_mitoProp', 'PCR', 'RT', 'bead_bc', 'sample', 'flags', 'passing_scaleplex', 'assigned_scaleplex', 'assigned_hash_index', 'sample_name', '!experiment_name', 'experiment_name', 'hash_index', 'nuclear_hash_oligo', 'RT_index', 'RT_index_oligo', 'lab_name', 'JUMBO-sci', 'pseudosample', 'is_pilot_data', 'donor_id', 'strain', 'target', 'organism', 'sex', 'self_reported_ethnicity', 'disease', '!tissue', 'tissue', 'preservation_method', 'dissociation_method', '!development_stage', 'development_stage', 'development_stage (time)', '!assay', 'assay', 'tissue_type', 'suspension_type', '!genetic_perturbation_strategy', 'experimental_perturbation_type', 'experimental_perturbation', 'experimental_perturbation_concentration', 'experimental_perturbation_time_point'\n",
       "    var: 'gene_id', 'gene_symbol', 'annotation'"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "experiment_04.all_pseudosamples[0].adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "2e523a36-4ff0-4193-a2aa-3be9de574212",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PCR\n",
       "QSR-8    46006\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "experiment_04.all_pseudosamples[0].adata.obs['PCR'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "99da15d5-394f-4393-ac3d-b1c9d1fcc26d",
   "metadata": {},
   "outputs": [],
   "source": [
    "DOWN_DIR = 'temp_plate'\n",
    "\n",
    "@dataclass\n",
    "class URIPath:\n",
    "    '''\n",
    "    Dataclass to hold various aspects of a single run of SeqSuite.\n",
    "    This path should include the entire URI path, including bucket name, of the SeqSuite rundate output files.\n",
    "    Ex: s3://czi-novogene/trapnell-seahub-bcp/NVUS2024101701-04/RNA3_096/processed/Run_2025-12-15_biohub/\n",
    "    sample_files includes all files associated with this URIPath that are needed\n",
    "    '''\n",
    "    \n",
    "    full_uri: str\n",
    "\n",
    "    def __post_init__(self):\n",
    "        if not self.full_uri.endswith('/'):\n",
    "            self.full_uri += '/'\n",
    "        self.bucket_name: str = self.full_uri.split('/')[2]\n",
    "        self.prefix: str = '/'.join(self.full_uri.split('/')[3:])\n",
    "        self.sample_files = fs.ls(f's3://{self.bucket_name}/{self.prefix}samples/')\n",
    "        self.sample_files.append(f'{self.bucket_name}/{self.prefix}samples.csv')\n",
    "        self.experiment = self.full_uri.split('/')[5]\n",
    "\n",
    "\n",
    "@dataclass\n",
    "class LatticeMetadata:\n",
    "    '''\n",
    "    Dataclass to hold and map ontologies to a Lattice metadata spreadsheet.\n",
    "    The functions allow for this object to read in a specific tab for a google sheet into a dataframe\n",
    "    '''\n",
    "    \n",
    "    sheet_id: str\n",
    "    tab_name: str\n",
    "    uri: URIPath\n",
    "\n",
    "    def get_gid(self):\n",
    "        '''\n",
    "        Given sheet id and tab name, return gid\n",
    "        '''\n",
    "        sheet_url = f'https://docs.google.com/spreadsheets/d/{self.sheet_id}'\n",
    "        req = Request(sheet_url, headers={'User-Agent' : \"Magic Browser\"})\n",
    "        s = urlopen(req)\n",
    "        soup = BeautifulSoup(s, 'html.parser')\n",
    "        tab_ids = {}\n",
    "        pattern = re.compile('var bootstrapData = (.*?)};')\n",
    "        for s in soup.find_all('script'):\n",
    "            if pattern.search(str(s)):\n",
    "                d = pattern.search(str(s)).group()[20:-1]\n",
    "                data = json.loads(d)\n",
    "                for t in data['changes']['topsnapshot']:\n",
    "                    u = t[1].split('\"')\n",
    "                    if len(u) > 5:\n",
    "                        tab_ids[u[5]] = u[1]\n",
    "        return tab_ids[self.tab_name]\n",
    "\n",
    "    def get_metadata_df(self):\n",
    "        '''\n",
    "        Given sheet id and gid, return lattice metadata in a dataframe, subset by experiment name\n",
    "        '''\n",
    "        url = f'https://docs.google.com/spreadsheets/d/{self.sheet_id}/export?format=csv&gid={self.gid}'\n",
    "        response = requests.get(url)\n",
    "        sample_df = pd.read_csv(BytesIO(response.content), comment=\"#\", dtype=str).dropna(axis=1,how='all')\n",
    "        sample_df = sample_df[sample_df['experiment_name'] == self.uri.experiment]\n",
    "        return sample_df\n",
    "\n",
    "    def __post_init__(self):\n",
    "        self.gid = self.get_gid()\n",
    "        self.metadata_df = self.get_metadata_df()\n",
    "\n",
    "\n",
    "@dataclass\n",
    "class Pseudosample:\n",
    "    '''\n",
    "    The pooled pseudosample entity that by which there is a single h5ad file.\n",
    "    This object will track associated metadata and AnnData object\n",
    "    '''\n",
    "    local_h5ad: str\n",
    "    parent: Experiment\n",
    "    pseudosample_id: str\n",
    "    sample_id: str\n",
    "    allcells: None\n",
    "    metadata_sample: pd.DataFrame\n",
    "\n",
    "    def load_h5ad(self):\n",
    "        '''\n",
    "        Load h5ad as AnnData object using local_h5ad variable\n",
    "        '''\n",
    "        return ad.read_h5ad(self.local_h5ad)\n",
    "\n",
    "    def assign_oligo(self):\n",
    "        '''\n",
    "        For SingleSublibrary experiments, the observations need to have the hash oligo mapped using the *allCells.csv file\n",
    "        '''\n",
    "        hash_df = pd.read_csv(self.allcells, index_col = 'cell_id')\n",
    "        self.adata.obs = pd.merge(self.adata.obs, hash_df[['passing_scaleplex', 'assigned_scaleplex']], left_index=True, right_index=True, how='left')\n",
    "        self.adata.obs['assigned_scaleplex'] = self.adata.obs['assigned_scaleplex'].astype('string')\n",
    "    \n",
    "    def drop_unassigned(self):\n",
    "        '''\n",
    "        given an allCell.csv, with \"cell_id\" colum for cellID and \"assigned_scaleplex\" for scaleplex hash assignment,\n",
    "        merge \"assigned_scaleplex\" to adata.obs, where the well number and letter need to be swapped (eg: 7E -> E7) and add prefix \"SCALE-\"\n",
    "        to match the Lattice metadata spreadsheet. Drop cells that have the following values as \"assigned_scaleplex\":\n",
    "         - Indeterminate\n",
    "         - Max_Fail\n",
    "         - Unexpected\n",
    "        '''\n",
    "        drop_values = ['Indeterminate','Max_Fail','Unexpected']\n",
    "        adata_subset = self.adata.copy()\n",
    "        adata_subset = adata_subset[~adata_subset.obs['assigned_scaleplex'].isin(drop_values)].copy()\n",
    "        self.adata = adata_subset\n",
    "\n",
    "    def rename_cols(self):\n",
    "        '''\n",
    "        For the cell level metrics, append 'cell_' to field name, as there will be sample level metrics\n",
    "        '''\n",
    "        metrics_list = [\n",
    "            'counts',\n",
    "            'genes',\n",
    "            'totalReads',\n",
    "            'countedReads',\n",
    "            'mappedReads',\n",
    "            'geneReads',\n",
    "            'exonReads',\n",
    "            'intronReads',\n",
    "            'antisenseReads',\n",
    "            'mitoReads',\n",
    "            'countedMultiGeneReads',\n",
    "            'Saturation',\n",
    "            'mitoProp'\n",
    "        ]\n",
    "        for col in metrics_list:\n",
    "            self.adata.obs.rename(columns = {col: f'cell_{col}'}, inplace = True)\n",
    "    \n",
    "    def map_metadata(self):\n",
    "        '''\n",
    "        Add additional metadata to obs from spreadsheet\n",
    "        '''\n",
    "        self.adata.obs['assigned_hash_index'] = self.adata.obs['assigned_scaleplex'].str.replace(r'(\\d+)([a-zA-Z])', r'SCALE-\\2\\1', regex=True).astype('string')\n",
    "        self.adata.obs = pd.merge(self.adata.obs, self.metadata_sample, left_on='assigned_hash_index', right_on='hash_index', how='left').set_index(self.adata.obs.index)\n",
    "\n",
    "    def __post_init__(self):\n",
    "        self.adata = self.load_h5ad()\n",
    "        if isinstance(self.parent, SingleSublibrary):\n",
    "            self.assign_oligo()\n",
    "        self.drop_unassigned()\n",
    "        self.rename_cols()\n",
    "        self.map_metadata()\n",
    "\n",
    "\n",
    "@dataclass\n",
    "class Experiment:\n",
    "    '''\n",
    "    Dataclass to hold various aspects of a single experiment, including\n",
    "    the pseudosamples and corresponding sample names found in SeqSuite run\n",
    "    '''\n",
    "    metadata: LatticeMetadata\n",
    "    uri: URIPath\n",
    "    all_pseudosamples: list[Pseudosample] = field(init=False, default_factory=list)\n",
    "    sample_map: dict[str, str] = field(init=False, default_factory=dict)\n",
    "\n",
    "    def expand_ranges(self, value):\n",
    "        '''\n",
    "        Given a value found in 'barcodes' column in samples.csv, return the expanded form, where no letters are skipped (ex: 1A-1G)\n",
    "        '''\n",
    "        expanded = []\n",
    "        for sub_string in value.split(';'):\n",
    "            if re.search('-', sub_string):\n",
    "                # Match the prefix and the range of letters (e.g., 1A-1C)\n",
    "                # Group 1: Number prefix (e.g., '1')\n",
    "                # Group 2: Start letter (e.g., 'A')\n",
    "                # Group 3: End letter (e.g., 'C')\n",
    "                match = re.match(r'(\\d+)([A-Z])-(\\d+)?([A-Z])', sub_string)\n",
    "                if match:\n",
    "                    prefix, start_char, _, end_char = match.groups()\n",
    "                    # Iterate through the ASCII values of the characters\n",
    "                    for char_code in range(ord(start_char), ord(end_char) + 1):\n",
    "                        expanded.append(f\"{prefix}{chr(char_code)}\")\n",
    "            else:\n",
    "                expanded.append(sub_string)\n",
    "        return ';'.join(expanded)\n",
    "\n",
    "    \n",
    "    def get_sample_ids_map(self):\n",
    "        '''\n",
    "        Given a lattice metadata and a samples.csv file, map the pseudosample and sample ids using 'RT_index' and 'barcodes' columns, respectively.\n",
    "        Need to split metadata by '-' and then swap plate row/column, and then also expand sample csv if there is shorthand (ex: 1A-1G,8A-8G)\n",
    "        '''\n",
    "        samples_csv_df = pd.read_csv(f'{DOWN_DIR}/samples.csv', index_col='sample')\n",
    "        metadata_df = self.metadata.metadata_df[['RT_index','pseudosample']].drop_duplicates().set_index('pseudosample')\n",
    "        \n",
    "        # After reading in csv and metadata df, parse barcode/RT_index so that they are comparable and store in dictionary\n",
    "        metadata_df['RT_transformed'] =  metadata_df['RT_index'].str.replace(r'[A-Za-z0-9]+-([A-H])([0-9]+)', r'\\2\\1', regex=True)\n",
    "        metadata_df['RT_transformed'] = metadata_df['RT_transformed'].apply(lambda x: ';'.join(sorted(x.split(','))) if isinstance(x, str) else x)\n",
    "        metadata_map = metadata_df['RT_transformed'].to_dict()\n",
    "        samples_csv_df['barcodes_transformed'] = samples_csv_df['barcodes'].apply(lambda x: self.expand_ranges(x))\n",
    "        samples_csv_df['barcodes_transformed'] = samples_csv_df['barcodes_transformed'].apply(lambda x: ';'.join(sorted(x.split(';'))) if isinstance(x, str) else x)\n",
    "        samples_csv_map = samples_csv_df['barcodes_transformed'].to_dict()\n",
    "\n",
    "        # Compare index values and return mapping of sample IDs\n",
    "        # This also checks that the samples.csv file is matching the metadata spreadsheet for wells\n",
    "        sample_map = {}\n",
    "        for pseudosample, value in metadata_map.items():\n",
    "            sample_ids = [k for k, v in samples_csv_map.items() if v==value]\n",
    "            if (len(sample_ids)>1):\n",
    "                print(f'ERROR: multimapping indices\\t{pseudosample}\\t{value}')\n",
    "            elif (len(sample_ids)==0):\n",
    "                print(f'ERROR: indices not matching\\t{pseudosample}\\t{value}')\n",
    "            else:\n",
    "                sample_map[pseudosample] = sample_ids[0]\n",
    "        self.sample_map = sample_map\n",
    "\n",
    "    \n",
    "    def download_files(self):\n",
    "        '''\n",
    "        Download all files needed for this experiment\n",
    "        '''\n",
    "        os.makedirs(DOWN_DIR, exist_ok=True)\n",
    "        for s3_uri, local_path in self.file_dict.items():\n",
    "            if not os.path.exists(local_path):\n",
    "                print(f'Downloading from {s3_uri} to {local_path}')\n",
    "                fs.get(s3_uri, local_path)\n",
    "\n",
    "\n",
    "    def add_pseudosamples(self):\n",
    "        '''\n",
    "        Create all pseudosample object given sample_map and determine metadata for each.\n",
    "        Only load allcells for SingleSublibrary object\n",
    "        '''\n",
    "        for pseudo_id, samp_id in self.sample_map.items():\n",
    "            allcells_val = find_file(self.file_dict, samp_id, 'allCells.csv') if isinstance(self, SingleSublibrary) else None\n",
    "            print(f'Loading {samp_id}')\n",
    "            pseudosample = Pseudosample(\n",
    "                local_h5ad = find_file(self.file_dict, samp_id, 'h5ad'),\n",
    "                pseudosample_id = pseudo_id,\n",
    "                parent = self,\n",
    "                sample_id = samp_id,\n",
    "                allcells = allcells_val,\n",
    "                metadata_sample = self.metadata.metadata_df[self.metadata.metadata_df['pseudosample']==pseudo_id]\n",
    "            )\n",
    "            self.all_pseudosamples.append(pseudosample)\n",
    "        \n",
    "    def __post_init__(self):\n",
    "        self.sample_map: dict = get_sample_ids_map()\n",
    "\n",
    "\n",
    "class SingleSublibrary(Experiment):\n",
    "    '''\n",
    "    Plate Experiment to store specific variables and methods for an experiment that is a single sublibrary,\n",
    "    which will include *allCell.csv files in addition to the sublibrary specific h5ad\n",
    "    '''\n",
    "    \n",
    "    def determine_files(self):\n",
    "        '''\n",
    "        Return the actual file URIs of files that need to be downloaded\n",
    "        '''\n",
    "        file_dict = {}\n",
    "        for file in self.uri.sample_files:\n",
    "            if file.endswith(('allCells.csv','_anndata.h5ad')):\n",
    "                filename = file.split('/')[-1]\n",
    "                local_path = os.path.join(DOWN_DIR, filename)\n",
    "                file_dict[file] = local_path\n",
    "        return file_dict\n",
    "        \n",
    "    def __post_init__(self):\n",
    "        self.file_dict = self.determine_files()\n",
    "\n",
    "\n",
    "class MultiSublibrary(Experiment):\n",
    "    '''\n",
    "    Plate Experiment to store specific variables and methods for an experiment that has multiple sublibraries,\n",
    "    which will include the sample merged h5ad files\n",
    "    '''\n",
    "    def determine_files(self):\n",
    "        '''\n",
    "        Return the actual file URIs of files that need to be downloaded.\n",
    "        file_dict: dictionary where key is uri and value is local_path\n",
    "        '''\n",
    "        file_dict = {}\n",
    "        for file in self.uri.sample_files:\n",
    "            if file.endswith('anndata.h5ad') and not re.search('QSR', file):\n",
    "                filename = file.split('/')[-1]\n",
    "                local_path = os.path.join(DOWN_DIR, filename)\n",
    "                file_dict[file] = local_path\n",
    "        return file_dict\n",
    "\n",
    "    def __post_init__(self):\n",
    "        self.file_dict = self.determine_files()      \n",
    "\n",
    "\n",
    "def is_single(uripath_obj):\n",
    "    '''\n",
    "    Given an URI object, determine if the experiment is a single sublibrary experiment\n",
    "    by searching for 'QSR-#' for files on S3\n",
    "    '''\n",
    "    files = uripath_obj.sample_files\n",
    "    pattern = r'\\.(QSR-[1-8])_anndata.h5ad'\n",
    "    \n",
    "    sublibraries = list({re.search(pattern, f).group(1) for f in files if re.search(pattern, f)})\n",
    "    return len(sublibraries) == 1\n",
    "\n",
    "\n",
    "def find_file(file_dict, prefix, suffix):\n",
    "    '''\n",
    "    Given prefix and suffix, return file in file_dict\n",
    "    '''\n",
    "    matches = (f for f in list(file_dict.values())\n",
    "       if f.split('/')[-1].startswith(prefix) \n",
    "       and f.endswith(suffix))\n",
    "    return next(matches, None)\n",
    "\n",
    "\n",
    "def cxg_add_labels(adata):\n",
    "    adata.obs['cell_type_ontology_term_id'] = 'unknown'\n",
    "    labeler = AnnDataLabelAppender(adata)\n",
    "    labeler._add_labels()\n",
    "    adata.obs.drop(columns=['cell_type_ontology_term_id','cell_type'],inplace=True)\n",
    "\n",
    "    schema_v = labeler.schema_version\n",
    "    adata.uns['schema_version'] = schema_v\n",
    "    adata.uns['schema_reference'] = labeler._build_schema_reference_url(schema_v)\n",
    "\n",
    "\n",
    "def add_guide_metadata(adata, sheet, guide_gid):\n",
    "    '''\n",
    "    Add guide metadata into adata.uns from Lattice wrangling sheet\n",
    "    \n",
    "    :param obj adata: the anndata object that is being transformed into the curated matrix\n",
    "    :param obj guide_df: the dataframe containing guide metadata from wrangling sheet\n",
    "    \n",
    "    :returns obj adata: modified adata to contain guide metadata\n",
    "    '''\n",
    "    url = f'https://docs.google.com/spreadsheets/d/{sheet}/export?format=csv&gid={guide_gid}'\n",
    "    response = requests.get(url)\n",
    "    guide_df = pd.read_csv(BytesIO(response.content), comment=\"#\", dtype=str)\n",
    "    genetic_perturbations = {}\n",
    "    \n",
    "    for row in guide_df.itertuples():\n",
    "        genetic_perturbations[row.guide_id] = {}\n",
    "        genetic_perturbations[row.guide_id]['role'] = 'targeting' if row.guide_role == 'Targeting a Gene' else 'control'\n",
    "        genetic_perturbations[row.guide_id]['protospacer_sequence'] = row.guide_protospacer\n",
    "        genetic_perturbations[row.guide_id]['protospacer_adjacent_motif'] = row.guide_PAM\n",
    "        if not pd.isna([row.start,row.end,row.strand]).all():\n",
    "            chr_loc = str(row.chromosome).replace(\"chr\",\"\") + \":\" + str(row.start) + \"-\" + str(row.end) + \"(\" + str(row.strand) + \")\"\n",
    "            genetic_perturbations[row.guide_id]['target_genomic_regions'] = [chr_loc]\n",
    "        if not pd.isna(row.overlapping_gene_ids):\n",
    "            genetic_perturbations[row.guide_id]['target_features'] = {}\n",
    "            for i in range(len(row.overlapping_gene_ids.split(\";\"))):\n",
    "                genetic_perturbations[row.guide_id]['target_features'][row.overlapping_gene_ids.split(\";\")[i]] = row.overlapping_gene_names.split(\";\")[i]\n",
    "                                                                             \n",
    "            \n",
    "    adata.uns['genetic_perturbations'] = genetic_perturbations\n",
    "    \n",
    "    return adata\n",
    "\n",
    "\n",
    "def determine_perturbation_strategy(adata):\n",
    "    '''\n",
    "    Assess feature_call from protospacer_calls_per_cell.csv, where if all guides\n",
    "    assigned to a single cell are all control, then 'control'. Otherwise, it is \"no perturbations\"\n",
    "    if no guids or one of the following if targeting:\n",
    "        - \"CRISPR activation screen\"\n",
    "        - \"CRISPR interference screen\"\n",
    "        - \"CRISPR knockout mutant\"\n",
    "        - \"CRISPR knockout screen\"\n",
    "    \n",
    "    :param obj adata: the anndata object that is being transformed into the curated matrix\n",
    "\n",
    "    :returns obj adata: modified adata to contain perturbation_strategy as cell metadata\n",
    "    '''\n",
    "    adata.obs['genetic_perturbation_strategy_calculated'] = adata.obs['genetic_perturbation_id']\n",
    "    adata.obs['genetic_perturbation_strategy_calculated'] = adata.obs['genetic_perturbation_strategy_calculated'].apply(\n",
    "        lambda x: x.split(' || ') if pd.notna(x) else 'no perturbations'\n",
    "    )\n",
    "    adata.obs['genetic_perturbation_strategy_calculated'] = adata.obs['genetic_perturbation_strategy_calculated'].apply(\n",
    "        lambda x: [adata.uns['genetic_perturbations'][i]['role'] for i in x] if isinstance(x, list)\n",
    "            else x        \n",
    "    )\n",
    "    adata.obs['genetic_perturbation_strategy_calculated'] = adata.obs['genetic_perturbation_strategy_calculated'].apply(\n",
    "         lambda x: 'control' if isinstance(x, list) and 'targeting' not in set(x)\n",
    "            else x\n",
    "    )\n",
    "    adata.obs.loc[adata.obs['genetic_perturbation_strategy_calculated']=='control', 'genetic_perturbation_strategy'] = 'control'\n",
    "    adata.obs.loc[adata.obs['genetic_perturbation_strategy_calculated']=='no perturbations', 'genetic_perturbation_strategy'] = 'no perturbations'\n",
    "    adata.obs.drop(columns=['genetic_perturbation_strategy_calculated'], inplace=True)\n",
    "    \n",
    "    return adata\n",
    "\n",
    "\n",
    "\n",
    "def map_ontologies(sample_df):\n",
    "    '''\n",
    "    Takes the sample metadata dataframe and standardizes ontologies\n",
    "    Also checks that standard fields are only filled out for appropriate organism\n",
    "\n",
    "    :param dataframe sample_df: the sample metadata from given google sheet\n",
    "\n",
    "    :returns dataframe sample_df: sample metadata with ontologies added\n",
    "    '''\n",
    "    col_ont_map = {\n",
    "        'organism':'NCBITaxon',\n",
    "        'sex':'PATO',\n",
    "        'self_reported_ethnicity':{'NCBITaxon:9606':'HANCESTRO',\n",
    "                                   'other':'none'},\n",
    "        'disease':'MONDO',\n",
    "        'assay':'EFO',\n",
    "        'development_stage':{'NCBITaxon:6239':'WBls', # C. Elegans\n",
    "                             'NCBITaxon:7227':'FBdv', # Drosophila\n",
    "                             'NCBITaxon:10090':'MmusDv', # Mouse\n",
    "                             'NCBITaxon:7955':'ZFS', # Zebrafish\n",
    "                             'other':'HsapDv' # For all other organisms, use HsapDv\n",
    "                            },\n",
    "        'tissue':{'NCBITaxon:6239':'WBbt', # C. Elegans\n",
    "                  'NCBITaxon:7227':'FBbt', # Drosophila\n",
    "                  'NCBITaxon:7955':'ZFA', # Zebrafish\n",
    "                  'other':'UBERON' # For all other organisms, use UBERON\n",
    "                 }\n",
    "    }\n",
    "    ontology_parser = OntologyParser()\n",
    "    ont_err_lst = []\n",
    "    \n",
    "    for col in col_ont_map:\n",
    "        map_dict = {}\n",
    "        for label in sample_df[col].unique():\n",
    "            term_id = None\n",
    "            if col == 'disease' and label == 'normal': # Normal is not in MONDO ontology\n",
    "                term_id = 'PATO:0000461'\n",
    "            elif label in ['unknown','na']: # Unknown and na won't be in ontologies, pass along\n",
    "                map_dict[label] = label\n",
    "                continue\n",
    "            elif col in ['tissue','development_stage','self_reported_ethnicity']:\n",
    "                if col == 'tissue':\n",
    "                    # Find what tissue type is at label row\n",
    "                    if sample_df.loc[sample_df[col] == label, 'tissue_type'].tolist()[0] != 'tissue':\n",
    "                        map_dict[label] = label # Don't map cell type in tissue\n",
    "                        continue\n",
    "                # Find what organism term id is at label row\n",
    "                org_term_id = sample_df.loc[sample_df[col] == label, 'organism_ontology_term_id'].tolist()[0]\n",
    "                if org_term_id in col_ont_map[col]:\n",
    "                    # Get ontology of specific organism and map label\n",
    "                    species_ont = col_ont_map[col][org_term_id]\n",
    "                    term_id = ontology_parser.get_term_id_by_label(label, species_ont)\n",
    "                else:\n",
    "                    if col_ont_map[col]['other'] == 'none':\n",
    "                        map_dict[label] = label\n",
    "                        continue\n",
    "                    else:\n",
    "                        term_id = ontology_parser.get_term_id_by_label(label, col_ont_map[col]['other'])\n",
    "            else:\n",
    "                term_id = ontology_parser.get_term_id_by_label(label, col_ont_map[col])\n",
    "            if term_id == None:\n",
    "                print(f'{label}\\t{col}\\t{term_id}\\t{org_term_id}')\n",
    "                if org_term_id:\n",
    "                    if org_term_id in col_ont_map[col]:\n",
    "                        ont_err_lst.append(f\"Error: Matching '{col_ont_map[col][org_term_id]}' term id not found for label '{label}' in column '{col}'\")\n",
    "                    else:\n",
    "                        ont_err_lst.append(f\"Error: Matching '{col_ont_map[col]['other']}' term id not found for label '{label}' in column '{col}'\")\n",
    "                else:\n",
    "                    ont_err_lst.append(f\"Error: Matching '{col_ont_map[col]}' term id not found for label '{label}' in column '{col}'\")\n",
    "                map_dict[label] = label\n",
    "                continue\n",
    "            map_dict[label] = term_id\n",
    "        sample_df[col + '_ontology_term_id'] = sample_df[col].map(map_dict)\n",
    "        del sample_df[col]\n",
    "    \n",
    "    ### Print out any errors from ontologizing\n",
    "    if ont_err_lst:\n",
    "        for e in ont_err_lst:\n",
    "            print(e)\n",
    "        sys.exit()\n",
    "\n",
    "    ### Convert string to boolean for is_pilot_data and donor_living_at_sample_collection\n",
    "    ### Check that donor_living_at_sample_collection is not filled out for non-human\n",
    "    b_type = ['is_pilot_data','donor_living_at_sample_collection']\n",
    "    for c in b_type:\n",
    "        if c in sample_df.columns:\n",
    "            if c == 'donor_living_at_sample_collection':\n",
    "                for val in sample_df[c].unique():\n",
    "                    if val != 'na' and sample_df.loc[sample_df[c] == val, \n",
    "                    'organism_ontology_term_id'].tolist()[0] != 'NCBITaxon:9606':\n",
    "                        print(f\"ERROR: donor_living_at_sample_collection for non-human data should be 'na' but '{val}' is present\")\n",
    "                        sys.exit()\n",
    "            sample_df[c] == sample_df[c].replace({'FALSE':False, 'TRUE':True})\n",
    "    \n",
    "    ### Blank fields in worksheet result in NaN values in dataframe, replacing these with na?\n",
    "    ### Could also replace with unknown for certain columns using fillna options?\n",
    "    sample_df.fillna('na', inplace=True)\n",
    "    sample_df.drop(columns=[c for c in sample_df.columns if c.startswith('!')], inplace=True)\n",
    "\n",
    "    return sample_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4e053c02-2877-48b3-baca-3a111f41aa7b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading in R109-25-0422\n",
      "Reading in R109-25-0423\n",
      "Reading in R109-25-0424\n",
      "Reading in R109-25-0425\n",
      "Reading in R109-25-0426\n",
      "Reading in R109-25-0427\n",
      "Reading in R109-25-0428\n",
      "Reading in R109-25-0429\n",
      "Reading in R109-25-0430\n",
      "Reading in R109-25-0431\n",
      "Reading in R109-25-0432\n",
      "Reading in R109-25-0433\n",
      "Reading in R109-25-0434\n"
     ]
    }
   ],
   "source": [
    "# Read in h5ad file, and assign scale hash oligo to each cell\n",
    "\n",
    "adatas = []\n",
    "for pseudosample in pseudosamples:\n",
    "    print(f'Reading in {pseudosample}')\n",
    "    adata = sc.read_h5ad(f'{down_dir}/{pseudosample}_anndata.h5ad')\n",
    "    adata = assign_oligo(adata, f'{down_dir}/{pseudosample}.merged.allCells.csv')\n",
    "    adatas.append(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "32aac163-8222-4882-b34e-db92b3755a36",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Curating:\tR096-25-0255.QSR-8\n",
      "Curating:\tR096-25-0256.QSR-8\n",
      "Curating:\tR096-25-0257.QSR-8\n",
      "Curating:\tR096-25-0258.QSR-8\n"
     ]
    }
   ],
   "source": [
    "# For each adata, determine which sample (scale_sample) it is for, then subset sample_df for:\n",
    "# - the corresponding sample (lab_sample)\n",
    "# - the experiment name of the SeqSuite run\n",
    "# Then do a merged based on hash_index and assigned_hash_index\n",
    "\n",
    "for adata in adatas:\n",
    "    scale_sample = adata.obs['sample'].unique()[0]\n",
    "    print(f'Curating:\\t{scale_sample}')\n",
    "    lab_sample = sample_map[scale_sample]\n",
    "    sample_df_subset = sample_df[(sample_df['pseudosample']==lab_sample) & (sample_df['experiment_name']== experiment_name)]\n",
    "    adata.obs = pd.merge(adata.obs, sample_df_subset, left_on='assigned_hash_index', right_on='hash_index', how='left').set_index(adata.obs.index)\n",
    "    adata.obs['is_primary_data'] = True\n",
    "\n",
    "    # Set uns metadata\n",
    "    adata.uns['organism_ontology_term_id'] = adata.obs['organism_ontology_term_id'].unique()[0]\n",
    "    adata.obs.drop(columns=['organism_ontology_term_id'], inplace=True)\n",
    "    prefix = f'{adata.obs[\"experiment_name\"].unique()[0]}__{adata.obs[\"sample\"].unique()[0]}'\n",
    "    adata.uns['title'] = prefix\n",
    "\n",
    "    # Write to file\n",
    "    out = 'curated_matrices'\n",
    "    if os.path.exists(out) == False:\n",
    "        os.mkdir(out)\n",
    "    adata.write(f'{out}/{prefix}.h5ad', compression=\"gzip\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6d4ad3c1-19cb-4da3-8954-68c4a658f36c",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "assay_ontology_term_id ['EFO:0022490', nan]\n",
      "\n",
      "\u001b[1m\u001b[31mERROR: cell_type_ontology_term_id not in obs\n",
      "\u001b[0m\n",
      "development_stage_ontology_term_id ['ZFS:0000035', nan]\n",
      "\n",
      "disease_ontology_term_id ['PATO:0000461', nan]\n",
      "\n",
      "self_reported_ethnicity_ontology_term_id ['na', nan]\n",
      "\n",
      "sex_ontology_term_id ['unknown', nan]\n",
      "\n",
      "tissue_ontology_term_id ['ZFA:0000103', nan]\n",
      "\n",
      "donor_id ['CHEM13_SCALE_E7_Bl1', 'CHEM13_SCALE_D8_Bl1', 'CHEM13_SCALE_G10_Bl1', 'CHEM13_SCALE_H1_Bl1', 'CHEM13_SCALE_A3_Bl1', 'CHEM13_SCALE_E9_Bl1', 'CHEM13_SCALE_H8_Bl1', 'CHEM13_SCALE_H3_Bl1', 'CHEM13_SCALE_B7_Bl1', 'CHEM13_SCALE_F2_Bl1', 'CHEM13_SCALE_D2_Bl1', 'CHEM13_SCALE_D3_Bl1', 'CHEM13_SCALE_C10_Bl1', 'CHEM13_SCALE_D1_Bl1', 'CHEM13_SCALE_E1_Bl1', 'CHEM13_SCALE_E8_Bl1', 'CHEM13_SCALE_A2_Bl1', 'CHEM13_SCALE_G8_Bl1', 'CHEM13_SCALE_F5_Bl1', 'CHEM13_SCALE_H10_Bl1', 'CHEM13_SCALE_B9_Bl1', 'CHEM13_SCALE_E6_Bl1', 'CHEM13_SCALE_G4_Bl1', 'CHEM13_SCALE_F3_Bl1', 'CHEM13_SCALE_C4_Bl1', 'CHEM13_SCALE_A1_Bl1', 'CHEM13_SCALE_D7_Bl1', 'CHEM13_SCALE_A4_Bl1', 'CHEM13_SCALE_B4_Bl1', 'CHEM13_SCALE_A9_Bl1', 'CHEM13_SCALE_B1_Bl1', 'CHEM13_SCALE_F7_Bl1', 'CHEM13_SCALE_D10_Bl1', 'CHEM13_SCALE_A6_Bl1', 'CHEM13_SCALE_D9_Bl1', 'CHEM13_SCALE_C9_Bl1', 'CHEM13_SCALE_B8_Bl1', 'CHEM13_SCALE_H6_Bl1', 'CHEM13_SCALE_E5_Bl1', 'CHEM13_SCALE_C3_Bl1', 'CHEM13_SCALE_H2_Bl1', 'CHEM13_SCALE_C1_Bl1', 'CHEM13_SCALE_C8_Bl1', 'CHEM13_SCALE_A5_Bl1', 'CHEM13_SCALE_H9_Bl1', 'CHEM13_SCALE_G6_Bl1', 'CHEM13_SCALE_E3_Bl1', 'CHEM13_SCALE_B2_Bl1', 'CHEM13_SCALE_C7_Bl1', 'CHEM13_SCALE_B5_Bl1', 'CHEM13_SCALE_H5_Bl1', 'CHEM13_SCALE_F4_Bl1', 'CHEM13_SCALE_H4_Bl1', 'CHEM13_SCALE_H7_Bl1', 'CHEM13_SCALE_F9_Bl1', 'CHEM13_SCALE_G2_Bl1', 'CHEM13_SCALE_G3_Bl1', 'CHEM13_SCALE_D4_Bl1', 'CHEM13_SCALE_E4_Bl1', 'CHEM13_SCALE_F1_Bl1', 'CHEM13_SCALE_D6_Bl1', 'CHEM13_SCALE_E2_Bl1', 'CHEM13_SCALE_B10_Bl1', 'CHEM13_SCALE_C6_Bl1', 'CHEM13_SCALE_G1_Bl1', 'CHEM13_SCALE_A7_Bl1', 'CHEM13_SCALE_F10_Bl1', 'CHEM13_SCALE_G5_Bl1', 'CHEM13_SCALE_A10_Bl1', 'CHEM13_SCALE_F8_Bl1', 'CHEM13_SCALE_G7_Bl1', 'CHEM13_SCALE_A8_Bl1', 'CHEM13_SCALE_D5_Bl1', 'CHEM13_SCALE_F6_Bl1', 'CHEM13_SCALE_B6_Bl1', 'CHEM13_SCALE_E10_Bl1', 'CHEM13_SCALE_C2_Bl1', 'CHEM13_SCALE_C5_Bl1', 'CHEM13_SCALE_B3_Bl1', nan, 'CHEM13_SCALE_G9_Bl1']\n",
      "\n",
      "suspension_type ['nucleus', nan]\n",
      "\n",
      "tissue_type ['tissue', nan]\n",
      "\n",
      "\u001b[1m\u001b[31mERROR: is_primary_data not in obs\n",
      "\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "sys.path.insert(1, '/home/jovyan/lattice-tools/cellxgene_resources')\n",
    "from cellxgene_mods import *\n",
    "evaluate_obs_schema(adatas[0].obs, labels=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "25701f1c-9a9f-41e5-8dc7-b8e7237a3e6c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
