{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cca26c9b-632f-4bed-a4bd-06b89cecf937",
   "metadata": {},
   "outputs": [],
   "source": [
    "import boto3\n",
    "import numpy as np\n",
    "import os\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import tarfile\n",
    "from cellxgene_schema.write_labels import AnnDataLabelAppender\n",
    "from urllib.parse import quote\n",
    "\n",
    "\n",
    "s3client = boto3.client('s3')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40f3f071-0b7c-4ef8-b7cb-49cdad9c153e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def download_files(s_dir):\n",
    "    mx_h5 = f'{s_dir}/count/sample_filtered_feature_bc_matrix.h5'\n",
    "    metrics_csv = f'{s_dir}/metrics_summary.csv'\n",
    "    cri_tar = f'{s_dir}/count/crispr_analysis.tar.gz'\n",
    "    \n",
    "    for file_path in [mx_h5, metrics_csv, cri_tar]:\n",
    "        f = file_path.split('/')[-1]\n",
    "        s3client.download_file(bucket_name, file_path, f)\n",
    "\n",
    "\n",
    "def custom_var_to_obs(adata):\n",
    "    moved = []\n",
    "    for gene_index in adata.var[\n",
    "        (adata.var['feature_types'] == 'Gene Expression') &\n",
    "        (adata.var['gene_ids'].str.startswith('ENS') == False)\n",
    "    ].index:\n",
    "        gene_id = adata.var.loc[gene_index]['gene_ids']\n",
    "        gene_values = adata.X[:, adata.var.index.get_loc(gene_index)]\n",
    "        adata.obs[gene_id] = gene_values.A.flatten() if hasattr(gene_values, \"A\") else gene_values.toarray().flatten().tolist()\n",
    "        moved.append(gene_id)\n",
    "\n",
    "    adata.var.set_index('gene_ids', inplace=True)\n",
    "    var_to_keep = [i for i in adata.var.index if i not in moved]\n",
    "    adata = adata[:, var_to_keep]\n",
    "    adata.var = adata.var.replace('', np.nan).dropna(axis=1, how='all')\n",
    "\n",
    "    return adata\n",
    "\n",
    "\n",
    "def gather_crispr():\n",
    "    tar = tarfile.open('crispr_analysis.tar.gz', 'r:gz')\n",
    "    f = tar.extractfile('protospacer_calls_per_cell.csv')\n",
    "    df = pd.read_csv(f).rename(columns={'num_umis':'num_umis_guide_id'})\n",
    "    tar.close()\n",
    "    df['genetic_perturbation_guide_id'] = df['feature_call'].apply(lambda x: x.replace('|',' || '))\n",
    "    df = df[['cell_barcode','genetic_perturbation_guide_id','num_umis_guide_id']].set_index('cell_barcode')\n",
    "\n",
    "    return df\n",
    "\n",
    "\n",
    "def gather_metrics(samp):\n",
    "    df = pd.read_csv('metrics_summary.csv')\n",
    "    df['Metric'] = df.apply(lambda x: f\"{x['Metric Name']}, {x['Library Type']}, {x['Category'].replace('Cells','Sample')}\", axis=1)\n",
    "\n",
    "    probe_barcodes = ' || '.join(df[\n",
    "        (df['Metric Name'] == 'Sample ID') &\n",
    "        (df['Metric Value'] == samp)\n",
    "    ]['Group Name'].unique())\n",
    "\n",
    "    df = df[\n",
    "        (df['Grouped By'].isin(['Fastq ID','Probe barcode ID']) == False) &\n",
    "        (df['Category'].isin(['Library','Cells']))\n",
    "    ]\n",
    "    df = df[['Metric','Metric Value']].set_index('Metric').transpose()\n",
    "\n",
    "    for c in df.columns:\n",
    "        v = df[c].iloc[0]\n",
    "        if v.endswith('%'):\n",
    "            df[c] = df[c].apply(lambda x: float(x.rstrip('%')) / 100)\n",
    "        elif '%' in v:\n",
    "            df.drop(columns=c, inplace=True)\n",
    "        else:\n",
    "            df[c] = df[c].apply(lambda x: int(x.replace(',','')))\n",
    "\n",
    "    df['Probe barcode IDs'] = probe_barcodes\n",
    "\n",
    "    return df\n",
    "\n",
    "\n",
    "def cxg_add_labels(adata):\n",
    "    adata.obs['cell_type_ontology_term_id'] = 'unknown'\n",
    "    labeler = AnnDataLabelAppender(adata)\n",
    "    labeler._add_labels()\n",
    "    adata.obs.drop(columns=['cell_type_ontology_term_id','cell_type'],inplace=True)\n",
    "\n",
    "    #IF DESIRED TO TRACK\n",
    "    schema_v = labeler.schema_version\n",
    "    adata.uns['schema_version'] = schema_v\n",
    "    adata.uns['schema_reference'] = labeler._build_schema_reference_url(schema_v)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dcca356f-0b0c-48cb-9ced-bc50f06f0678",
   "metadata": {},
   "outputs": [],
   "source": [
    "#read in the sample metadata\n",
    "sheet = ''\n",
    "tab = 'sample template'\n",
    "url = f'https://docs.google.com/spreadsheets/d/{sheet}/gviz/tq?tqx=out:csv&sheet={quote(tab)}'\n",
    "sample_df = pd.read_csv(url).dropna(axis=1,how='all')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6914a64-c8dd-4866-8a57-c7ca0f8746e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "bucket_name = 'czi-psomagen'\n",
    "lab = 'weissman'\n",
    "proj = 'scaling-in-vivo-perturb-seq-in-the-liver-and-beyond'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d78fdbfc-45b7-4069-a300-09aeb813f756",
   "metadata": {},
   "outputs": [],
   "source": [
    "my_dir = f'{lab}-{proj}/'\n",
    "r = s3client.list_objects(Bucket=bucket_name, Prefix=my_dir, Delimiter='/')\n",
    "orders = [o['Prefix'].replace(my_dir,'') for o in r['CommonPrefixes']]\n",
    "\n",
    "samples = []\n",
    "for o in orders:\n",
    "    r = s3client.list_objects(Bucket=bucket_name, Prefix=f'{my_dir}{o}', Delimiter='/')\n",
    "    libs = [l['Prefix'].replace(f'{my_dir}{o}','') for l in r['CommonPrefixes']]\n",
    "    for l in libs:\n",
    "        r = s3client.list_objects(Bucket=bucket_name, Prefix=f'{my_dir}{o}{l}processed/cellranger/', Delimiter='/')\n",
    "        dates = [d['Prefix'].replace(f'{my_dir}{o}{l}processed/cellranger/','') for d in r['CommonPrefixes']]\n",
    "        for d in dates:\n",
    "            r = s3client.list_objects(Bucket=bucket_name, Prefix=f'{my_dir}{o}{l}processed/cellranger/{d}outs/per_sample_outs/', Delimiter='/')\n",
    "            subs = [s['Prefix'].replace(f'{my_dir}{o}{l}processed/cellranger/{d}outs/per_sample_outs/','') for s in r['CommonPrefixes']]\n",
    "            for s in subs:\n",
    "                samples.append({\n",
    "                    'order': o,\n",
    "                    'library': l.rstrip('/'),\n",
    "                    'date': d,\n",
    "                    'sample': s.rstrip('/')\n",
    "                })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff24c2b5-49db-4f56-b6d1-afe038409ae3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#identify which sample to curate\n",
    "n = 0\n",
    "s = samples[n]\n",
    "s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35d329fa-62b9-49da-aedb-c844aec6407b",
   "metadata": {},
   "outputs": [],
   "source": [
    "order = s['order']\n",
    "lib = s['library']\n",
    "run_date = s['date']\n",
    "samp = s['sample']\n",
    "\n",
    "s_dir = f'{lab}-{proj}/{order}{lib}/processed/cellranger/{run_date}outs/per_sample_outs/{samp}'\n",
    "download_files(s_dir)\n",
    "\n",
    "adata = sc.read_10x_h5('sample_filtered_feature_bc_matrix.h5', gex_only=True)\n",
    "\n",
    "adata = custom_var_to_obs(adata)\n",
    "\n",
    "adata.obs['lab'] = lab #this can go away once it is no longer in the directory structure\n",
    "adata.obs['project'] = proj\n",
    "#ADD IN REFERENCE INFO - genome/annotation\n",
    "adata.obs['sample_name'] = samp\n",
    "adata.obs['library_id'] = lib\n",
    "adata.obs = adata.obs.merge(sample_df, on='sample_name', how='left').set_index(adata.obs.index)\n",
    "adata.uns['organism_ontology_term_id'] = adata.obs['organism_ontology_term_id'].unique()[0]\n",
    "adata.obs.drop(columns=['organism_ontology_term_id'],inplace=True)\n",
    "\n",
    "crispr_df = gather_crispr()\n",
    "adata.obs = adata.obs.merge(\n",
    "    crispr_df, left_index=True, right_index=True, how='left'\n",
    ").set_index(adata.obs.index)\n",
    "\n",
    "metrics_df = gather_metrics(samp)\n",
    "for c in metrics_df.columns:\n",
    "    adata.obs[c] = metrics_df[c].values[0]\n",
    "\n",
    "#ADD IN MITO %\n",
    "#ADD IN DOUBLETS\n",
    "#maybe https://scanpy.readthedocs.io/en/stable/generated/scanpy.pp.calculate_qc_metrics.html\n",
    "\n",
    "#add labels\n",
    "cxg_add_labels(adata)\n",
    "\n",
    "adata.write(filename=f'{samp}_curated.h5ad', compression='gzip')\n",
    "\n",
    "for f in ['sample_filtered_feature_bc_matrix.h5','crispr_analysis.tar.gz','metrics_summary.csv']:\n",
    "    os.remove(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec74612d-2493-4744-95ea-ca828e90f61c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
