{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cca26c9b-632f-4bed-a4bd-06b89cecf937",
   "metadata": {},
   "outputs": [],
   "source": [
    "import boto3\n",
    "import numpy as np\n",
    "import os\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import tarfile\n",
    "from cellxgene_schema.write_labels import AnnDataLabelAppender"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8abbeeb-c9a8-4733-b60a-ecb6453fdc69",
   "metadata": {},
   "outputs": [],
   "source": [
    "s3client = boto3.client('s3')\n",
    "paginator = s3client.get_paginator('list_objects')\n",
    "\n",
    "bucket_name = 'czi-psomagen'\n",
    "lab = 'Marson'\n",
    "proj = 'mapping-grns-perturb-seq'\n",
    "run_dates = [\n",
    "    'Run_2025-05-23',\n",
    "    'Run_2025-05-28',\n",
    "    'Run_2025-06-03'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40f3f071-0b7c-4ef8-b7cb-49cdad9c153e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def download_files(samp):\n",
    "    mx_h5 = f'{lab}/{proj}/processed/cellranger/{run_date}/{lib}/outs/per_sample_outs/{samp}/count/sample_filtered_feature_bc_matrix.h5'\n",
    "    metrics_csv = f'{lab}/{proj}/processed/cellranger/{run_date}/{lib}/outs/per_sample_outs/{samp}/metrics_summary.csv'\n",
    "    cri_tar = f'{lab}/{proj}/processed/cellranger/{run_date}/{lib}/outs/per_sample_outs/{samp}/count/crispr_analysis.tar.gz'\n",
    "    \n",
    "    for file_path in [mx_h5, metrics_csv, cri_tar]:\n",
    "        f = file_path.split('/')[-1]\n",
    "        s3client.download_file(bucket_name, file_path, f)\n",
    "\n",
    "\n",
    "def custom_var_to_obs(adata):\n",
    "    moved = []\n",
    "    for gene_index in adata.var[\n",
    "        (adata.var['feature_types'] == 'Gene Expression') &\n",
    "        (adata.var['gene_ids'].str.startswith('ENS') == False)\n",
    "    ].index:\n",
    "        gene_id = adata.var.loc[gene_index]['gene_ids']\n",
    "        gene_values = adata.X[:, adata.var.index.get_loc(gene_index)]\n",
    "        adata.obs[gene_id] = gene_values.A.flatten() if hasattr(gene_values, \"A\") else gene_values.toarray().flatten().tolist()\n",
    "        moved.append(gene_id)\n",
    "\n",
    "    adata.var.set_index('gene_ids', inplace=True)\n",
    "    var_to_keep = [i for i in adata.var.index if i not in moved]\n",
    "    adata = adata[:, var_to_keep]\n",
    "    adata.var = adata.var.replace('', np.nan).dropna(axis=1, how='all')\n",
    "\n",
    "    return adata\n",
    "\n",
    "\n",
    "def gather_crispr():\n",
    "    tar = tarfile.open('crispr_analysis.tar.gz', 'r:gz')\n",
    "    f = tar.extractfile('protospacer_calls_per_cell.csv')\n",
    "    df = pd.read_csv(f).rename(columns={'num_umis':'num_umis_guide_id'})\n",
    "    tar.close()\n",
    "    df['genetic_perturbation_guide_id'] = df['feature_call'].apply(lambda x: x.replace('|',' || '))\n",
    "    df = df[['cell_barcode','genetic_perturbation_guide_id','num_umis_guide_id']].set_index('cell_barcode')\n",
    "\n",
    "    return df\n",
    "\n",
    "\n",
    "def gather_metrics(samp):\n",
    "    df = pd.read_csv('metrics_summary.csv')\n",
    "    df['Metric'] = df.apply(lambda x: f\"{x['Metric Name']}, {x['Library Type']}, {x['Category'].replace('Cells','Sample')}\", axis=1)\n",
    "\n",
    "    probe_barcodes = ' || '.join(df[\n",
    "        (df['Metric Name'] == 'Sample ID') &\n",
    "        (df['Metric Value'] == samp)\n",
    "    ]['Group Name'].unique())\n",
    "\n",
    "    df = df[\n",
    "        (df['Grouped By'].isin(['Fastq ID','Probe barcode ID']) == False) &\n",
    "        (df['Category'].isin(['Library','Cells']))\n",
    "    ]\n",
    "    df = df[['Metric','Metric Value']].set_index('Metric').transpose()\n",
    "\n",
    "    for c in df.columns:\n",
    "        v = df[c].iloc[0]\n",
    "        if v.endswith('%'):\n",
    "            df[c] = df[c].apply(lambda x: float(x.rstrip('%')) / 100)\n",
    "        elif '%' in v:\n",
    "            df.drop(columns=c, inplace=True)\n",
    "        else:\n",
    "            df[c] = df[c].apply(lambda x: int(x.replace(',','')))\n",
    "\n",
    "    df['Probe barcode IDs'] = probe_barcodes\n",
    "\n",
    "    return df\n",
    "\n",
    "\n",
    "def cxg_add_labels(adata):\n",
    "    adata.obs['cell_type_ontology_term_id'] = 'unknown'\n",
    "    labeler = AnnDataLabelAppender(adata)\n",
    "    labeler._add_labels()\n",
    "    adata.obs.drop(columns=['cell_type_ontology_term_id','cell_type'],inplace=True)\n",
    "\n",
    "    #IF DESIRED TO TRACK, AS IT PERTAINS TO ADD-LABELS\n",
    "    schema_v = labeler.schema_version\n",
    "    adata.uns['schema_version'] = schema_v\n",
    "    adata.uns['schema_reference'] = labeler._build_schema_reference_url(schema_v)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dcca356f-0b0c-48cb-9ced-bc50f06f0678",
   "metadata": {},
   "outputs": [],
   "source": [
    "#read in the sample metadata\n",
    "#for now, it's stored in a Google Sheet\n",
    "sheet = '1VhASg4SaisOuZ3LUiTwkSkpaxdInBXquF0Lbkik5_os'\n",
    "tab = 'samples'\n",
    "url = f'https://docs.google.com/spreadsheets/d/{sheet}/gviz/tq?tqx=out:csv&sheet={tab}'\n",
    "sample_df = pd.read_csv(url).dropna(axis=1,how='all')\n",
    "\n",
    "#temp solution: replace any terms not in the CxG schema to avoid add-labels failure\n",
    "sample_df.replace({\n",
    "    'self_reported_ethnicity_ontology_term_id': {\n",
    "        'HANCESTRO:0612': 'HANCESTRO:0014'\n",
    "    }\n",
    "}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d78fdbfc-45b7-4069-a300-09aeb813f756",
   "metadata": {},
   "outputs": [],
   "source": [
    "# build lists of files, libraries, samples\n",
    "files = []\n",
    "for run_date in run_dates:\n",
    "    my_dir = f'{lab}/{proj}/processed/cellranger/{run_date}/'\n",
    "    for page in paginator.paginate(Bucket=bucket_name, Prefix=my_dir):\n",
    "        if 'Contents' in page:\n",
    "            for obj in page['Contents']:\n",
    "                files.append(obj['Key'])\n",
    "\n",
    "subs = {}\n",
    "for file_path in files:\n",
    "    file_path = file_path.split('/')\n",
    "    cri = file_path.index('cellranger')\n",
    "    datei = cri + 1\n",
    "    lib = file_path[datei + 1]\n",
    "    if len(file_path) > (datei + 3):\n",
    "        if file_path[datei + 3] == 'per_sample_outs':\n",
    "            sub = file_path[datei + 4]\n",
    "            subs[sub] = {\n",
    "                'library': lib,\n",
    "                'date': file_path[datei]\n",
    "            }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35d329fa-62b9-49da-aedb-c844aec6407b",
   "metadata": {},
   "outputs": [],
   "source": [
    "for samp,v in subs.items():\n",
    "    lib = v['library']\n",
    "    run_date = v['date']\n",
    "    print(samp)\n",
    "    download_files(samp)\n",
    "\n",
    "    adata = sc.read_10x_h5('sample_filtered_feature_bc_matrix.h5', gex_only=True)\n",
    "\n",
    "    adata = custom_var_to_obs(adata)\n",
    "\n",
    "    adata.obs['lab'] = lab #this can go away once it is no longer in the directory structure\n",
    "    adata.obs['project'] = proj\n",
    "    #ADD IN REFERENCE INFO - genome/annotation\n",
    "    adata.obs['sample_id'] = samp\n",
    "    adata.obs['library_id'] = lib\n",
    "    adata.obs = adata.obs.merge(sample_df, on='sample_id', how='left').set_index(adata.obs.index)\n",
    "    #STARTING 6.0.0\n",
    "    #adata.uns['organism_ontology_term_id'] = adata.obs['organism_ontology_term_id'].unique()[0]\n",
    "    #adata.obs.drop(columns=['organism_ontology_term_id'],inplace=True)\n",
    "\n",
    "    crispr_df = gather_crispr()\n",
    "    adata.obs = adata.obs.merge(\n",
    "        crispr_df, left_index=True, right_index=True, how='left'\n",
    "    ).set_index(adata.obs.index)\n",
    "\n",
    "    metrics_df = gather_metrics(samp)\n",
    "    for c in metrics_df.columns:\n",
    "        adata.obs[c] = metrics_df[c].values[0]\n",
    "\n",
    "    #ADD IN MITO %\n",
    "    #ADD IN DOUBLETS\n",
    "    #maybe https://scanpy.readthedocs.io/en/stable/generated/scanpy.pp.calculate_qc_metrics.html\n",
    "\n",
    "    #add labels - somehow\n",
    "    cxg_add_labels(adata)\n",
    "\n",
    "    #adata.write(filename=f'{samp}_curated.h5ad', compression='gzip')\n",
    "    #adata.write_zarr(f'{samp}_curated.zarr')\n",
    "    #WHAT TO NAME IT? WHERE TO SEND IT?\n",
    "\n",
    "    for f in ['sample_filtered_feature_bc_matrix.h5','crispr_analysis.tar.gz','metrics_summary.csv']:\n",
    "        os.remove(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec74612d-2493-4744-95ea-ca828e90f61c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
