{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "112b7dbb-1d5e-46de-9b1a-884727d5ec6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import boto3\n",
    "import json\n",
    "import os\n",
    "import re\n",
    "from bs4 import BeautifulSoup"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de4e5b49-7fb5-41f9-b207-eafffa767ab1",
   "metadata": {},
   "source": [
    "**[Cell Ranger multi](https://www.10xgenomics.com/support/software/cell-ranger/latest/analysis/outputs/cr-3p-outputs-cellplex)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "f0053339-d821-43d9-91b2-d0c6afb7dca5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#outs/\n",
    "multi_outs = [\n",
    "    'config.csv',\n",
    "    'multi/count/feature_reference.csv',\n",
    "    'multi/count/raw_cloupe.cloupe',\n",
    "    'multi/count/raw_feature_bc_matrix.h5',\n",
    "    'multi/count/raw_feature_bc_matrix.tar.gz',\n",
    "    'multi/count/raw_molecule_info.h5',\n",
    "    'multi/count/unassigned_alignments.bam',\n",
    "    'multi/count/unassigned_alignments.bam.bai',\n",
    "    'multi/multiplexing_analysis.tar.gz'\n",
    "]\n",
    "\n",
    "#outs/per_sample_outs/<sample_id>/\n",
    "multi_per_sample_outs = [\n",
    "    'count/analysis.tar.gz',\n",
    "    'count/feature_reference.csv',\n",
    "    'count/sample_cloupe.cloupe',\n",
    "    'count/sample_alignments.bam',\n",
    "    'count/sample_alignments.bam.bai',\n",
    "    'count/sample_filtered_barcodes.csv',\n",
    "    'count/sample_filtered_feature_bc_matrix.h5',\n",
    "    'count/sample_filtered_feature_bc_matrix.tar.gz',\n",
    "    'count/sample_molecule_info.h5',\n",
    "    'metrics_summary.csv',\n",
    "    'web_summary.html'\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0f372f6-91c8-4765-a3ee-ab4f777c68ea",
   "metadata": {},
   "source": [
    "**[Cell Ranger multi - 10x Flex](https://www.10xgenomics.com/support/software/cell-ranger/latest/analysis/outputs/cr-flex-outputs-frp)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ca7a2d7-7bf0-4ec1-bb38-c5d918c6dd39",
   "metadata": {},
   "outputs": [],
   "source": [
    "#outs/\n",
    "multi_flex_outs = [\n",
    "    'config.csv',\n",
    "    'multi/count/raw_cloupe.cloupe',\n",
    "    'multi/count/raw_feature_bc_matrix.h5',\n",
    "    'multi/count/raw_feature_bc_matrix.tar.gz',\n",
    "    'multi/count/raw_molecule_info.h5',\n",
    "    'multi/count/raw_probe_bc_matrix.h5',\n",
    "    'multi/multiplexing_analysis.tar.gz'\n",
    "]\n",
    "\n",
    "#outs/per_sample_outs/<sample_id>/\n",
    "multi_flex_per_sample_outs = [\n",
    "    'count/analysis.tar.gz',\n",
    "    'count/probe_set.csv',\n",
    "    'count/sample_cloupe.cloupe',\n",
    "    'count/sample_filtered_barcodes.csv',\n",
    "    'count/sample_filtered_feature_bc_matrix.h5',\n",
    "    'count/sample_filtered_feature_bc_matrix.tar.gz',\n",
    "    'count/sample_molecule_info.h5',\n",
    "    'count/sample_raw_feature_bc_matrix.h5',\n",
    "    'count/sample_raw_feature_bc_matrix.tar.gz',\n",
    "    'count/sample_raw_probe_bc_matrix.h5',\n",
    "    'metrics_summary.csv',\n",
    "    'web_summary.html'\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d534275-cdd9-421a-9ee8-5289951556df",
   "metadata": {},
   "source": [
    "**[Cell Ranger count](https://www.10xgenomics.com/support/software/cell-ranger/latest/analysis/outputs/cr-outputs-gex-overview)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2013a802-0df8-444e-96b7-a5ffe131625d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#outs/\n",
    "count_outs = [\n",
    "    'analysis.tar.gz',\n",
    "    'cloupe.cloupe',\n",
    "    'filtered_feature_bc_matrix.tar.gz',\n",
    "    'filtered_feature_bc_matrix.h5',\n",
    "    'metrics_summary.csv',\n",
    "    'molecule_info.h5',\n",
    "    'possorted_genome_bam.bam',\n",
    "    'possorted_genome_bam.bam.bai',\n",
    "    'raw_feature_bc_matrix.tar.gz',\n",
    "    'raw_feature_bc_matrix.h5',\n",
    "    'web_summary.html'\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6237bddf-2d8c-4e2f-a0d0-a5459ff59fb0",
   "metadata": {},
   "source": [
    "**[Cell Ranger ARC (multiome)](https://www.10xgenomics.com/support/software/cell-ranger-arc/latest/analysis/outputs/understanding-output)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b646092-0d46-4b99-835a-3418d5055d12",
   "metadata": {},
   "outputs": [],
   "source": [
    "#outs/\n",
    "multiome_outs = [\n",
    "    'analysis.tar.gz',\n",
    "    'cloupe.cloupe',\n",
    "    'filtered_feature_bc_matrix.tar.gz',\n",
    "    'filtered_feature_bc_matrix.h5',\n",
    "    'summary.csv',\n",
    "    'per_barcode_metrics.csv',\n",
    "    'gex_molecule_info.h5',\n",
    "    'gex_possorted_bam.bam',\n",
    "    'gex_possorted_bam.bam.bai',\n",
    "    'raw_feature_bc_matrix.tar.gz',\n",
    "    'raw_feature_bc_matrix.h5',\n",
    "    'web_summary.html',\n",
    "    'atac_possorted_bam.bam',\n",
    "    'atac_possorted_bam.bam.bai',\n",
    "    'atac_peaks.bed',\n",
    "    'atac_peak_annotation.tsv',\n",
    "    'atac_cut_sites.bigwig',\n",
    "    'atac_fragments.tsv.gz',\n",
    "    'atac_fragments.tsv.gz.tbi'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d09a65a-707b-4cd4-8f84-a88996579a8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#trying to enumerate the possible \"chemistry\" values from the web_summary.html\n",
    "non_flex_chem = [\n",
    "    \"Single Cell 5' R2-only v3\",\n",
    "    \"Single Cell 5' R2-only\",\n",
    "    \"Single Cell 3' v4 (polyA)\",\n",
    "    \"Single Cell 3' v3\"\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a628229-cda9-4f5e-94ee-2d9a161fc272",
   "metadata": {},
   "source": [
    "**[ScaleRna](https://github.com/ScaleBio/ScaleRna/blob/master/docs/outputs.md)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10552ff4-56e6-469f-bb83-2e6b573989ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "#ScaleRna.out/\n",
    "scale_rna = [\n",
    "    'reports/multiqc_report.html',\n",
    "    'reports/<sample>.<libIndex2>.report.html',\n",
    "    'reports/<sample>_libraries/', #For QuantumScale runs, need files within\n",
    "    'reports/allSamples.reportStatistics.csv',\n",
    "    'reports/csv/', #need files within\n",
    "    'reports/library/library_<libIndex2>.report.html',\n",
    "    'reports/library/csv/', #need files within\n",
    "    'samples/<sample>.<libIndex2>.filtered.matrix/', #need files within\n",
    "    'samples/<sample>.<libIndex2>.allCells.csv',\n",
    "    'samples/<sample>_libraries/', #For QuantumScale runs, need files within\n",
    "    'fastq/fastqc/*_fastqc.html',\n",
    "    'fastq/Reports/', #need files within\n",
    "    'barcodes/split_bcparser_jobs/bcparser.<libIndex2>/<sample>.bam',\n",
    "    'barcodes/<libIndex2>.metrics.json',\n",
    "    'alignment/<sample>.<libIndex2>/<sample>.star.align/', #need files within\n",
    "    'alignment/<sample>.<libIndex2>/<sample>.star.solo/' #need files within\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac77a191-360a-4e78-a9c2-f9857b0a3c51",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_web_summ(f):\n",
    "    with open(f) as html_doc:\n",
    "        soup = BeautifulSoup(html_doc, 'html.parser')\n",
    "    for x in soup.find_all('script'):\n",
    "        match = re.search(\"const data = \", x.string)\n",
    "        if match:\n",
    "            end = match.end()\n",
    "            data = json.loads(x.string[end:])\n",
    "\n",
    "    extra = []\n",
    "    if 'library' in data:\n",
    "        sub = data['sample']['subcommand']\n",
    "    \n",
    "        gex_tab = {row[0]:row[1] for row in data['library']['data']['gex_tab']['content']['parameters_table']['rows']}\n",
    "        chem = gex_tab['Chemistry']\n",
    "\n",
    "        if data['library']['data']['crispr_tab']:\n",
    "            extra.append('CRISPR')\n",
    "        if data['library']['data']['antibody_tab']:\n",
    "            extra.append('Antibody')\n",
    "    \n",
    "        #location of some additional info to QA\n",
    "        ref = gex_tab['Transcriptome']\n",
    "        if chem != 'Flex Gene Expression':\n",
    "            incl_int = gex_tab['Include Introns']\n",
    "        cr_v = data['library']['data']['header_info']['Pipeline Version']\n",
    "        if cr_v != 'cellranger-9.0.1':\n",
    "            print(f'ERROR:version is {cr_v} but should be 9.0.1')\n",
    "\n",
    "    else:\n",
    "        sub = None\n",
    "        info = {row[0]:row[1] for row in data['joint_pipeline_info_table']['rows']}\n",
    "        chem = info['Chemistry']\n",
    "\n",
    "        #location of some additional info to QA\n",
    "        ref = info['Reference path']\n",
    "        cr_v = info['Pipeline version']\n",
    "        if cr_v != 'cellranger-arc-2.0.2':\n",
    "            print(f'ERROR:version is {cr_v} but should be 2.0.2')\n",
    "        \n",
    "\n",
    "    return sub,chem,extra"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "504e0a04-906f-47f7-81f9-4d492fa2ab8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "s3client = boto3.client('s3')\n",
    "paginator = s3client.get_paginator('list_objects')\n",
    "\n",
    "\n",
    "# define scope of QA\n",
    "bucket_name = 'czi-psomagen'\n",
    "lab = 'Marson'\n",
    "proj = 'mapping-grns-perturb-seq'\n",
    "run_dates = [\n",
    "    'Run_2025-05-23',\n",
    "    'Run_2025-05-28',\n",
    "    'Run_2025-06-03'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b83516e6-35f2-4db1-8e16-1e56bdd5e7f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# build lists of files\n",
    "files = []\n",
    "for run_date in run_dates:\n",
    "    my_dir = f'{lab}/{proj}/processed/cellranger/{run_date}/'\n",
    "    for page in paginator.paginate(Bucket=bucket_name, Prefix=my_dir):\n",
    "        if 'Contents' in page:\n",
    "            for obj in page['Contents']:\n",
    "                files.append(obj['Key'])\n",
    "\n",
    "# build a list of libraries and their samples & run dates\n",
    "libs = {}\n",
    "for file_path in files:\n",
    "    file_path = file_path.split('/')\n",
    "    cri = file_path.index('cellranger')\n",
    "    datei = cri + 1\n",
    "    lib = file_path[datei + 1]\n",
    "    if lib not in libs:\n",
    "        libs[lib] = {\n",
    "            'date': file_path[datei],\n",
    "            'samples': []\n",
    "        }\n",
    "    if len(file_path) > (datei + 3):\n",
    "        if file_path[datei + 3] == 'per_sample_outs':\n",
    "            sub = file_path[datei + 4]\n",
    "            if sub not in libs[lib]['samples']:\n",
    "                libs[lib]['samples'].append(sub)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6a952ed-0783-4e9e-b4a5-12b44848bd80",
   "metadata": {},
   "outputs": [],
   "source": [
    "for lib,v in libs.items():\n",
    "    print(lib)\n",
    "    run_date = v['date']\n",
    "    samples = v['samples']\n",
    "    outs_path = f'{lab}/{proj}/processed/cellranger/{run_date}/{lib}/outs/'\n",
    "    file_path = f'{outs_path}per_sample_outs/{samples[0]}/web_summary.html'\n",
    "    f = file_path.split('/')[-1]\n",
    "    s3client.download_file(bucket_name, file_path, f)\n",
    "\n",
    "    sub,chem,extra = parse_web_summ(f)\n",
    "\n",
    "    if chem == 'Single Cell Multiome ATAC + Gene Expression v1':\n",
    "        outs_expected = multiome_outs\n",
    "        per_samp_expected = []\n",
    "\n",
    "    else:\n",
    "        if sub == 'multi':\n",
    "            if chem == 'Flex Gene Expression':\n",
    "                outs_expected = multi_flex_outs.copy()\n",
    "                per_samp_expected = multi_flex_per_sample_outs.copy()\n",
    "            elif chem in non_flex_chem:\n",
    "                outs_expected = multi_outs.copy()\n",
    "                per_samp_expected = multi_per_sample_outs.copy()\n",
    "        elif sub == 'count':\n",
    "            outs_expected = count_outs.copy()\n",
    "            per_samp_expected = []\n",
    "\n",
    "        #https://www.10xgenomics.com/support/software/cell-ranger/latest/analysis/outputs/cr-outputs-ab-overview\n",
    "        #https://www.10xgenomics.com/support/software/cell-ranger/latest/analysis/outputs/cr-outputs-crispr-overview\n",
    "        if 'CRISPR' in extra or 'Antibody' in extra:\n",
    "            outs_expected.append('multi/count/feature_reference.csv')\n",
    "            if sub == 'multi':\n",
    "                per_samp_expected.append('count/feature_reference.csv')\n",
    "            if 'CRISPR' in extra:\n",
    "                if sub == 'count':\n",
    "                    outs_expected.append('crispr_analysis.tar.gz')\n",
    "                if sub == 'multi':\n",
    "                    per_samp_expected.append('count/crispr_analysis.tar.gz')\n",
    "            if 'Antibody' in extra:\n",
    "                if sub == 'count':\n",
    "                    outs_expected.append('antibody_analysis.tar.gz')\n",
    "                if sub == 'multi':\n",
    "                    per_samp_expected.append('count/antibody_analysis.tar.gz')\n",
    "\n",
    "        #https://www.10xgenomics.com/support/software/cell-ranger/latest/analysis/running-pipelines/cr-cell-annotation-pipeline\n",
    "        for line in data['experimental_design']['csv'].split('\\n'):\n",
    "            if ',' in line:\n",
    "                path = line.strip().split(',')\n",
    "                if path[0] == 'skip-cell-annotation' and path[1] == 'false':\n",
    "                    per_samp_expected.append('count/cell_types.tar.gz')\n",
    "                    if sub == 'count':\n",
    "                        outs_expected.append('count/web_summary_cell_types.html')\n",
    "\n",
    "    actual = [f for f in files if f.startswith(outs_path)]\n",
    "\n",
    "    #check multi/lib\n",
    "    missing = [f for f in outs_expected if f'{outs_path}{f}' not in actual]\n",
    "    if missing:\n",
    "        print('missing')\n",
    "        print(missing)\n",
    "    extra = [f for f in actual if f.replace(outs_path,'') not in outs_expected and 'per_sample_outs' not in f]\n",
    "    if extra:\n",
    "        print('extra')\n",
    "        print(extra)\n",
    "\n",
    "    #check per sample\n",
    "    if per_samp_expected:\n",
    "        for s in samples:\n",
    "            print(s)\n",
    "            samp_outs_path = f'{outs_path}per_sample_outs/{s}/'\n",
    "            missing = [f for f in per_samp_expected if f'{samp_outs_path}{f}' not in actual]\n",
    "            if missing:\n",
    "                print('missing')\n",
    "                print(missing)\n",
    "            extra = [f for f in actual if f.replace(samp_outs_path,'') not in per_samp_expected and f.startswith(samp_outs_path)]\n",
    "            if extra:\n",
    "                print('extra')\n",
    "                print(extra)\n",
    "\n",
    "    os.remove(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eae92cb5-61a8-4716-a45f-8f74cbac9260",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
