{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2b8b98ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "# https://github.com/chanzuckerberg/single-cell-curation/issues/516\n",
    "# https://github.com/chanzuckerberg/single-cell-curation/blob/main/schema/4.0.0/schema.md#feature_length\n",
    "\n",
    "import numpy as np\n",
    "import os\n",
    "import scanpy as sc\n",
    "import subprocess\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75849952",
   "metadata": {},
   "outputs": [],
   "source": [
    "def validate(input_file, output_file):\n",
    "    ''' \n",
    "    Input: h5ad file\n",
    "    Output: h5ad file with additional var metadata fields\n",
    "    \n",
    "    '''\n",
    "    validate_process = subprocess.run(['cellxgene-schema', 'validate', '--add-labels', input_file, output_file], stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
    "    for line in validate_process.stdout.decode('utf-8').split('\\n'):\n",
    "        print(line)\n",
    "    for line in validate_process.stderr.decode('utf-8').split('\\n'):\n",
    "        print(line)\n",
    "        if 'is_valid=' in line:\n",
    "            valid = line.split('=')[-1]\n",
    "            print(valid)\n",
    "            return output_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19365f72",
   "metadata": {},
   "outputs": [],
   "source": [
    "def save(adata):\n",
    "    adata.write(filename='test.h5ad')\n",
    "    adata = sc.read_h5ad('test.h5ad')\n",
    "    if ('feature_biotype' not in adata.var.columns) and ('feature_biotype' not in adata.var.columns):\n",
    "        print('feature_length and feature_biotype are absent - CELLxGENE should annotate these two var fields')\n",
    "    \n",
    "    elif ('feature_length' in adata.var.columns) and ('feature_biotype' in adata.var.columns):\n",
    "        print(adata.var[['feature_biotype','feature_length']].info())\n",
    "        print('-------------------')\n",
    "        print(adata.var[['feature_length','feature_biotype']].value_counts())\n",
    "    \n",
    "    else: \n",
    "        print('feature_biotype or feature_length is present')\n",
    "\n",
    "    print('------------------')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19d531be",
   "metadata": {},
   "outputs": [],
   "source": [
    "def concat_expected_length_dfs(list_of_csvs):\n",
    "    df_list = []\n",
    "    for c in list_of_csvs:\n",
    "        df = pd.read_csv(c,header=None)\n",
    "        df_list.append(df)\n",
    "    \n",
    "    return pd.concat(df_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "9fbf9679",
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_var_metadata(output_file, ensg_id_to_check_list, list_of_csvs):\n",
    "    ''' \n",
    "    Input: 1) h5ad validate output file, 2) list of ensembl gene ids to check, 3) list of csvs with expected gene lengths \n",
    "    Output: Dataframe of var metadata checks of both adata.raw.var and adata.var for feature_biotype and feature_length\n",
    "    '''\n",
    "\n",
    "    adata = sc.read_h5ad(output_file)\n",
    "    gene_df = concat_expected_length_dfs(list_of_csvs)\n",
    "    gene_df.set_index(0, inplace=True)\n",
    "    result_dict = {}\n",
    "\n",
    "    for e in ensg_id_to_check_list:\n",
    "        result_list = []\n",
    "        expected_length = gene_df.loc[e, 3]\n",
    "        result_list.append(expected_length)\n",
    "        if adata.raw.var:\n",
    "            print('Checking adata.raw.var')\n",
    "            print('---------------------------')\n",
    "            if ('feature_biotype' in adata.raw.var.columns) and ('feature_length' in adata.raw.var.columns):\n",
    "                print('feature_biotype and feature_length are present in raw.var fields')\n",
    "                print('------------------------------------------------------------')\n",
    "                print('Checking feature_length of: ', e)\n",
    "                biotype = adata.raw.var.loc[e, 'feature_biotype']\n",
    "                if biotype == 'gene':\n",
    "                    print('biotype = gene')\n",
    "                    result_list.append(biotype)\n",
    "                    length = adata.raw.var.loc[e, 'feature_length']\n",
    "                    result_list.append(length)\n",
    "                    if length == expected_length:\n",
    "                        print(f'feature_length {length} is the same as expected {expected_length}')\n",
    "                        result_list.append(True)\n",
    "                    else:\n",
    "                        result_list.append(False)\n",
    "                else:\n",
    "                    print(f'biotype is {biotype}')\n",
    "            \n",
    "\n",
    "            elif ('feature_biotype' in adata.raw.var.columns) and ('feature_length' not in adata.raw.var.columns):\n",
    "                print('raw.var exists however feature_length is missing')\n",
    "\n",
    "            else:\n",
    "                print('raw.var exists however fields are missing')\n",
    "\n",
    "        else:\n",
    "            print('Checking adata.var')\n",
    "            print('---------------------------')\n",
    "            if ('feature_biotype' in adata.var.columns) and ('feature_length' in adata.var.columns):\n",
    "                print('feature_biotype and feature_length are present in var fields')\n",
    "                print('------------------------------------------------------------')\n",
    "                print('Checking feature_length of: ', e)\n",
    "                biotype = adata.var.loc[e, 'feature_biotype']\n",
    "                result_list.append(biotype)\n",
    "                if biotype == 'gene':\n",
    "                    print('feature_biotype = gene')\n",
    "                    length = adata.var.loc[e, 'feature_length']\n",
    "                    result_list.append(length)\n",
    "                    if length == expected_length:\n",
    "                        print(f'feature_length {length} is the same as expected {expected_length}')\n",
    "                        result_list.append(True)\n",
    "                    else:\n",
    "                        result_list.append(False)\n",
    "\n",
    "            elif ('feature_biotype' in adata.var.columns) and ('feature_length' not in adata.var.columns):\n",
    "                print('feature_length is missing')\n",
    "\n",
    "            else:\n",
    "                print('fields are missing')\n",
    "\n",
    "        result_dict[e] = result_list\n",
    "    return pd.DataFrame(result_dict, columns=['ensg_id', 'expected_length','feature_biotype','feature_length','matching_lengths'])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8caeb0ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "directory = '/Users/corinnsmall/GitClones/CZI/single-cell-curation/cellxgene_schema_cli/cellxgene_schema/ontology_files/'\n",
    "expected_df = concat_expected_length_dfs([directory + 'genes_homo_sapiens.csv.gz', directory + 'genes_mus_musculus.csv.gz', directory + 'genes_sars_cov_2.csv.gz'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "92a073af",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad('../valid.h5ad') #backed='r' would be slightly quicker but produces an error with multiple writes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e0b17ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.raw.var"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c289720",
   "metadata": {},
   "source": [
    "### Test info\n",
    "- feature_length will be a new var field that the portal will add to each dataset upon submission: Annotator =\tCELLxGENE Discover\n",
    "\n",
    "- the --add-labels option for cellxgene-schema should be adding this too. \n",
    "\n",
    "- validation will be to run --add-labels to a .h5ad, then open the resulting .h5ad and do some checks:\n",
    "\n",
    "    1. make sure var.feature_length is present\n",
    "\n",
    "    2. if thereâ€™s a raw layer, make sure raw.var.feature_length is present\n",
    "\n",
    "    3. then do some checks to make sure the lengths filled in are expected (we may need to pull some genes from v38 GENCODE gtf to use as checks here)\n",
    "\n",
    "        - Lengths depend on feature_biotype = 'gene' or 'spike-in'\n",
    "        - Gene lengths are calculated by creating non-overlapping concatenated exons across all isoforms of the gene, and then adding up their length in base-pairs.\n",
    "        - Testing both var and raw.var\n",
    "\n",
    "        <br> Valid cases:\n",
    "        1. gene = summed length of non-overlapping concatenated exons across all isoforms of the gene\n",
    "           \n",
    "        2. spike-in = 0\n",
    "\n",
    "        <br> Invalid cases:\n",
    "        \n",
    "        1. when feature_biotype = `gene` -> is gene length calculated as expected?\n",
    "            - test when there are overlapping exons of various lengths (isoforms):\n",
    "                - using gene found in v38 gencode gtf with multiple exons that vary in length\n",
    "            \n",
    "            - does it matter what strands exons are found on?\n",
    "            \n",
    "            - test null values?\n",
    "\n",
    "            - test datatypes: str, float\n",
    "            - any investigation into the use of cds vs exons?\n",
    "\n",
    "        2. when feature_biotype = `spike-in` -> is length calculated as 0?\n",
    "            - test non-zero value\n",
    "            - test null value\n",
    "            - test datatypes: str, float\n",
    "            \n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
